---
title: "New_analysis_piRNA"
author: "Konstantinos G"
date: "January 17, 2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# install BiocManager                                                       
```{r}
if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()
```

```{m packages}
install.packages("tidyverse")
install.packages("data.table")
install.packages("reshape2")
install.packages("bit64")
install.packages("ggseqlogo")


BiocManager::install(c("ComplexHeatmap", "AnnotationHub", "rtracklayer","GenomicAlignments", "Biostrings", "SummarizedExperiment", "IRanges", "edgeR", "Glimma", "limma", "DESeq2", "GenomicRanges", "rafalib", "biomaRt", "BSgenome", "ggbio","ComplexHeatmap","circlize", "Rsubread","TxDb.Hsapiens.UCSC.hg38.knownGene", "seqLogo","NOISeq", "Homo.sapiens", "stephenturner/annotables", "clusterProfiler","gridExtra","ggpmisc","wesanderson"))

library(devtools)
install_github("vqv/ggbiplot")
#[]
```

# data merging                                                              
```{r}
library(data.table);library(tidyverse)
# 1. First create a new file with noAnnot from *_output.txt                 ----
setwd("/home/rstudio/Documents/SPORTS_results/Result_furlan_with_DBs/") # CHANGE this
outfiles <-list.files(pattern = "_output.txt",recursive = TRUE ) # CHANGE this

for (file in outfiles){

gsub("_output.txt","",basename(file))->ofile
  
  as_tibble(fread(file,header = TRUE)) %>%
  filter(Match_Genome=="Yes") %>% 
  filter(Annotation=="NO_Annotation") %>%
  dplyr::transmute(V1=ID,V2=Reads,V3= Length,
                   V4=Sequence,V5=Match_Genome, V6=Sequence)-> DT
  
patt <- paste0(gsub(paste0(ofile,"_result/",basename(file)),
                    "",file),ofile,"_processed/",ofile,"_output_noAnnot_match_genome")

pat2 <- paste0(gsub(basename(file),"",file),ofile,"_summary.txt")

as_tibble(fread(pat2,header = TRUE)) %>%
  filter(Class=="Unannotated_Match_Genome") -> isit
stopifnot(identical(as.numeric(isit$Reads),as.numeric(sum(DT$V2))))
  fwrite(DT,file = patt,sep = "\t",col.names = FALSE)
  }

#-------------------Start Here--------------------------------------------------
#--------------------Creating DIRECTORIES---------------------------------------
fileout <- "furlan" # CHANGE this
folders <- c("Data","PNF_plots","Outputs")
lapply(paste0("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/",
              fileout,"/",folders),dir.create)

#------------find the files------------------------------------------
setwd("/media/sf_R_virtual_shared_folder/furlan/")# CHANGE this

#list.files(path = "/media/sf_R_virtual_shared_folder/", pattern = ".fastq", recursive = TRUE)

temp <- list.files(pattern = ".fastq", recursive = TRUE)

gsub(".fastq","_processed",basename(temp))->temp1

gsub(".fastq","",basename(temp))->temp

setwd("/home/rstudio/Documents/SPORTS_results/Result_furlan_with_DBs/")# change this

output <- list.files(pattern = "output_(ensembl|miRNA|piRNA|rfam|rRNA|tRNA|tRNA_mature|noAnnot)_match_genome$", recursive = TRUE)
(myfiles <- lapply(temp1,function(x)output[grepl(x,output)]))

#-------merge the files for small RNAs from different DBs-------------

for (i in seq(along=myfiles)){
DT <- rbindlist(sapply(myfiles[[i]],fread,
                        simplify=FALSE,select = c(1,2,4,6),
                        verbose=getOption("datatable.verbose", TRUE)),
                 use.names= TRUE,idcol="RNA_type") 

DT$RNA_type <- gsub(paste0(temp[i],"_output_"),"",
                    gsub("_genome","",basename(DT$RNA_type),perl=TRUE),perl=TRUE)

print(myfiles[[i]])

write_delim(DT,
            path = paste0("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/furlan/Data/",temp[i],"_smallRNA.txt")) #-------CHANGE this!-DATA DIR-^^^^^
}


```

# Search for nucleotide bias percentages in 1st and 10th base               
```{r}
library(data.table);library(tidyverse);library(Biostrings);library(wesanderson)

#######-----------------Create dirs--------------------------------
setwd("/home/rstudio/Documents/new_analysis_workflow/1_5_files_noDBs/")# change this

fileout <- list.files()
folders <- c("Data","PNF_plots")
lapply(paste0("/home/rstudio/Documents/new_analysis_workflow/1_5_files_noDBs/",# change this
              fileout,"/",folders),dir.create)

######--------- copy the outputs to new workflow--------------------------
setwd("/home/rstudio/Documents/SPORTS_results/Result_IPP_PIWIL1_noDBs/")
filesn <- list.files(pattern = "_output.txt", recursive = TRUE)
outfol="/home/rstudio/Documents/new_analysis_workflow/1_5_files_noDBs/3_IPP_PIWIL1_no_DBs/Outputs/"
input=getwd()
file.copy(file.path(input,filesn),outfol, overwrite = "TRUE",
          recursive=TRUE)

#----------------------- Start here--------------------------------------------
setwd("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/")# change this
filesnames<- list.files(pattern = "_output.txt", recursive = TRUE,full.names =TRUE)
for (filename in filesnames) {
  filesnames1 <- gsub("_output.txt","",basename(filename))
  dirname <- gsub(paste0("Outputs/",filesnames1,"_output.txt"),"",filename)
  #-------use the total output of SPORTS1.0--------------------------

  DT <- fread(filename,header = TRUE)
  
  #------subset by annotation piRNA and match genome----------------------
  DT  <- DT[Annotation=="piRNA" & Match_Genome=="Yes",] 
  
  b <- BStringSet(DT$Sequence,start = 1)
  
  conMat <- as_tibble(t(consensusMatrix(b, baseOnly=TRUE)))

  #-----create position column and then melt data ---------------------------

  conMat$position <- c(1:nrow(conMat))
  #
  conMat <- conMat %>% 
    gather(key=base,value=freq,-position) 

  
  #------color palette to use it later -------------------------------------

  pal1 <- c("#9986A5", "#79402E", "#CCBA72", "#FD6467")

#------------------------------ plot1-----------------------------------
p1=ggplot(conMat, aes(fill=base, y=freq, x=position)) +
  geom_bar(stat="identity",position = "fill") +
  scale_fill_manual(values = pal1) +
  ggtitle(paste0("Positional Nucleotide Frequencies of ",filesnames1," sample"))+
  scale_y_continuous("Frequency", labels = scales::percent)
#----------------------save the plot-----------------------------------
pathf <- getwd()
ggsave(filename = paste0(filesnames1,"_plot_PNF.png"),plot=p1,device = "png",
         path = paste0(dirname,"PNF_plots/"),
         width= 180,
         height = 98,2,
         units = "mm",dpi = "retina")
# plot1
p2=ggplot(conMat, aes(fill=base, y=freq, x=position)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = pal1) +
  ggtitle(paste0("Positional Nucleotide Abundance of absolute read count for ",filesnames1," sample")) +
  scale_y_continuous("read count")

#save the plot

ggsave(filename = paste0(filesnames1,"_plot_PNA.png"),plot=p2,device = "png",
         path = paste0(dirname,"PNF_plots/"),
         width= 180,
         height = 98,2,
         units = "mm",dpi = "retina")

  # create the percentage table
  conMat %>% 
    group_by(position) %>% 
    mutate(perc= paste0(round(freq/sum(freq)*100,2)," %")) %>% 
    select(-freq) %>% 
    spread(base,perc) %>% 
    write_delim(path = paste0(dirname,
          "PNF_plots/",filesnames1,"_PNF.txt"),delim = "\t")
}

#---------------testing if same freq, evaluation-------------------------------
Abase <- length(DT[str_detect(DT$Sequence,"^[A]"),ID:Reads]$ID)/length(DT$ID)
Tbase <- length(DT[str_detect(DT$Sequence,"^[T]"),ID:Reads]$ID)/length(DT$ID)
Cbase <- length(DT[str_detect(DT$Sequence,"^[C]"),ID:Reads]$ID)/length(DT$ID)
Gbase <- length(DT[str_detect(DT$Sequence,"^[G]"),ID:Reads]$ID)/length(DT$ID)

# using ggseqlogo
library(ggseqlogo)
data(ggseqlogo_sample)

seqs_dna
ggseqlogo(str_sub(DT$Sequence,1,15),method = "prob")
```


# multiplot function                                                        
```{r}
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

##------- Create Datasets-----------------------------------------------------
# 1st Dataset, Cell lines and testis sample, one Biol rep                   
```{r}
library(data.table);library(tidyverse);library(rafalib)
setwd("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/")

#  1. list files merged and outputs-------
list_smallRNA <- list.files(pattern = "_smallRNA", recursive = TRUE)
#list_output<- list.files(pattern = "_output", recursive = TRUE)
# find the file about the experiments Testis
list_smallRNA[str_detect(list_smallRNA,"Testis")] -> list_smallRNA_Testis
list_smallRNA_Testis[str_detect(list_smallRNA_Testis,"Non")]->list_smallRNA_Testis
#list_output[str_detect(list_output,"Testis")] -> list_output_Testis
#list_output_Testis[str_detect(list_output_Testis,"Non")]->list_output_Testis
# find the file about the experiments Cell lines
list_smallRNA[str_detect(list_smallRNA,"1_Cell")] -> list_smallRNA_Cell
#list_output[str_detect(list_output,"1_Cell")] -> list_output_Cell

# combine lists
listRNA <- c(list_smallRNA_Cell,list_smallRNA_Testis)
#listout <- c(list_output_Testis,list_output_Cell)

#load data to data.table
DTRNA <- rbindlist( sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")
#  2.0 make a tibble grouped by file names and smallRNA--------------------
#----------example
DTRNA %>% as_tibble %>% sample_n(25) %>% dput -> DT_sample
DT_sample %>% as_tibble %>% group_by(FileName) %>% 
  group_by(V6,add = TRUE) %>% mutate(V2 = V2 / n()) ->DT_sample
DT_sample %>% group_by(FileName) %>% 
  group_by(V4,add = TRUE) %>% mutate( counts = sum(V2)) ->DT_sample
DT_sample %>% group_by(FileName) %>% 
  group_by(V4,counts,add = TRUE) %>% 
  summarise() %>%
  arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT_sample
DT_sample
## 2.1 Using dplyr--------------------------------------------

DTRNA %>% 
  as_tibble %>% 
  dplyr::group_by(FileName,V6) %>% 
  dplyr::mutate(count = V2 / n()) %>% 
  dplyr::group_by(FileName,V4) %>% 
  dplyr::mutate( counts = sum(count))  %>% 
  dplyr::group_by(FileName,RNA_type,V4,counts) %>% 
  dplyr::summarise() %>%
  dplyr::arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT2
#------evaluate the resulted summaries
DT2 %>% 
 dplyr::group_by(FileName) %>% 
 dplyr::summarise(sums=sum(counts))

#--------change table from long to wide
DT2 %>% spread(FileName,RNA_type,counts, fill = 0) -> DT2
gsub("_smallRNA.txt","",basename(colnames(DT2))) ->colnames(DT2)
gsub("_Pool_Non_Treated","",colnames(DT2)) ->colnames(DT2)

# write for all
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt"
DT2 %>%
 write_delim(path,delim = "\t")



## 2.2 using data table------------------
DT <- DTRNA
DT <- DT %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

DT_last %>% group_by(FileName) %>% summarise(sums=sum(counts))

setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

gsub("_smallRNA.txt","",basename(colnames(DT_last))) ->colnames(DT_last)
gsub("_Pool_Non_Treated","",colnames(DT_last)) ->colnames(DT_last)

colnames(DT_last)
sum(DT_last[,Human_Testis])

path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt" # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  3.1.0 data pre-processing with edger----------------------------------------
library(edgeR);library(wesanderson)
# DT2 <- as_tibble(fread("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt" ,header = TRUE))
#DT2 <- as_tibble(DT_last)
DT2 <- DT2 %>%as_tibble %>%  column_to_rownames("smallRNA")

#mat <- as.matrix(DT2)
mat <- as.matrix(DT2 %>% select(CACO2:RKO,SW1417,SW403,Human_Testis))
group <- as.factor(colnames(mat))
batch <- as.factor(1:9)
status <- as.factor(c("MSS_low","MSS_high",
                      "MSI_no","MSS_low","MSS_low","MSI_no","MSS_high","MSS_low","Testis"))
samples <- data.frame(group,batch,status,row.names = colnames(mat))
x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))

## 3.1.1 filtering 1 cpm------------------------------------
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
table(rowSums(x$counts==0)==9)
keep.exprs <- rowSums(cpm>1)>=1
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
#x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
#lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

#cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_TMMwzpiRNA

#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_testis_normalized_TMM_logcpm_noSK_counts.txt",delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
#  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "all_mat_Cell_testis_noSK_norm_TMMwzp_logcpm_counts.txt",delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
#  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "all_mat_Cell_testis_noSK_norm_TMMwzp_cpm_counts.txt",delim = "\t")
  
  
## 3.1.2 density plot-------------------------  
pal1 <- c("#9986A5", "#79402E", "#CCBA72",
          "#FD6467","#F3DF6C","#CEAB07","#D5D5D3","#24281A","#FF0000")
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=2, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")
lcpm <- cpm(x1, log=TRUE, prior.count=2)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=2, 
     main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

## 3.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)
hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 3.1.4 boxplots---------------------------------------------------
par(mfrow=c(1,3))
par(mar=c(8,3,2,3))
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col=pal1, main="")
title(main="A. Unnormalized data",ylab="Log-cpm")


#-----------------TMM
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMM, las=2, col=pal1, main="")
title(main="B. TMM Normalized data",ylab="Log-cpm")


#-----------------TMMwzp
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMMwzp, las=2, col=pal1, main="")
title(main="C. TMMwzp Normalized data",ylab="Log-cpm")

## 3.1.5 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(x1_TMMwzp$counts))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")
legend("topleft",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )


hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

## 3.1.6 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(1,2))

legend("topleft",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS.DGEList(x1_TMMwzp,labels = x1$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(5,6))

legend("topleft",c("4","5","6"),
       fill = as.numeric(x1_TMMwzp$targets$status),cex=1.5,inset = c(.01,.09))

## 3.1.7 mat filter for piRNAs-------------------------------------------
mat_piR <- as.matrix(x1_TMM$counts %>% 
  as_tibble(rownames=NA) %>% 
  rownames_to_column(var="smallRNA") %>%   
  filter(str_detect(smallRNA,"piRNA_m"))%>%  
  column_to_rownames("smallRNA"))
 x1_TMMwzpiRNA<- x1_TMMwzp[rownames(mat_piR),,keep.lib.sizes=TRUE]
 lcpm_TMMwzpiRNA <- lcpm_TMMwzp[rownames(mat_piR),]
## 3.1.8 piRNAs Hierarchical, clustering---------------------------------------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMMwzpiRNA))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")



hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with only piRNAs")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")
legend("topright",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 3.1.9 piRNAs MDS plot mat1------------------------------------------------
mat_piR


par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(1,2))

legend("topleft",legend=levels(x1_TMMwzpiRNA$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzpiRNA$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(5,6))
## 3.1.10 filter for pirnas only in testis or in cell lines----------
mat_piR <- as.matrix(x1_TMM$counts %>% 
  as_tibble(rownames=NA) %>% 
  rownames_to_column(var="smallRNA") %>%   
  filter(str_detect(smallRNA,"piRNA_m"))%>%  
  column_to_rownames("smallRNA"))
# common testis cell lines 1cpm
cpm_TMMwzpiRNA <- cpm_TMMwzp[rownames(mat_piR),]
com_piR <- rowSums(cpm_TMMwzpiRNA>1)==9
common_piRNAs <- cpm_TMMwzpiRNA[com_piR,]

common_piRNAs %>% 
  as_tibble(rownames="smallRNA") %>% 
  write_delim(path = "common_piRNA_Cell_Testis_1cpm_norm.txt",delim = "\t")

# only in testis lines 1cpm

cpm_TMMwzpiRNA %>% 
  as_tibble(rownames="smallRNA") %>% 
  filter(Human_Testis>1,CACO2==0,COLO205==0,HCT116==0,HT115==0,HT29==0,
         RKO==0,SW1417==0,SW403==0) %>% 
  write_delim(path = "only_piRNA_Testis_1cpm_norm.txt",delim = "\t")

# common in cell lines 1cpm
cpm_TMMwzpiRNA %>% 
  as_tibble(rownames="smallRNA") %>% 
  filter(Human_Testis==0,CACO2>1,COLO205>1,HCT116>1,HT115>1,HT29>1,
         RKO>1,SW1417>1,SW403>1) %>% 
   write_delim(path = "only_piRNA_Cell_lines_1cpm_norm.txt",delim = "\t")
  
  

#  3.2.0 mat2 with only COLO cell lines---------------------------
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/")
mat2 <- as.matrix(DT2 %>% select(CACO2:RKO,SW1417,SW403))

group <- as.factor(colnames(mat2))
batch <- as.factor(1:8)
status <- as.factor(c("MSS_low","MSS_high",
                      "MSI_no","MSS_low","MSS_low","MSI_no","MSS_high","MSS_low"))
samples <- data.frame(group,batch,status,row.names = colnames(mat2))
x <- DGEList(counts = mat2,samples = samples,lib.size = colSums(mat2),norm.factors = rep(1,ncol(mat2)))

## 3.2.1 filtering 1 cpm------------------
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
table(rowSums(x$counts==0)==8)
keep.exprs <- rowSums(cpm>1)>=1
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)


#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_noSK_TMM_logcpm_piRNA.txt",delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_noSK_TMMwzp_logcpm_piRNA.txt",delim = "\t")
#cpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_noSK_TMM_cpm_piRNA.txt",delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_noSK_TMMwzp_cpm_piRNA.txt",delim = "\t")  
  

## 3.2.2 density plot-------------------------  
pal1 <- c("#9986A5", "#79402E", "#CCBA72",
          "#FD6467","#F3DF6C","#CEAB07","#D5D5D3","#24281A","#FF0000","#00A08A")
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

#par(srt=0)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")
lcpm <- cpm(x1, log=TRUE, prior.count=2)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=1, 
     main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
text(srt=180)
legend("topright", samplenames, text.col=pal1, bty="n", adj=1, x.intersp = 0, xjust = 0,yjust = 0)

## 3.2.3 histogram------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)
hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 3.2.4 boxplots--------
par(mfrow=c(1,3))
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col=pal1, main="")
title(main="A. Unnormalized data",ylab="Log-cpm")

#TMM
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMM, las=2, col=pal1, main="")
title(main="B. TMM Normalized data",ylab="Log-cpm")

#TMMwzp
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMMwzp, las=2, col=pal1, main="")
title(main="C. TMMwzp Normalized data",ylab="Log-cpm")

## 3.2.5 Hierarchical, clustering, allRNA-------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMMwzp$counts))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")
#legend("topright",legend=levels(x1_TMMwzp$samples$status),
#       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
 #      bty="n",
 #      cex = 1.5, inset = c(.01,.09) )

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.01) )

## 3.2.6 MDS plot, allRNA-------
par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(1,2))

legend("topleft",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS.DGEList(x1_TMMwzp,labels = x1$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(5,6))

## 3.2.7 mat filter for piRNAs-------------------------------------------
mat_piR <- as.matrix(x1_TMM$counts %>% 
  as_tibble(rownames="smallRNA") %>% 
  filter(str_detect(smallRNA,"piRNA_m"))%>%  
  column_to_rownames("smallRNA"))

x1_TMMwzpiRNA <- x1_TMMwzp[rownames(mat_piR),,keep.lib.sizes=TRUE]
lcpm_TMMwzpiRNA <- lcpm_TMMwzp[rownames(mat_piR),]

## 3.2.8 piRNAs Hierarchical, clustering---------------------------------------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMMwzpiRNA))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,
          xlab = "euclidean-ward.D2",main="",hang = 0.03)

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with only piRNAs",hang = 0.03 )

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,
          xlab = "euclidean-average",main="",hang = 0.02)

legend("topright",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="o",
       cex = 1.5, inset = c(.01,.01) )
## 3.2.9 piRNAs MDS plot mat1------------------------------------------------
mat_piR

par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(1,2))

legend("topright",legend=levels(x1_TMMwzpiRNA$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzpiRNA$samples$status))),
       bty="o",
       cex = 1.5, inset = c(.01,.01) )

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(5,6))

## 3.2.10 DE analysis------
design <- model.matrix(~0+status)
colnames(design) <- levels(status)
rownames(design) <- rownames(samples)

### estimate dispersion
plotMD(x1_TMMwzp,column = 1)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp$samples$group <- x1_TMMwzp$samples$status
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp1,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)

### DE MSS_h_MSI
MSS_h_MSI <- makeContrasts(MSS_high-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_h_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
 # write_delim(path = "MSS_h_v_MSI_top100.txt",delim = "\t")

### DE MSS_H_v_L
MSS_H_v_L <- makeContrasts(MSS_high-MSS_low,levels = design)
res <- glmQLFTest(fit,contrast = MSS_H_v_L)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA")  
 # write_delim(path = "MSS_H_v_MSS_L_top100.txt",delim = "\t")
  
### DE MSS_L_v_MSI
MSS_L_v_MSI <- makeContrasts(MSS_low-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_L_v_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") #%>% 
 write_delim(path = "MSS_L_v_MSI_top100.txt",delim = "\t")  
  
 
#  3.3.0 limma quality weights--------
 
DT<- read_delim("~/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt",delim = "\t")

mat<-DT %>% 
  select(smallRNA:RKO,SW1417,SW403) %>% 
  column_to_rownames("smallRNA")
group <- as.factor(c("MSS_low","MSS_high",
                      "MSI_no","MSS_low","MSS_low","MSI_no","MSS_high","MSS_low"))
batch <- as.factor(1:8)
samples <- data.frame(group,batch,row.names = colnames(mat))

design <- model.matrix(~0+group)
colnames(design) <- levels(group)
rownames(design) <- rownames(samples)

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))

cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
table(rowSums(x$counts==0)==8)
keep.exprs <- rowSums(cpm>1)>=1
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
#x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
#lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
#lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
voom_TMMwzp <- voomWithQualityWeights(x1_TMMwzp,design = design,plot = TRUE)
vfit <- lmFit(voom_TMMwzp)
cm <- makeContrasts(
  MSSH_V_MSI = MSS_high-MSI_no,
  MSSH_V_MSSL = MSS_high-MSS_low,
  MSSL_V_MSI = MSS_low-MSI_no,
  levels = design
)
vfit <- contrasts.fit(vfit,cm)
vfit <- eBayes(vfit,robust = TRUE,trend = TRUE)

topMSSH_V_MSI <- topTable(vfit,coef = "MSSH_V_MSI",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p")
head(topMSSH_V_MSI)

topMSSH_V_MSSL <- topTable(vfit,coef = "MSSH_V_MSSL",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p")
head(topMSSH_V_MSSL)

## 3.3.1
```

# 3rd Dataset, IPP_PIWIL                                                    
```{3rd Dataset:IPP_PIWIL}
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/")
#  1.1   list files merged and outputs of PIWIL                             ----
library(data.table);library(tidyverse);library(rafalib)
setwd("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/")

path_IPP <- "R_virtual_shared_folder/Documents/new_analysis_workflow/1_5_files_withDBs/3_IPP_PIWIL1_all_DBs/"
list_smallRNA <- list.files(path = path_IPP, pattern = "_smallRNA", recursive = TRUE, full.names = TRUE)
list_smallRNA <- list.files(pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments IPP_PIWIL1
listRNA <- list_smallRNA[str_detect(list_smallRNA,"IPP_PIWIL1")]  

#load data to data.table
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

## 1.1.a Using dplyr                                                        ----
DTRNA %>% filter(RNA_type,"piR")

DTRNA <- DTRNA %>% 
  as_tibble() %>%  
  unite("smallRNA",RNA_type,V4,sep = "_")

piRNAT_reads %>% 
  distinct(V6, .keep_all = TRUE) %>% 
  group_by(V1) %>% 
  mutate( width = str_length(V6)) %>% 
  arrange(desc(width)) %>% ungroup() %>% 
  distinct(piRNA, .keep_all = TRUE)
#clean piRNA entries
#1
piR_DTRNA <- DTRNA %>% filter(str_detect(smallRNA,"piR"))
piR_DTRNA %>% filter(!duplicated(.$V6))  %>% group_by(smallRNA) %>% summarise(counts = sum(V2))  %>% arrange(desc(counts))
#2
piR_DTRNA_dist <- piR_DTRNA %>% 
  distinct(V6, .keep_all = TRUE) %>%
  group_by(smallRNA)
#3
piR_DTRNA_length <- piR_DTRNA_dist %>%
  mutate( width = str_length(V6)) %>% 
  arrange(desc(width)) %>% ungroup()
#4
piR_DTRNA_filtered <- piR_DTRNA_length %>% 
  distinct(smallRNA, .keep_all = TRUE) 

#piR_DTRNA_filteredgroup <- piR_DTRNA_length %>% group_by(FileName) %>% 
  distinct(smallRNA, .keep_all = TRUE)
#5
DTpir <- piR_DTRNA %>% 
  filter(!smallRNA %in% piR_DTRNA_filtered$smallRNA)
#####in sanity check
DTcleanpir <- piR_DTRNA %>% 
  filter(smallRNA %in% piR_DTRNA_filtered$smallRNA)
filter(piR_DTRNA,str_detect(FileName,"COLO205_IPP_C_Ab_1_small"))$smallRNA %>% base::duplicated() %>% table
filter(DTcleanpir,str_detect(FileName,"COLO205_IPP_C_Ab_1_small"))$smallRNA %>% base::duplicated() %>% table

#final data
DTRNA_cor <- DTRNA %>% 
  filter(!smallRNA %in% DTpir$smallRNA)
DTRNA_clear <- DTRNA %>% 
  filter(!smallRNA %in% DTcleanpir$smallRNA)

DTRNA %>% 
  as_tibble %>% 
  dplyr::group_by(FileName,V6) %>% 
  dplyr::mutate(count = V2 / n()) %>% 
  dplyr::group_by(FileName,V4) %>% 
  dplyr::mutate( counts = sum(count))  %>% 
  dplyr::group_by(FileName,RNA_type,V4,counts) %>% 
  dplyr::summarise() %>%
  dplyr::arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT2
#------evaluate the resulted summaries
DT2 %>% 
 dplyr::group_by(FileName) %>% 
 dplyr::summarise(sums=sum(counts))

#--------change table from long to wide
DT2 %>% spread(FileName,RNA_type,counts, fill = 0) -> DT2
gsub("_smallRNA.txt","",basename(colnames(DT2))) ->colnames(DT2)
gsub("_Pool_Non_Treated","",colnames(DT2)) ->colnames(DT2)

# write for all
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt"
DT2 %>%
 write_delim(path,delim = "\t")

## 1.1.b using data table                                                   ----
DT <- DTRNA
DT <- DT %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

DT_last %>% group_by(FileName) %>% summarise(sums=sum(counts))

setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

gsub("_smallRNA.txt","",basename(colnames(DT_last))) ->colnames(DT_last)
gsub("_Pool_Non_Treated","",colnames(DT_last)) ->colnames(DT_last)

colnames(DT_last)
sum(DT_last[,COLO205_IPP_C_Ab_1])

path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/1_IPPIWIL_raw_counts.txt" # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)
## 2.1.0 Data pre-processing with edger, packaging                          ----
library(edgeR);library(wesanderson)
#DT2 <- as_tibble(fread("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/1_IPPIWIL_raw_counts.txt" ,header = TRUE))
#DT2 <- as_tibble(DT_last)
DT2 <- DT2 %>%as_tibble %>%  column_to_rownames("smallRNA")

mat <- as.matrix(DT2)

group <- as.factor(str_replace(colnames(mat),"_[:digit:]$","")  %>%
                     str_replace("COLO205_","") %>% 
                     str_replace("Control_",""))
batch <- as.factor(rep(c(1:3),4))

samples <- data.frame(group,batch,row.names = colnames(mat))
x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)


## 2.1.1 filtering                                                          ----
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)

log2(5/L)

summary(lcpm)

table(rowSums(x$counts==0)==12)

keep.exprs <- filterByExpr.DGEList(x,group = group)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

## 2.1.2 density plot                                                       ----  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,3))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="B. Filtered data, 0.4 cpm", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


lcpm.cutoff <- log2(5/M + 2/L)

lcpm <- cpm(x11, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="B. Filtered data, 0.2 cpm", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")



## 2.1.3 histogram                                                          --------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)
AveLogCpm_Filtered_Data2 <-aveLogCPM(x11)

par(mfrow=c(1,3))

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)
hist(AveLogCpm_Filtered_Data2)


## 2.1.4 normalization                                                      ----
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

write_rds(x1_voomTMM,"x1_voomTMM.rds")
x1_voomTMM <- read_rds("~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/x1_voomTMM.rds")

## 2.1.5 boxplots                                                           ----
par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(1,3))
#voomQ data
boxplot(x1_voomQ$E, las=2, col= wes_palettes$Royal1[1:3], main="",,names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")


## save the matrices
#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "IPP_PIWIL_TMM_logcpm_piRNAcounts.txt",delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "IPP_PIWIL_TMMwzp_logcpm_piRNAcounts.txt",delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "IPP_PIWIL_TMMwzp_cpm_piRNAcounts.txt",delim = "\t")
#x1_voomQ$E %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "IPP_PIWIL_voomQ_QualW_lcpm_piRNAcounts.txt",delim = "\t")
  

## 2.1.6 Hierarchical, clustering, allRNA                                   ----
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMM))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, TMM")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 2.1.7 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,3))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomQ$E,labels = group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(1,2))

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomQ$E,labels = x1$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(3,4),main=" MDS plot of IPP PIWIL1 with voom Quantile normalized counts considering batch")

plotMDS(x1_voomQ$E,labels = x1_TMMwzp$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(5,6))




## 3.1   DE analysis------

### estimate dispersion
par(mfrow=c(1,1))
plotMD(x1_TMMwzp,column = 2)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)
#make contrasts
con 
### DE MSS_h_MSI
MSS_h_MSI <- makeContrasts(MSS_high-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_h_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
 # write_delim(path = "MSS_h_v_MSI_top100.txt",delim = "\t")

### DE MSS_H_v_L
MSS_H_v_L <- makeContrasts(MSS_high-MSS_low,levels = design)
res <- glmQLFTest(fit,contrast = MSS_H_v_L)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA")  
 # write_delim(path = "MSS_H_v_MSS_L_top100.txt",delim = "\t")
  
### DE MSS_L_v_MSI
MSS_L_v_MSI <- makeContrasts(MSS_low-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_L_v_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") #%>% 
 write_delim(path = "MSS_L_v_MSI_top100.txt",delim = "\t")  
  
 
## 3.3.0 limma quality weights--------

vfit <- lmFit(x1_voomTMM,design=design)
con_mat <- makeContrasts(
  IPP_C_vs_noAb = IPP_C_Ab-IPP_noAb,
  IPP_N_vs_noAb = IPP_N_Ab-IPP_noAb,
  IPP_INPUT_vs_noAb = IPP_INPUT-IPP_noAb,
  IPP_C_vs_INPUT = IPP_C_Ab-IPP_INPUT,
  IPP_N_vs_INPUT = IPP_N_Ab-IPP_INPUT,
  levels = design
)
vfit <- contrasts.fit(vfit,con_mat)
vfit <- eBayes(vfit,robust = TRUE,trend = TRUE)

write_rds(vfit,"vfit_x1_voomTMM.rds")
vfit <- read_rds("vfit_x1_voomTMM.rds")

#IPP_C_vs_noAb
topIPP_C_vs_noAb <- topTable(vfit,coef = "IPP_C_vs_noAb",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "smallRNA")

#IPP_N_vs_noAb
topIPP_N_vs_noAb <- topTable(vfit,coef = "IPP_N_vs_noAb",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "smallRNA")

#IPP_INPUT_vs_noAb
topIPP_INPUT_vs_noAb <- topTable(vfit,coef = "IPP_INPUT_vs_noAb",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "smallRNA") %>% 
  select(smallRNA,
         Input_vs_noAb_logFC = "logFC",
         AveExpr ,
         Input_vs_noAb_P.Value = "P.Value",
         Input_vs_noAb_adj.P.Val = "adj.P.Val")

#IPP_C_vs_INPUT
topIPP_C_vs_INPUT <- topTable(vfit,coef = "IPP_C_vs_INPUT",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "smallRNA") %>% 
  select(smallRNA,
         ab1_vs_noAb_logFC = "logFC",
         AveExpr ,
         ab1_vs_noAb_P.Value = "P.Value",
         ab1_vs_noAb_adj.P.Val = "adj.P.Val")

#IPP_N_vs_INPUT
topIPP_N_vs_INPUT <- topTable(vfit,coef = "IPP_N_vs_INPUT",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "smallRNA") %>% 
  select(smallRNA,
         ab2_vs_noAb_logFC = "logFC",
         AveExpr ,
         ab2_vs_noAb_P.Value = "P.Value",
         ab2_vs_noAb_adj.P.Val = "adj.P.Val")
# make table for PAPER Journal PUB
topIPP_INPUT_vs_noAb %>% 
  inner_join(topIPP_C_vs_INPUT,by = c("smallRNA","AveExpr")) %>% 
  inner_join(topIPP_N_vs_INPUT,by = c("smallRNA","AveExpr")) %>%  select(smallRNA,AveExpr,everything()) %>% 
  separate(smallRNA,c("DB","smallRNA"), sep = "_match_") %>% 
  arrange(DB) %>% 
  write_tsv("Table_IPP_journal.txt")


topIPP_C_vs_INPUT %>% 
  as_tibble(rownames = "smallRNA") %>% 
  filter(str_detect(smallRNA,"piRNA_m"),logFC>1,adj.P.Val<0.05)

l <- list(topIPP_C_vs_noAb,topIPP_N_vs_noAb,
          topIPP_INPUT_vs_noAb,topIPP_C_vs_INPUT,
            topIPP_N_vs_INPUT) 
fname <- c("topIPP_C_vs_noAb","topIPP_N_vs_noAb",
           "topIPP_INPUT_vs_noAb","topIPP_C_vs_INPUT",
            "topIPP_N_vs_INPUT")
names(l) <- fname
lapply(fname,
       function(name)write_delim(l[[name]],
                                 path=paste0(name,"_ENRICH.txt") ,delim = "\t"))

LFCs <- rbindlist(l,use.names= TRUE,idcol="comparison") %>% as_tibble()

LFCs %>% as_tibble() %>% 
  filter(adj.P.Val<0.01) %>%
  select(comparison,smallRNA,logFC) %>% 
  spread(comparison,logFC,drop = TRUE )-> LFCstable

write_delim(LFCstable,path="LFCs_tables_FDR001_comparisons.txt",delim = "\t")
# LFCstable <- read_delim("LFCs_tables_FDR001_comparisons.txt",delim = "\t")

# again as before less columns
l <- list(topIPP_INPUT_vs_noAb,topIPP_C_vs_INPUT,
            topIPP_N_vs_INPUT) 
names(l) <- c("topIPP_INPUT_vs_noAb","topIPP_C_vs_INPUT",
            "topIPP_N_vs_INPUT")

LFCs <- rbindlist(l,use.names= TRUE,idcol="comparison") 

LFCs %>% as_tibble() %>% 
  filter(adj.P.Val<0.01) %>%
  select(comparison,smallRNA,logFC) %>% 
  spread(comparison,logFC,drop = TRUE )-> LFCstable

LFCstable %>% 
  filter(topIPP_C_vs_INPUT > 2, topIPP_N_vs_INPUT > 2) 

exclude <- LFCstable %>% 
  filter(topIPP_INPUT_vs_noAb < (-2)) 

(RNAs <- LFCstable  %>% 
  filter(topIPP_C_vs_INPUT>2,topIPP_N_vs_INPUT>2) %>% 
  filter(!smallRNA %in% exclude$smallRNA))
smallRNAs585 <- as_tibble(fread("~/Downloads/smallRNAs585.txt"))

write_delim(RNAs,path="LFCs_C_N_2_input_2_results.txt",delim = "\t")

LFCstable %>% 
  filter(topIPP_C_vs_INPUT > 2, topIPP_N_vs_INPUT > 2,
         topIPP_INPUT_vs_noAb < (-2)) %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

## 3.3.1 visualization of results-------

LFCs %>% as_tibble() %>% 
    select(comparison,smallRNA,logFC) %>% 
    spread(comparison,logFC,drop = TRUE )-> LFCstable

Expr <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA") 

my_data <- inner_join(Expr,LFCstable,by="smallRNA")

#LFCs %>% as_tibble() %>% 
  filter(adj.P.Val < 0.01) %>%
  select(comparison,smallRNA,logFC) %>% 
  spread(comparison,logFC,drop = TRUE ) -> LFCstable

filN <- LFCstable %>% 
  filter(topIPP_N_vs_INPUT > 2,!smallRNA  %in% exclude$smallRNA ) %>% 
  write_delim("N_enriched_RNA.txt",delim = "\t") 

filC <- LFCstable %>% 
  filter(topIPP_C_vs_INPUT > 2,!smallRNA  %in% exclude$smallRNA ) %>% 
  write_delim("C_enriched_RNA.txt",delim = "\t") 

my_dataN <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(filtered=(smallRNA %in% filN$smallRNA),
                   Input_meanExpr=mean(c(COLO205_IPP_INPUT_1,
                                       COLO205_IPP_INPUT_2,
                             COLO205_IPP_INPUT_3)))

my_dataC <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(filtered=(smallRNA %in% filC$smallRNA),
                   Input_meanExpr=mean(c(COLO205_IPP_INPUT_1,
                                       COLO205_IPP_INPUT_2,
                             COLO205_IPP_INPUT_3)))

my_dataN %>% 
  select(smallRNA,topIPP_N_vs_INPUT,Input_meanExpr,filtered) %>% 
  filter(topIPP_N_vs_INPUT > 0) %>% 
ggplot()+
  geom_point(mapping=aes(x=Input_meanExpr,
                         y=topIPP_N_vs_INPUT,colour= filtered))+
labs(title = str_glue("{nrow(filN)} Enriched smallRNAs, FDR < 0.01"),
     x="RNA expression level (log2, limma-voom transformed expression)",
     y="N ab log2 enrichment factor",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=c("not-Enriched","Enriched"),values = c("black","red"))
  
my_dataC %>% 
  filter(topIPP_C_vs_INPUT > 0) %>% 
ggplot()+
  geom_point(mapping=aes(x=Input_meanExpr,
                         y=topIPP_C_vs_INPUT,colour= filtered))+
labs(title = str_glue("{nrow(filC)} Enriched smallRNAs, FDR<0.01"),
     x="RNA expression level (log2, limma-voom transformed expression)",
     y="C ab log2 enrichment factor",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=c("not-Enriched","Enriched"),values = c("black","blue"))


fil <- LFCstable %>% 
  filter(topIPP_N_vs_INPUT>2 ,topIPP_C_vs_INPUT>2 , !smallRNA %in% exclude$smallRNA)

my_dataCN <- my_data %>% 
  filter(topIPP_N_vs_INPUT>2|topIPP_C_vs_INPUT>2) %>% 
  group_by(smallRNA) %>% 
  mutate(Both=(smallRNA %in% fil$smallRNA),
         C=(smallRNA %in% filC$smallRNA),
         N=(smallRNA %in% filN$smallRNA),
         ench_ratio= log2( ((round(2^topIPP_C_vs_INPUT,3)) / (round(2^topIPP_N_vs_INPUT,3)) )+0.001)) %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,ench_ratio,Both,C,N)

sum(is.na(my_dataCN$ench_ratio))

#between two abs
my_dataCN <- my_data %>% 
  filter(topIPP_N_vs_INPUT>2|topIPP_C_vs_INPUT>2) %>% 
  group_by(smallRNA) %>% 
  mutate(ab= ifelse(smallRNA %in% fil$smallRNA,"Both",
                       ifelse(smallRNA %in% filC$smallRNA,"C",
                               ifelse(smallRNA %in% filN$smallRNA,"N","FDR > 0.01"))),
         Both=(smallRNA %in% fil$smallRNA),
         C=(smallRNA %in% filC$smallRNA),
         N=(smallRNA %in% filN$smallRNA),ench_ratio= log2( ((round(2^topIPP_C_vs_INPUT,3)) / (round(2^topIPP_N_vs_INPUT,3)) )+0.001)) %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,ench_ratio,ab,Both,C,N)

my_dataCN %>% 
  filter(!ab=="FDR > 0.01") %>% 
ggplot()+
  geom_point(mapping=aes(x=topIPP_C_vs_INPUT,
                         y=topIPP_N_vs_INPUT,
                         colour= ab))+
labs(title = str_glue("{nrow(fil)} Enriched common smallRNAs, FDR < 0.01"),
     x="C ab log2 enrichment factor",
     y="N ab log2 enrichment factor",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(values = c("purple","blue","red","black","red"))

#ench ratio
my_dataCN %>% 
#  replace_na(list(ench_ratio=0,topIPP_C_vs_INPUT=0)) %>% 
ggplot()+
  geom_point(mapping=aes(x=ench_ratio,
                         y=topIPP_N_vs_INPUT,colour= Both))+
labs(title = "777 Enriched common smallRNAs, FDR<0.01",
     x="log2 enrichment ratio (C/N ab)",
     y="C ab log2 enrichment factor",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=c("Enriched","Enriched in both"),values = c("black","purple"))



#  3.4   5,3UTR and CDS-------------------
library(biomaRt)
mart <- useDataset("hsapiens_gene_ensembl",useMart("ensembl"))

setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/")

#UTR3
UTR3 <- fread("3UTR.txt",select = c(1,3)) %>% 
  as_tibble() %>% rename("UTR3"=V3) %>% 
  separate(UTR3,c("UTR3","mRNA")) %>% 
  distinct(V1,UTR3)

G_listUTR3 <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = UTR3$UTR3,mart = mart)

UTR3gen <- inner_join(UTR3,G_listUTR3,by= c("UTR3"="ensembl_gene_id"))  %>% distinct()
UTR3gen %>% write_delim("UTR3genes.txt",delim = "\t")
#UTR5
UTR5 <- fread("5UTR.txt",select = c(1,3)) %>% 
  as_tibble() %>% rename("UTR5"=V3) %>% 
  separate(UTR5,c("UTR5","mRNA")) %>% 
  distinct(V1,UTR5)

G_listUTR5 <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = UTR5$UTR5,mart = mart)

UTR5gen <- inner_join(UTR5,G_listUTR5,by= c("UTR5"="ensembl_gene_id"))  %>% distinct()
UTR5gen %>% write_delim("UTR5genes.txt",delim = "\t")


#CDS
CDS  <- fread("CDS.txt" ,select = c(1,3)) %>% 
  as_tibble() %>% rename("CDS"=V3) %>% 
  separate(CDS,c("CDS","mRNA")) %>% 
  distinct(V1,CDS)

G_listCDS <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = CDS$CDS,mart = mart)

CDSgen <- inner_join(CDS,G_listCDS,by= c("CDS"="ensembl_gene_id")) %>% distinct()
CDSgen %>% write_delim("CDSgenes.txt",delim = "\t")



genelist <- list(UTR3,UTR5,CDS)
names(genelist) <- c("UTR3","UTR5","CDS")
genReg <- rbindlist(genelist,use.names = TRUE,idcol = "region",fill = TRUE)
genReg %>% write_delim("genesIPP",delim = "\t")



utrs <- UTR3gen %>% filter(UTR3 %in% UTR5gen$UTR5) %>% 
  filter(UTR3 %in% CDS$CDS)

utr <- inner_join(UTR3,UTR5,by= c("V1","UTR3"="UTR5"))
cdsutr <- inner_join(utr,CDS,by= c("V1","UTR3"="CDS"))
#  3.5    -----------


# DESeq2-------------------------------------------------------------------

library(DESeq2);library(tidyverse);library(data.table)
setwd("/media/sf_R_virtual_shared_folder/3_IPP_PIWIL1/")
temp <- list.files(pattern = ".fastq", recursive = TRUE)
gsub(".fastq","",basename(temp))->temp
gsub(".fastq","_processed",temp)->temp1
setwd("/home/rstudio/Documents/SPORTS_results/Result/")
output <- list.files(pattern = "output_(ensembl|miRNA|piRNA|rfam|rRNA|tRNA|tRNA_mature|noAnnot)_match_genome$", recursive = TRUE)

```

# ensembl transcripts                                                       
```{r}
setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/")
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(data.table);library(tidyverse)

# load the tables
RawCounts  <- fread("1_IPPIWIL_raw_counts.txt") %>% as_tibble()
x1_voomTMM <- read_rds("x1_voomTMM.rds")
normCounts <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA")

# subset for ensembl transcript
RawTran <- RawCounts %>% 
  filter(str_detect(smallRNA,"processed_transcript"))
normTran <- normCounts %>% 
  filter(str_detect(smallRNA,"processed_transcript"))

# find medians
normTran <- normTran %>% 
  group_by(smallRNA) %>% 
  mutate(medC = median(c(COLO205_IPP_C_Ab_1,COLO205_IPP_C_Ab_2,COLO205_IPP_C_Ab_3)),
        medNoAb = median(c(COLO205_IPP_Control_noAb_1,COLO205_IPP_Control_noAb_2,COLO205_IPP_Control_noAb_3)),
        medInput = median(c(COLO205_IPP_INPUT_1,COLO205_IPP_INPUT_2,COLO205_IPP_INPUT_3)),
        medN = median(c(COLO205_IPP_N_Ab_1,COLO205_IPP_N_Ab_2,COLO205_IPP_N_Ab_3))) %>% 
  select(smallRNA,medC,medN,medInput,medNoAb)

# extract genes names
genNames <- str_extract(normTran$smallRNA,"gene_symbol:.+") %>% 
  str_replace("gene_symbol:", "") %>% 
   str_replace(" description.+", "") 

# load the targets
path1 <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/targets/all_IPP_genes_3C5__Mar_05_2019.txt"
path2 <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/targets/NUC_IPP_genes_3C5__Mar_05_2019.txt"
targetsAll <- read_delim(path1,delim = "\t")
targetsNUC <- read_delim(path2,delim = "\t")

targetsAll %>% filter(UTR5 %in% genNames)
targetsNUC %>% filter(CDS %in% genNames)
```


# 2nd Dataset, TOTAL Periodate Treatment                                    
```{2nd Dataset:Periodate Treatment TOTAL}

library(data.table);library(tidyverse);library(rafalib)
name_file <- "TOTAL"
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
#  1.1   List files merged and outputs of Periodate TOTAL ----------
# working dir
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/")

# first select the correct tables
path <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/2_Periodate_treat_all_DBs/"
list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments TOTAL
listRNA <- list_smallRNA[str_detect(list_smallRNA,name_file)] 
listRNA <- sapply(listRNA, function(x)str_glue({path},{x}),simplify=FALSE)

#load data to data.table
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

## 1.1.a Using dplyr--------------------------------------------
DTRNA <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DTRNA %>% 
  as_tibble %>% 
  dplyr::group_by(FileName,V6) %>% 
  dplyr::mutate(count = V2 / n()) %>% 
  dplyr::group_by(FileName,V4) %>% 
  dplyr::mutate( counts = sum(count))  %>% 
  dplyr::group_by(FileName,RNA_type,V4,counts) %>% 
  dplyr::summarise() %>%
  dplyr::arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT2
#------evaluate the resulted summaries
DT2 %>% 
 dplyr::group_by(FileName) %>% 
 dplyr::summarise(sums=sum(counts))

#--------change table from long to wide
DT2 %>% spread(FileName,RNA_type,counts, fill = 0) -> DT2
gsub("_smallRNA.txt","",basename(colnames(DT2))) ->colnames(DT2)
gsub("_Pool_Non_Treated","",colnames(DT2)) ->colnames(DT2)

# write for all
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt"
DT2 %>%
 write_delim(path,delim = "\t")



## 1.1.b using data table------------------
DT <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,name_file)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as_vector(sumsDT$Reads),as_vector(round(newSUms$sums))))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,TOTAL_1_NaIO4])

# write the table
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/TOTAL_raw_counts.txt" # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  2.1   Data pre-processing with edger----------------------------------------
## 2.1.0 Data packaging----------
library(edgeR);library(wesanderson)

#load the file
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/TOTAL_raw_counts.txt"
#DT2 <- as_tibble(DT_last) %>%  column_to_rownames("smallRNA")
DT2 <- as_tibble(fread(path ,header = TRUE)) %>%  
  column_to_rownames("smallRNA")

#prepare the matrix, targets df and design
mat <- as.matrix(DT2)
group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)

#make the DGElist object
x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))


## 2.1.1 filtering ------------------------------------
# compute tables for dens plot 
cpm(10,mean(x$samples$lib.size))
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
summary(lcpm)
table(rowSums(x$counts==0)==6)

# find which are the filtered data
keep.exprs <- filterByExpr.DGEList(x,design = group)
# filter the data
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

## 2.1.2 density plot-------------------------  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data of {name_file}, ~ {round(lcpm.cutoff,3)} log-cpm"), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


## 2.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 2.1.4 normalization---------------------------------------------------------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voom(x1_TMM,design,plot = TRUE)
x1_voomTMM_QW <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voomWithQualityWeights(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{name_file}_filt2_{date}.txt"),sep = "\t",col.names = TRUE)
lcpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("lcpm_TMM_{name_file}_filt2_{date}.txt"),sep = "\t",col.names = TRUE)
cpm_TMMwzp %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMMwzp_{name_file}.txt"),sep = "\t",col.names = TRUE)


write_rds(x1_voomQ,str_glue("x1_voomQ_filt2_{name_file}_{date}.rds"))
write_rds(x1_voomTMM,str_glue("x1_voomTMM_filt2_{name_file}_{date}.rds"))
x1_voomQ <- read_rds(str_glue("x1_voomQ_filt2_{name_file}.rds"))
x1_voomTMM <- read_rds(str_glue("x1_voomTMM_filt2_{name_file}.rds"))
## 2.1.5 boxplots---------------------------------------------------
par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(1,4))
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#voomQ data
boxplot(x1_voomQ$E, las=2, col= wes_palettes$Royal1[1:3], main="",,names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")
#voomTMMwzp data
boxplot(x1_voomTMMwzp$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMMwzp Normalized filtered data",ylab="Log-cpm")

## 2.1.6 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(x1_voomTMM$E))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, voomTMM")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 2.1.7 MDS plot, all RNA----------------------------------------------
par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomTMM$E,labels = group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],2),dim.plot = c(1,2))

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomTMM$E,labels = x1$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],2),dim.plot = c(3,4),main=" MDS plot of TOTAL treatment with \nvoom TMM normalized counts considering batch")


## 3.1   DE analysis------

### estimate dispersion
par(mfrow=c(1,1))
plotMD(x1_TMMwzp,column = 2)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)
#make contrasts
con 
### DE MSS_h_MSI
MSS_h_MSI <- makeContrasts(MSS_high-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_h_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
 # write_delim(path = "MSS_h_v_MSI_top100.txt",delim = "\t")

### DE MSS_H_v_L
MSS_H_v_L <- makeContrasts(MSS_high-MSS_low,levels = design)
res <- glmQLFTest(fit,contrast = MSS_H_v_L)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA")  
 # write_delim(path = "MSS_H_v_MSS_L_top100.txt",delim = "\t")
  
### DE MSS_L_v_MSI
MSS_L_v_MSI <- makeContrasts(MSS_low-MSI_no,levels = design)
res <- glmQLFTest(fit,contrast = MSS_L_v_MSI)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top<- topTags(res,n=100)
#### save the results top100
top$table %>% 
  as_tibble(rownames = "smallRNA") #%>% 
 write_delim(path = "MSS_L_v_MSI_top100.txt",delim = "\t")  
  
 
#  3.2   limma quality weights--------
vfit <- lmFit(x1_voomTMM,design=design)
con_mat <- makeContrasts(
  tre_non = treated-non_treated,
  levels = design
)
vfit <- contrasts.fit(vfit,con_mat)
efit <- eBayes(vfit,robust = TRUE)

write_rds(efit,str_glue("efit_x1_voomTMM_filt2_{date}.rds"))
efit <- read_rds("efit_x1_voomTMM_filt2.rds")

#treated-non_treated
topTreat_vs_no <- topTable(efit,coef = "tre_non",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA") %>% as_tibble()

topTreat_vs_no  %>% 
  filter(adj.P.Val<0.01,logFC>2) %>%
  select(smallRNA,logFC) 

write_delim(topTreat_vs_no,path=str_glue("Complete_list_Treat_vs_No_{name_file}_{date}.txt"),delim = "\t")

#  3.3   visualization of results--------
fil <- read_delim(str_glue("Complete_list_Treat_vs_No_TOTAL_Mar_24_2019.txt"),
                  delim = "\t") 

# set values of filtering
lfc = 2
fdr = 0.01

methylated <- fil %>% 
  filter(logFC > lfc, adj.P.Val < fdr)
not_methyl <- fil %>% 
  filter(logFC < -lfc, adj.P.Val < fdr)
partMeth <- fil %>% 
  filter(between(logFC,-lfc,lfc))

x1_voomTMM <- read_rds(str_glue("x1_voomTMM_filt2_TOTAL_Mar_24_2019.rds"))
Exs <- read_delim("cpm_TMM_TOTAL_filt2_Mar_24_2019.txt", delim = "\t")
Exs <- read_delim("lcpm_TMM_TOTAL_filt2_Mar_24_2019.txt", delim = "\t")
Exs <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA")


my_data <- inner_join(Exs,fil,by="smallRNA")
my_fdata <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(TOTAL_meanExpr = mean(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         TOTAL_sd = sd(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         TOTAL_mad = mad(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         TOTAL_NAIO4_meanExpr = mean(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         TOTAL_NAIO4_sd = sd(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         TOTAL_NAIO4_mad = mad(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         pval = ifelse(adj.P.Val< fdr,
                      "signi","notsignif"),
         Status = ifelse(smallRNA %in% methylated$smallRNA,"methylated",
                  ifelse(smallRNA %in% not_methyl$smallRNA,"not-methylated",
                  ifelse(smallRNA %in% partMeth$smallRNA && TOTAL_NAIO4_meanExpr > -0.55,
                 # ifelse(,
                                       "partially-methylated","not-significant"))))   %>% 
        
  select(smallRNA, TOTAL_meanExpr, TOTAL_sd,TOTAL_mad,
         TOTAL_NAIO4_meanExpr, TOTAL_NAIO4_sd, TOTAL_NAIO4_mad,
         logFC, AveExpr, adj.P.Val, pval ,Status) %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_")

         EnRNA=ifelse(smallRNA %in% enriched$smallRNA,
                      smallRNA,"not_Enriched_match_genome")) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  select(smallRNA,TOTAL_meanExpr,TOTAL_NAIO4_meanExpr,topTreat_vs_no,AveExpr,DB)



my_data %>% filter(str_detect(smallRNA,"piR")) %>%  
  ggplot()+
  geom_point(mapping=aes(x=TOTAL_meanExpr,
                         y=TOTAL_sd,
                         colour= pval))

# plot the enrichment 
my_fdata %>% 
ggplot()+
  geom_point(mapping=aes(x=TOTAL_meanExpr,
                         y=TOTAL_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(methylated)} Methylated smallRNAs in {name_file} after treament, FDR < {fdr}, LFC > {lfc}"),
     x="Non-Treated average log2 counts per million",
     y="Treated log2 average counts per million",
     color="small RNAs") +
  theme_bw()+
scale_color_manual(values = c("gold","cornflowerblue","grey","chartreuse4", "darkorchid1","firebrick1","darkorange1",  "orange"))
#1 ensembl #2 miRNA #3 noAnnot #4 not_Enriched ,"black"
#5 piRNA #6 rfam #7 rRNA #8 tRNA #9 tRNA_mature

# only pirnas
filtered <- filter(methylated,str_detect(smallRNA,"piRNA"),)
filtered <- filter(enriched,str_detect(smallRNA,"spike|piRNA"))
# plot the enrichment 

my_fdata %>% 
    filter(str_detect(DB,"piR")) %>%
ggplot()+
  geom_point(mapping=aes(x=TOTAL_meanExpr,
                         y=TOTAL_NAIO4_meanExpr,
                         colour= Status))+
labs(title =str_glue("{nrow(filtered)} Methylated piRNAs in {name_file} after treament: FDR < {fdr}, LFC > {lfc}\n partially methylated : cpm > 0.55, |LFC| >= 2"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="piRNA Status") +
  theme_bw()+
scale_color_manual(
                   values = c("chartreuse4", "black", "gold","cornflowerblue","grey", "darkorchid1","firebrick1","darkorange1",  "orange"))


#  4.1   Compare results to find methylated IPP ---------
setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/")

LFCstable <- read_delim("LFCs_tables_FDR001_treatvsnon.txt",delim = "\t")

#LFCstable %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% filter(DB=="noAnnot")
#filter(str_detect(smallRNA,"piRNA_m")) %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% select(sRNA) %>%  write_delim(path="IPP_pirna_targets.txt",delim = "\t")

MetOtal <- LFCstable %>% 
  filter(logFC >= 0) %>% 
  rename(topTreat_vs_no=logFC)

setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/")

IPP <- read_delim("LFCs_C_N_2_input_2_results.txt",delim = "\t")
IPP %>% 
  filter(smallRNA %in% MetOtal$smallRNA) %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

MethIPP <- inner_join(IPP,MetOtal,by="smallRNA")
write_delim(MethIPP,path="Methylated_IPP_tables_FDR001_filt2.txt",delim = "\t")

#  4.2   Create a file with sequences-----
fasta <- readDNAStringSet("/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa")
IPPpirna<- IPP %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

IPPpirna$smallRNA <- str_replace(IPPpirna$smallRNA,"piRNA_match_","")

fasta <- fasta[IPPpirna$smallRNA]
writeXStringSet(fasta,"IPPpirna.fa")


```

# 2nd Dataset, CYTOSOL Periodate Treatment                                  
```{2nd Dataset:Periodate Treatment CYTOSOL}
#  1.1   List files merged and outputs of all ----------
library(data.table);library(tidyverse);library(rafalib)
setwd("/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/")

list_smallRNA <- list.files(pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments IPP_PIWIL1
list_smallRNA[str_detect(list_smallRNA,"CITOSOL")] -> listRNA


#load data to data.table
DTRNA <- rbindlist( sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

## 1.1.a Using dplyr--------------------------------------------
DTRNA <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DTRNA %>% 
  as_tibble %>% 
  dplyr::group_by(FileName,V6) %>% 
  dplyr::mutate(count = V2 / n()) %>% 
  dplyr::group_by(FileName,V4) %>% 
  dplyr::mutate( counts = sum(count))  %>% 
  dplyr::group_by(FileName,RNA_type,V4,counts) %>% 
  dplyr::summarise() %>%
  dplyr::arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT2
#------evaluate the resulted summaries
DT2 %>% 
 dplyr::group_by(FileName) %>% 
 dplyr::summarise(sums=sum(counts))

#--------change table from long to wide
DT2 %>% spread(FileName,RNA_type,counts, fill = 0) -> DT2
gsub("_smallRNA.txt","",basename(colnames(DT2))) ->colnames(DT2)
gsub("_Pool_Non_Treated","",colnames(DT2)) ->colnames(DT2)

# write for all
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt"
DT2 %>%
 write_delim(path,delim = "\t")



## 1.1.b using data table------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,"CITOSOL")]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as.numeric(sumsDT$Reads),as.numeric(newSUms$sums)))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)

colnames(DT_last)
sum(DT_last[,CITOSOL_5C_NaIO4])

path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/Treated_non_Rawcounts.txt" # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  2.1   Data pre-processing with edger--------------
## 2.1.0 Data packaging----------
library(edgeR);library(wesanderson)
#DT2 <- as_tibble(fread(path ,header = TRUE))
#DT2 <- as_tibble(DT_last)
DT2 <- DT2 %>%
  as_tibble %>%  column_to_rownames("smallRNA")

mat <- as.matrix(DT2)

group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))

design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)

## 2.1.1 filtering ------------------------------------
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
summary(lcpm)
table(rowSums(x$counts==0)==6)

keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

## 2.1.2 density plot-------------------------  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data, ~{round(lcpm.cutoff,3)} cpm"), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


## 2.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 2.1.4 normalization---------------------------------------------------------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/")

nafi <- "CYTOSOL"
cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{nafi}.txt"),sep = "\t",col.names = TRUE)

write_rds(x1_voomTMM,str_glue("x1_voomTMM_filt2_{nafi}.rds"))
write_rds(x1_voomQ,str_glue("x1_voomQ_filt2_{nafi}.rds"))
x1_voomTMM <- read_rds(str_glue("x1_voomTMM_filt2_{nafi}.rds"))

## 2.1.5 boxplots---------------------------------------------------
par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(1,3))
#voomQ data
boxplot(x1_voomQ$E, las=2, col= wes_palettes$Royal1[1:3], main="",,names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")


## save the matrices
#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMM_logcpm_{nafi}.txt"),delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_logcpm_{nafi}.txt"),delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_cpm_{nafi}.txt"),delim = "\t")

## 2.1.6 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMM))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, TMM")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 2.1.7 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomQ$E,labels = group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(1,2))

legend("topleft",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomQ$E,labels = x1$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(3,4),main=" MDS of Cytosol Periodate with voom Quantile \n normalized counts considering batch")


## 3.1   DE analysis------

### estimate dispersion
par(mfrow=c(1,1))
plotMD(x1_TMMwzp,column = 2)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)
# make contrasts
# DE tre_non
tre_non <- makeContrasts(treated-non_treated,levels = design)
res <- glmQLFTest(fit,contrast = tre_non)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top <- topTags(res,n=nrow(res$table))

#### save the results 
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
  filter(logFC >= 0, FDR < 0.01) %>%
  select(smallRNA,logFC,logCPM,FDR) %>% 
write_delim(path = str_glue("DEnriched_FDR001_{nafi}_edgeR.txt"),delim = "\t")

 
#  3.3.0 limma quality weights--------
vfit <- lmFit(x1_voomTMM,design=design)
con_mat <- makeContrasts(
  tre_non = treated-non_treated,
  levels = design)
vfit <- contrasts.fit(vfit,con_mat)
# check if to use trend = TRUE
max(x1_voomTMM$targets$lib.size)/min(x1_voomTMM$targets$lib.size)>3

# if TRUE (>3) don't use trend = TRUE
# apply eBayes 
efit <- eBayes(vfit,robust = TRUE)
plotSA(efit,main=str_glue("{nafi} Final model: Mean-variance trend"))
summary(decideTests(efit,p.value = 0.05))
summary(decideTests(efit,p.value = 0.05,lfc=2))
summary(decideTests(efit,p.value = 0.01))
summary(decideTests(efit,p.value = 0.01,lfc=2))

#tfit <- treat(vfit,lfc = 2)
#summary(decideTests(tfit))
  
write_rds(efit,str_glue("efit_x1_voomTMM_filt2_{nafi}.rds"))
efit <- read_rds("efit_x1_voomQ_filt2_CYTOSOL.rds")
efit <- read_rds(str_glue("efit_x1_voomTMM_filt2_{nafi}.rds"))

#treated-non_treated
topTreat_vs_no <- topTable(efit,coef = "tre_non",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA")

topTreat_vs_no %>% 
  filter(adj.P.Val<0.05,logFC>2) %>%
  select(smallRNA,logFC) -> LFCstable

write_delim(topTreat_vs_no,path=str_glue("LFCs_voomTMM_filt2_treatvsnon_{nafi}.txt"),delim = "\t")

#  3.3.1 visualization of results--------
fil <- read_delim(str_glue("LFCs_voomTMM_filt2_treatvsnon_{nafi}.txt"),delim = "\t") %>% 
  rename(topTreat_vs_no=logFC) 

# set values of filtering
lfc = 0
fdr = 0.01

enriched <- fil %>% 
  filter( topTreat_vs_no >= lfc, adj.P.Val < fdr)

x1_voomTMM <- read_rds(str_glue("x1_voomTMM_filt2_{nafi}.rds"))

Exs <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA")

my_data <- inner_join(Exs,fil,by="smallRNA")

my_data <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(CYTOSOL_meanExpr = mean(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         CYTOSOL_NAIO4_meanExpr = mean(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4))) %>% 
  select(smallRNA,CYTOSOL_meanExpr,CYTOSOL_NAIO4_meanExpr,topTreat_vs_no)

my_data %>%  
ggplot()+
  geom_point(mapping=aes(x=CYTOSOL_meanExpr,
                         y=CYTOSOL_NAIO4_meanExpr,
                         colour= smallRNA %in% enriched$smallRNA))+
labs(title =str_glue("{nrow(enriched)} Enriched smallRNAs in {nafi} after treament, FDR < {fdr}, LFC >= {lfc}"),
     x="Non-Treated log2 average expression",
     y="Treated log2 average expression",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=c("not Enriched","Enriched"),values = c("black","orange"))

#  4.1   Compare results to find methylated IPP ---------
setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/")

enriched <- read_delim(str_glue("LFCs_voomTMM_filt2_treatvsnon_{nafi}.txt"),delim = "\t") %>% 
  rename(topTreat_vs_no=logFC) %>% 
  filter( topTreat_vs_no >= lfc, adj.P.Val < fdr)

#LFCstable %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% filter(DB=="noAnnot")
#filter(str_detect(smallRNA,"piRNA_m")) %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% select(sRNA) %>%  write_delim(path="IPP_pirna_targets.txt",delim = "\t")

path <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/LFCs_C_N_2_input_2_results.txt"

IPP <- read_delim(path,delim = "\t")
IPP %>% 
  filter(smallRNA %in% enriched$smallRNA) %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

MethIPP <- inner_join(IPP,enriched,by="smallRNA") %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,topIPP_INPUT_vs_noAb,
         topTreat_vs_no_LFC = topTreat_vs_no,adj.P.Val)
write_delim(MethIPP,path=str_glue("Methylated_IPP_{nafi}_FDR001_filt2.txt"),delim = "\t")

#  4.2   Create a file with sequences-----
fasta <- readDNAStringSet("/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa")
IPPpirna<- IPP %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

IPPpirna$smallRNA <- str_replace(IPPpirna$smallRNA,"piRNA_match_","")


fasta <- fasta[IPPpirna$smallRNA]
writeXStringSet(fasta,"IPPpirna.fa")


```


# 2nd Dataset, NUCLEUS Periodate Treatment                                  
```{2nd Dataset:Periodate Treatment NUCLEUS}
#  1.1   List files merged and outputs of all ----------
# first select the correct tables
name_file <- "NUCLEUS"
data_filt <- "NUCLEO"
library(data.table);library(tidyverse);library(rafalib)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_NUCLEUS_treatment/")

path <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/2_Periodate_treat_all_DBs/"
list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments IPP_PIWIL1
listRNA <- list_smallRNA[str_detect(list_smallRNA,data_filt)]

#load data to data.table
listRNA <- sapply(listRNA, function(x)str_glue({path},{x}),simplify=FALSE)
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

## 1.1.a Using dplyr--------------------------------------------
DTRNA <- DTRNA %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DTRNA %>% 
  as_tibble %>% 
  dplyr::group_by(FileName,V6) %>% 
  dplyr::mutate(count = V2 / n()) %>% 
  dplyr::group_by(FileName,V4) %>% 
  dplyr::mutate( counts = sum(count))  %>% 
  dplyr::group_by(FileName,RNA_type,V4,counts) %>% 
  dplyr::summarise() %>%
  dplyr::arrange(desc(counts)) %>% 
  dplyr::rename(smallRNA=V4) -> DT2
#------evaluate the resulted summaries
DT2 %>% 
 dplyr::group_by(FileName) %>% 
 dplyr::summarise(sums=sum(counts))

#--------change table from long to wide
DT2 %>% spread(FileName,RNA_type,counts, fill = 0) -> DT2
gsub("_smallRNA.txt","",basename(colnames(DT2))) ->colnames(DT2)
gsub("_Pool_Non_Treated","",colnames(DT2)) ->colnames(DT2)

# write for all
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt"
DT2 %>%
 write_delim(path,delim = "\t")



## 1.1.b using data table------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,data_filt)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as.integer(sumsDT$Reads),as.integer(newSUms$sums)))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,NUCLEO_5N])

path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_NUCLEUS_treatment/Treated_non_{name_file}_raw_counts.txt") # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  2.1   Data pre-processing with edger----------------------------------------
## 2.1.0 Data packaging----------
library(edgeR);library(wesanderson)
#DT2 <- as_tibble(fread(path ,header = TRUE))
#DT2 <- as_tibble(DT_last) %>%  column_to_rownames("smallRNA")
DT2 <- DT2 %>%
  as_tibble %>%  column_to_rownames("smallRNA")

mat <- as.matrix(DT2)

group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))

design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)

## 2.1.1 filtering ------------------------------------
cpm <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
summary(lcpm)
table(rowSums(x$counts==0)==6)

keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

## 2.1.2 density plot-------------------------  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data of {name_file}, ~ {round(lcpm.cutoff,3)} log-cpm"), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


## 2.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 2.1.4 normalization---------------------------------------------------------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{name_file}.txt"),sep = "\t",col.names = TRUE)
cpm_TMMwzp %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMMwzp_{name_file}.txt"),sep = "\t",col.names = TRUE)

write_rds(x1_voomTMM,str_glue("x1_voomTMM_filt2_{name_file}.rds"))
write_rds(x1_voomQ,str_glue("x1_voomQ_filt2_{name_file}.rds"))
x1_voomTMM <- read_rds("x1_voomTMM_filt2_CYTOSOL.rds")

## 2.1.5 boxplots---------------------------------------------------
par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(1,3))
#raw data
boxplot(lcpm, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= wes_palettes$Royal1[1:3], main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")
#voomQ data
boxplot(x1_voomQ$E, las=2, col= wes_palettes$Royal1[1:3], main="",,names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")


## save the matrices
#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMM_logcpm_{name_file}.txt"),delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_logcpm_{name_file}.txt"),delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_cpm_{name_file}.txt"),delim = "\t")

## 2.1.6 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMM))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, TMM")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],4),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 2.1.7 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomTMM$E,labels = group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(1,2))

legend("topleft",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomQ$E,labels = x1$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],4),dim.plot = c(3,4),main=" MDS of Cytosol Periodate with voom Quantile \n normalized counts considering batch")


## 3.1   DE analysis------

### estimate dispersion
par(mfrow=c(1,1))
plotMD(x1_TMMwzp,column = 2)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)
# make contrasts
# DE tre_non
tre_non <- makeContrasts(treated-non_treated,levels = design)
res <- glmQLFTest(fit,contrast = tre_non)
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top <- topTags(res,n=nrow(res$table))

#### save the results 
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
  filter(logFC >= 0, FDR < 0.01) %>%
  select(smallRNA,logFC,logCPM,FDR) %>% 
write_delim(path = str_glue("DEnriched_FDR001_{nafi}.txt"),delim = "\t")

 
#  3.3.0 limma quality weights--------
vfit <- lmFit(x1_voomTMM,design=design)
con_mat <- makeContrasts(
  tre_non = treated-non_treated,
  levels = design)
vfit <- contrasts.fit(vfit,con_mat)
vfit <- eBayes(vfit,robust = TRUE)

write_rds(vfit,str_glue("vfit_x1_voomTMM_filt2_{name_file}.rds"))
vfit <- read_rds("vfit_x1_voomQ_filt2_CYTOSOL.rds")
vfit <- read_rds("vfit_x1_voomTMM_filt2_CYTOSOL.rds")

#treated-non_treated
topTreat_vs_no <- topTable(vfit,coef = "tre_non",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA")

topTreat_vs_no %>% 
  filter(adj.P.Val<0.05) %>%
  select(smallRNA,logFC) -> LFCstable

write_delim(topTreat_vs_no,path=str_glue("LFCs_voomTMM_filt2_treatvsnon_{name_file}.txt"),delim = "\t")


#  3.3.1 visualization of results--------
fil <- read_delim(str_glue("LFCs_voomTMM_filt2_treatvsnon_{name_file}.txt"),delim = "\t") %>% 
  rename(topTreat_vs_no=logFC) 

# set values of filtering
lfc = 0
fdr = 0.01

enriched <- fil %>% 
  filter( topTreat_vs_no >= lfc, adj.P.Val < fdr)

x1_voomTMM <- read_rds(str_glue("x1_voomTMM_filt2_{name_file}.rds"))

Exs <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA")

my_data <- inner_join(Exs,fil,by="smallRNA")

my_data <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(NUCLEUS_meanExpr = mean(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         NUCLEUS_NAIO4_meanExpr = mean(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4))) %>% 
  select(smallRNA,NUCLEUS_meanExpr,NUCLEUS_NAIO4_meanExpr,topTreat_vs_no,AveExpr)

my_data %>%  
ggplot()+
  geom_point(mapping=aes(x=NUCLEUS_meanExpr,
                         y=NUCLEUS_NAIO4_meanExpr,
                         colour= smallRNA %in% enriched$smallRNA))+
labs(title =str_glue("{nrow(enriched)} Enriched smallRNAs in {name_file} after treament, FDR < {fdr}, LFC >= {lfc}"),
     x="Non-Treated log2 average expression",
     y="Treated log2 average expression",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=c("not Enriched","Enriched"),values = c("black","orange"))

#  4.1   Compare results to find methylated IPP ---------
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_NUCLEUS_treatment/")

enriched <- read_delim(str_glue("LFCs_voomTMM_filt2_treatvsnon_{name_file}.txt"),delim = "\t") %>% 
  rename(topTreat_vs_no=logFC) %>% 
  filter( topTreat_vs_no >= lfc, adj.P.Val < fdr)

#LFCstable %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% filter(DB=="noAnnot")
#filter(str_detect(smallRNA,"piRNA_m")) %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% select(sRNA) %>%  write_delim(path="IPP_pirna_targets.txt",delim = "\t")

path <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/LFCs_C_N_2_input_2_results.txt"

IPP <- read_delim(path,delim = "\t")
IPP %>% 
  filter(smallRNA %in% enriched$smallRNA) %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

MethIPP <- inner_join(IPP,enriched,by="smallRNA") %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,topIPP_INPUT_vs_noAb,
         topTreat_vs_no_LFC = topTreat_vs_no,adj.P.Val)
write_delim(MethIPP,path=str_glue("Methylated_IPP_{name_file}_FDR001_filt2.txt"),delim = "\t")

#  4.2   Create a file with sequences-----
fasta <- readDNAStringSet("/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa")
IPPpirna<- IPP %>% 
  filter(str_detect(smallRNA,"piRNA_m"))

IPPpirna$smallRNA <- str_replace(IPPpirna$smallRNA,"piRNA_match_","")


fasta <- fasta[IPPpirna$smallRNA]
writeXStringSet(fasta,"IPPpirna.fa")


```

# 2nd Dataset, CYT VS NUC Periodate Treatment                               
```{r}
name_file <- "CYT_vs_NUC"
library(data.table);library(tidyverse);library(rafalib)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")
#  1.1   List files merged and outputs of all ----------
# first select the correct tables
data_filt <- "CITOSOL|NUCLEO"
path <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/2_Periodate_treat_all_DBs/"

list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments CITOSOL|NUCLEO
listRNA <- list_smallRNA[str_detect(list_smallRNA,data_filt)]

#load data to data.table
listRNA <- sapply(listRNA, function(x)str_glue({path},{x}),simplify=FALSE)
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

## 1.1.b using data table------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>%as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,data_filt)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as_vector(sumsDT$Reads),as_vector(round(newSUms$sums))))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,NUCLEO_5N])

path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/Treated_non_{name_file}_raw_counts.txt") # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  2.1   Data pre-processing with edger----------------------------------------
## 2.1.0 Data packaging----------
library(edgeR);library(wesanderson)
#DT2 <- as_tibble(DT_last) %>%  column_to_rownames("smallRNA")
path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/Treated_non_{name_file}_raw_counts.txt")

#load the file
DT2 <- as_tibble(fread(path ,header = TRUE)) %>%
   column_to_rownames("smallRNA")

mat <- as.matrix(DT2)

treatment <- as.factor(rep(c("treated","non_treated"),6))
component <- as.factor(rep(c("Cyt","Nuc"),each=6))
group <- as.factor(str_glue("{treatment}_{component}"))
batch <- as.factor(rep(c(1:3),each=2,2))
samples <- data.frame(group,batch,row.names = colnames(mat))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)

x <- DGEList(counts = mat,
             samples = samples,
             lib.size = colSums(mat),
             norm.factors = rep(1,ncol(mat)))

## 2.1.1 filtering ------------------------------------
cpm  <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
summary(lcpm)
table(rowSums(x$counts==0)==12)

keep.exprs <- filterByExpr.DGEList(x,group = group)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)
table(rowSums(x1$counts==0)==6)

## 2.1.2 density plot-------------------------  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data of {name_file}, ~ {round(lcpm.cutoff,3)} log-cpm"), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


## 2.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 2.1.4 normalization---------------------------------------------------------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voomWithQualityWeights(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

#new
x1_voomTMM <- voom(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voom(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voom(x1,design = design,normalize.method = "quantile",plot = TRUE)


lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{name_file}.txt"),sep = "\t",col.names = TRUE)
lcpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("lcpm_TMM_{name_file}_{date}.txt"),sep = "\t",col.names = TRUE)
cpm_TMMwzp %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMMwzp_{name_file}.txt"),sep = "\t",col.names = TRUE)


write_rds(x1_voomQ,str_glue("x1_voomQ_filt2_{name_file}.rds"))
x1_voomQ <- read_rds(str_glue("x1_voomQ_filt2_{name_file}.rds"))
x1_TMM %>% write_rds(str_glue("x1_TMM_{name_file}_{date}.rds"))
## 2.1.5 boxplots---------------------------------------------------
par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(2,2))
#raw data
boxplot(lcpm, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")
#voomTMMwzp data
boxplot(x1_voomTMMwzp$E, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Voom TMMwzp Normalized filtered data",ylab="Log-cpm")
#voomQ data
boxplot(x1_voomQ$E, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")

## save the matrices
#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMM_logcpm_{name_file}.txt"),delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_logcpm_{name_file}.txt"),delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_cpm_{name_file}.txt"),delim = "\t")
#x1_voomTMMwzp
x1_voomTMMwzp$E %>% as_tibble(rownames="smallRNA") %>% 
  write_delim(path = str_glue("x1_voomTMMwzp_lcpm_{name_file}.txt"),delim = "\t")
  
## 2.1.6 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(x1_voomQ$E))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],each=2,2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],each=2,2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, voomQ")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = rep(wes_palettes$Royal1[c(1,2,4)],each=2,2),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

legend("topright",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 2.1.7 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomQ$E,labels = group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],each=2,2),dim.plot = c(1,2))

legend("left",legend=levels(batch),
       fill =wes_palettes$Royal1[c(1,2,4)],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomQ$E,labels = x1$samples$group,
                col=rep(wes_palettes$Royal1[c(1,2,4)],each=2,2),dim.plot = c(3,4),main=" MDS plot of Cyt Nuc Periodate with voom Quantile \n normalized counts considering batch")

## 3.1   DE analysis edgeR------

### estimate dispersion
par(mfrow=c(1,1))
plotMD(x1_TMMwzp,column = 2)
abline(h=0, col="red", lty=2, lwd=2)
x1_TMMwzp<- estimateDisp(x1_TMMwzp,design=design,  robust=TRUE)
plotBCV(x1_TMMwzp)
fit <- glmQLFit(x1_TMMwzp,design, robust = TRUE)
head(fit$coefficients)
plotQLDisp(fit)
# make contrasts
# DE tre_non
con_mat <- makeContrasts(
  TrevsUntCyt = treated_Cyt-non_treated_Cyt,
  TrevsUntNuc = treated_Nuc-non_treated_Nuc,
  TreCytvsNuc = treated_Cyt-treated_Nuc,
  UntCytvsNuc = non_treated_Cyt-non_treated_Nuc,
  TrvsNTr_Cyt_VS_TrvsNTr_Nuc = (treated_Cyt-non_treated_Cyt)-(treated_Nuc-non_treated_Nuc),
  TUCYTNUC = (treated_Cyt-non_treated_Nuc)-(treated_Nuc-non_treated_Cyt),
  levels = design)
res <- glmQLFTest(fit,contrast = con_mat[,"TrevsUntCyt"])
topTags(res)
is.de <- decideTests(res)
summary(is.de)
plotMD(res,status = is.de,values = c(1,-1),
       col = c("red","blue"),legend="topright")
top <- topTags(res,n=nrow(res$table))

#### save the results 
nafi <- "edgeR"
top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
#  filter(logFC >= 0, FDR < 0.01) %>%
#  select(smallRNA,logFC,logCPM,FDR) %>% 
write_delim(path = str_glue("DE_{nafi}_TrevsUntCyt.txt"),delim = "\t")


DE_edgeR <- top$table %>% 
  as_tibble(rownames = "smallRNA") %>% 
  select(smallRNA,logFC,logCPM,FDR) 

 
#  3.2.0 limma quality weights--------
corfit <- duplicateCorrelation(x1_voomQ,design,block = batch)

fit <- lmFit(x1_voomQ,design,block = batch,correlation = corfit$consensus.correlation)
con_mat <- makeContrasts(
  TrevsUntCyt = treated_Cyt-non_treated_Cyt,
  TrevsUntNuc = treated_Nuc-non_treated_Nuc,
  TreCytvsNuc = treated_Cyt-treated_Nuc,
  UntCytvsNuc = non_treated_Cyt-non_treated_Nuc,
  TrvsNTr_Cyt_VS_TrvsNTr_Nuc = (treated_Cyt-non_treated_Cyt)-(treated_Nuc-non_treated_Nuc),
  TUCYTNUC = (treated_Cyt-non_treated_Nuc)-(treated_Nuc-non_treated_Cyt),
  levels = design)
vfit <- contrasts.fit(fit,con_mat)
efit <- eBayes(vfit,robust = TRUE)

write_rds(efit,str_glue("efit_x1_voomQ_{name_file}_{date}.rds"))

contrast_list<- sapply(as.list(colnames(efit)),
       function(x)topTable(efit,coef = x,
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA"),simplify = FALSE)
names(contrast_list) <- colnames(con_mat)

# save the list
write_rds(contrast_list,str_glue("contrast_list_x1_voomQ_CYT_NUC_{date}.rds"))
# contrast_list <- read_rds(str_glue("contrast_list_x1_voomQ_{name_file}.rds"))

#  3.2.1 extract DEs for Unt & Tre Cyt vs Nuc-----
efit <- read_rds(str_glue("efit_x1_voomQ_{name_file}.rds"))
contrast_list <- read_rds(str_glue("contrast_list_x1_voomQ_{name_file}.rds"))
# list with DEs


# find DEs for Unt Cyt vs Nuc & Tre Cyt vs Nuc
DE_Unt_CytvsNuc <- contrast_list[["UntCytvsNuc"]] %>% 
  filter(logFC > 2 , adj.P.Val < 0.01) %>% 
  dplyr::select(smallRNA,
                UntlogFC=logFC,
                UntAveExpr=AveExpr,
                Untadj.P.Val=adj.P.Val) #3692

DE_Tre_CytvsNuc <- contrast_list[["TreCytvsNuc"]] %>% 
  filter(logFC > 2 , adj.P.Val < 0.01) %>% 
 dplyr::select(smallRNA,
               TrelogFC=logFC,
               TreAveExpr=AveExpr,
               Treadj.P.Val=adj.P.Val) #429

# find intersection of DEs Unt & Tre, Cyt vs Nuc
Common_Tre_Unt_CytvsNuc <- inner_join(DE_Tre_CytvsNuc,DE_Unt_CytvsNuc, by = "smallRNA") # edw exoume 344

# the same as above but through limma
DE_Tre_Unt_CytvsNuc <- contrast_list[["TUCYTNUC"]] %>% 
    filter(logFC > 2 , adj.P.Val < 0.01) %>% 
   dplyr::select(smallRNA,logFC,AveExpr,adj.P.Val)#  4,160

# check the intersection of the two above
inner_join(DE_Tre_Unt_CytvsNuc,Common_Tre_Unt_CytvsNuc, by = "smallRNA")#344 

# save table DE_UntCytvsNuc
write_delim(DE_Unt_CytvsNuc,path=str_glue("DE_Unt_CytvsNuc_x1_voomQ_{name_file}.txt"),delim = "\t")

# save table DE_TreCytvsNuc
write_delim(DE_Tre_CytvsNuc,path=str_glue("DE_Tre_CytvsNuc_x1_voomQ_{name_file}.txt"),delim = "\t")

# save table Common_Tre_Unt_CytvsNuc
write_delim(Common_Tre_Unt_CytvsNuc,path=str_glue("Common_Tre_Unt_CytvsNuc_x1_voomQ_{name_file}.txt"),delim = "\t")

#  3.2.2 extract DEs for Tre vs Unt Cyt-----

DE_TrevsUntCyt <- contrast_list[["TrevsUntCyt"]] %>% 
  filter(logFC > 2 , adj.P.Val < 0.01) 

# save table DE_TrevsUntCyt
write_delim(DE_TrevsUntCyt,path=str_glue("DE_TrevsUnt_Cyt_x1_voomQ_{name_file}.txt"),delim = "\t")

# find common between the small and complete dataset
path = "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/LFCs_voomTMM_filt2_treatvsnon_CYTOSOL.txt"

DE_CYT_trVSuntr <- read_delim(path,delim = "\t") %>% 
  filter(logFC > 1 , adj.P.Val < 0.01) %>% dplyr::select(smallRNA,CYTlogFC=logFC,CYTAveExpr=AveExpr,CYTadj.P.Val=adj.P.Val)

commonDE_CYTtrVSuntr <- inner_join(DE_CYT_trVSuntr,DE_TrevsUntCyt, by = "smallRNA")

nrow(commonDE_CYTtrVSuntr)/nrow(DE_TrevsUntCyt)*100
nrow(commonDE_CYTtrVSuntr)/nrow(DE_CYT_trVSuntr)*100

# write the complete file for visualization later
complete_TrevsUntCyt <- contrast_list[["TrevsUntCyt"]] 
# save table complete_TrevsUntCyt
write_delim(complete_TrevsUntCyt,path=str_glue("complete_TrevsUnt_Cyt_x1_voomQ_{name_file}_{date}.txt"),delim = "\t")

#common nuc and cyt
inner_join(DE_TrevsUntCyt,DE_TrevsUntNuc, by = "smallRNA")
#union nuc and cyt
full_join(DE_TrevsUntCyt,DE_TrevsUntNuc, by = "smallRNA") %>% 
  filter(str_detect(smallRNA,"piRNA"))


#  3.2.3 extract DEs for Tre vs Unt Nuc-----
DE_TrevsUntNuc <- contrast_list[["TrevsUntNuc"]] %>% 
  filter(logFC > 2, adj.P.Val < 0.01)

# save table DE_TrevsUntNuc
write_delim(DE_TrevsUntNuc,path=str_glue("DE_TrevsUnt_Nuc_x1_voomQ_{name_file}.txt"),delim = "\t")

# find common between the small and complete dataset
path = "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_NUCLEUS_treatment/LFCs_voomTMM_filt2_treatvsnon_NUCLEUS.txt"

DE_NUC_trVSuntr <- read_delim(path,delim = "\t") %>% 
  filter(logFC > 1 , adj.P.Val < 0.01) %>% dplyr::select(smallRNA,NUClogFC=logFC,NUCAveExpr=AveExpr,NUCadj.P.Val=adj.P.Val)

commonDE_NUCtrVSuntr <- inner_join(DE_NUC_trVSuntr,DE_TrevsUntNuc, by = "smallRNA")

nrow(commonDE_NUCtrVSuntr)/nrow(DE_TrevsUntNuc)*100
nrow(commonDE_NUCtrVSuntr)/nrow(DE_NUC_trVSuntr)*100

# write the complete file for Visualization later
complete_TrevsUntNuc <- contrast_list[["TrevsUntNuc"]] 
# save table complete_TrevsUntNuc
write_delim(complete_TrevsUntNuc,path=str_glue("complete_TrevsUnt_Nuc_x1_voomQ_{name_file}_{date}.txt"),delim = "\t")

#  3.3.1 visualization of results prepare data Cytosol --------
name_file <- "CYT_vs_NUC"
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")
# load files
filCyt <- read_delim(str_glue("complete_TrevsUnt_Cyt_x1_voomQ_CYT_vs_NUC_Mar_24_2019.txt"),delim = "\t") 
x1_TMM <- read_rds("x1_TMM_CYT_vs_NUC_Mar_24_2019.rds")
# set values of filtering
lfc = 2
fdr = 0.01
exprss <- cpm(10,mean(x1$samples$lib.size))

methylated <- filCyt %>% 
  filter(logFC > lfc, adj.P.Val < fdr)
not_methyl <- filCyt %>% 
  filter(logFC < -lfc, adj.P.Val < fdr)
partMeth <- filCyt %>% 
  filter(between(logFC,-lfc,lfc))

#load complete dataset
cpm_TMM <- read_delim(str_glue("cpm_TMM_CYT_vs_NUC.txt"),delim = "\t")
lcpm_TMM <- read_delim("lcpm_TMM_CYT_vs_NUC_Mar_24_2019.txt",delim = "\t")
  
# subset it to 2 for Cyt and Nuc
filExsCyt <- cpm_TMM %>% 
  select(smallRNA,starts_with("CIT")) %>% 
  filter_at(vars(starts_with("CIT")),all_vars(. == 0))

ExsCyt <- cpm_TMM %>% 
  select(smallRNA,starts_with("CIT")) %>% filter(!smallRNA %in% filExsCyt$smallRNA)

#join them for the visualization
my_data <- inner_join(ExsCyt,filCyt,by="smallRNA")

#find mean expr and subset columns for both datasets
my_dataCyt <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(Cyt_mean = log2(mean(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C))+0.1),
         Cyt_sd = sd(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_mad = mad(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_NAIO4_mean = log2(mean(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4))+0.1),
         Cyt_NAIO4_sd = sd(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         Cyt_NAIO4_mad = mad(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         pval = ifelse(adj.P.Val< fdr,
                      "signi","notsignif"),
         EnRNA=ifelse(smallRNA %in% methylated$smallRNA,
                      smallRNA,"not_significant_match_genome")) %>% 
  mutate(Status = ifelse(smallRNA %in% methylated$smallRNA,"methylated",
                  ifelse(smallRNA %in% not_methyl$smallRNA,"not-methylated",
                  ifelse(smallRNA %in% partMeth$smallRNA && Cyt_NAIO4_mean > log2(exprss),
                  "partially-methylated","not-significant")))) %>% 
  select(smallRNA, Cyt_mean, Cyt_sd,Cyt_mad,
         Cyt_NAIO4_mean, Cyt_NAIO4_sd, Cyt_NAIO4_mad,
         logFC, AveExpr, adj.P.Val, pval , Status, EnRNA) %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_") %>% 
  separate(EnRNA,c("DBs","sRNAs"), sep = "_match_")

my_dataNuc <- my_dataNuc %>% 
  group_by(smallRNA) %>% 
  mutate(NUCLEUS_meanExpr = mean(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         NUCLEUS_NAIO4_meanExpr = mean(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4)),
         EnRNA=ifelse(smallRNA %in% enrichedNuc$smallRNA,
                      smallRNA,"not_significant_match_genome")) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>%
  select(smallRNA,NUCLEUS_meanExpr,NUCLEUS_NAIO4_meanExpr,Cyt_NuclogFC,Cyt_NucAveExpr,DB)
#  3.3.2 visualization of results scatter of smallRNA Cytosol --------
my_dataCyt %>%  
ggplot()+
  geom_point(mapping=aes(x=Cyt_mean,
                         y=Cyt_NAIO4_mean,
                         colour= DBs ))+
labs(title =str_glue("{nrow(methylated)} smallRNAs found methylated  in Cytosol after treament, FDR < {fdr}, Relative ratio > {2^lfc}"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="smallRNAs") +
  theme_bw()+
scale_color_manual(values = c("gold","cornflowerblue","grey","black",
                              "chartreuse4", "darkorchid1","firebrick1",
                              "darkorange1",  "orange"))
#1 ensembl #2 miRNA #3 noAnnot #4 not_Enriched 
#5 piRNA #6 rfam #7 rRNA #8 tRNA #9 tRNA_mature
#  3.3.3 visualization of results scatter of piRNA Cytosol ---------
filteredCyt <- methylated %>%  filter(str_detect(smallRNA,"piRNA"))

my_dataCyt %>%  filter(str_detect(DB,"piR")) %>% 
ggplot()+
  geom_point(mapping=aes(x=Cyt_mean,
                         y=Cyt_NAIO4_mean,
                         colour= Status ))+
labs(title =str_glue("{nrow(filteredCyt)} Methylated piRNAs in {name_file} after treament: FDR < {fdr}, Relative ratio > {2^lfc}\n partially methylated : cpm > {round(exprss,2)}, |Relative ratio| >= {2^lfc}"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="piRNA Status") +
  theme_bw()+
scale_color_manual(
                   values = c("chartreuse4", "black", "gold","cornflowerblue","grey", "darkorchid1","firebrick1","darkorange1",  "orange"))

#  3.3.4 visualization of results scatter of piRNA with no FDR, but colour by LFC,Cytosol ---------
my_dataCyt_piR <- my_dataCyt %>%
  filter(str_detect(smallRNA,"piR")) %>% 
  mutate(EnRNA = ifelse(logFC > lfc,"methylated_piRNA_match_genome",
                        ifelse(logFC < -(lfc),"not_methylated_piRNA_match_genome",
                               "partially_methylated_match_genome"))) %>% 
  separate(EnRNA,c("Methylation","sRNA"), sep = "_match_") 

#add final column
my_dataCyt_piR <- my_dataCyt_piR %>% 
  mutate(sRNA = ifelse(DB == "not_significant",DB,Methylation))

my_dataCyt_piR %>% 
#  filter(Cyt_meanExp < 15,Cyt_NAIO4_meanExp < 15) %>% 
ggplot()+
  geom_point(mapping=aes(x=Cyt_meanExp,
                         y=Cyt_NAIO4_meanExp,
                         colour= sRNA ))+
labs(title =str_glue("{nrow(enrichedCyt)} piRNAs found statistically significant  in {nmC} after treament, FDR < {fdr} \n{nrow(filteredCyt)} piRNAs found methylated with relative log2 ratio > {lfc}"),
     x="Non-Treated log2 median cpm",
     y="Treated log2 median cpm",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(values = c("forestgreen","black","grey", "darkseagreen3","firebrick1",
                              "darkorange1"  ))

#  3.3.5 visualization of results Volcano plot treated--------
name_file <- "CYT_vs_NUC"
#threshold
FC <- 2
FDR <- 0.01

setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")
contrast_list <- read_rds(str_glue("contrast_list_x1_voomQ_CYT_NUC_Mar_24_2019.rds"))

# treated data
mydata <- contrast_list[["TreCytvsNuc"]] 
signifi <- mydata %>% filter(adj.P.Val < FDR)
enrch <- mydata %>% filter(abs(logFC) > FC)
finDat <- mydata %>%  group_by(smallRNA) %>% 
  mutate(EnRNA=ifelse(smallRNA %in% enrch$smallRNA &&
                      smallRNA %in% signifi$smallRNA,
                      smallRNA,
            ifelse(smallRNA %in% enrch$smallRNA, "not_significant_match_genome",
                   "not_Enriched_match_genome"))) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  mutate(DB=as.factor(DB)) %>% 
  select(smallRNA,logFC,adj.P.Val,DB)

#multiplot
cyt <- mydata %>% filter(adj.P.Val < FDR,logFC > FC)
nuc <- mydata %>% filter(adj.P.Val < FDR,logFC < -FC)
enrch2 <- enrch %>% filter(adj.P.Val < FDR) %>% 
  mutate(EnRNA=smallRNA) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  mutate(DB=as.factor(DB)) %>% 
  select(smallRNA,logFC,adj.P.Val,DB)

newList <- list(histogram = enrch2,
                'volcano plot'= finDat) %>% 
  rbindlist(use.names = TRUE, idcol = "facet") %>%  as_tibble()

ggplot()+
  facet_grid(rows = vars(facet), scales = "free_y")+
  geom_point(data = filter(newList,facet=="volcano plot"),
             mapping=aes(x = finDat$logFC,
                         y=-log10(adj.P.Val),
                         colour= DB),position = "jitter")+
  geom_histogram(data = filter(newList,facet=="histogram"), 
                 mapping=aes(x= enrch2$logFC, fill= enrch2$DB))+
  theme_bw()+
  scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "chartreuse4", "darkorchid1", "firebrick1",
                                "darkorange1",  "orange", "black","bisque4"))+
  scale_fill_manual(values = c("gold","cornflowerblue","grey","chartreuse4", "darkorchid1","firebrick1","darkorange1", "orange"))+
labs(title =str_glue("{nrow(cyt)} smallRNAs identified enriched in treated Cytosol,
                      {nrow(nuc)} smallRNAs identified enriched in treated Nucleus
                      FDR < {FDR}, |LFC| > 2"))

# piRNAs only
finDatpiR <- finDat %>% 
  filter(str_detect(smallRNA,"piRNA")) 

enrchpiR <- enrch2 %>% 
  filter(str_detect(smallRNA,"piRNA"))

newList1 <- list(histogram = enrchpiR,
                'volcano plot'= finDatpiR) %>% 
  rbindlist(use.names = TRUE, idcol = "facet") %>%  as_tibble()

ggplot()+
  facet_grid(rows = vars(facet), scales = "free_y")+
  geom_point(data = filter(newList1,facet=="volcano plot"),
             mapping=aes(x = finDatpiR$logFC,
                         y=-log10(adj.P.Val),
                         colour= DB),position = "jitter")+
  geom_histogram(data = filter(newList1,facet=="histogram"), 
                 mapping=aes(x= enrchpiR$logFC, fill= enrchpiR$DB))+
  theme_bw()+
  scale_color_manual(values = c("chartreuse4","black","grey"))+
  scale_fill_manual(values = c("chartreuse4"))+
  labs(title =str_glue("{nrow(filter(enrchpiR,logFC>0))} Enriched piRNAs in Cytosol, {nrow(filter(enrchpiR,logFC<0))} Enriched piRNAs in Nucleus after treament\n FDR < {FDR}, |LFC| > {FC}"),
     x="LogFC",
     color="Selected RNAs")
#  3.3.6 visualization of results Volcano plot non-treated--------
name_file <- "CYT_vs_NUC"
#threshold
FC <- 2
FDR <- 0.01

setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")
contrast_list <- read_rds(str_glue("contrast_list_x1_voomQ_CYT_NUC_Mar_24_2019.rds"))

mydata <- contrast_list[["UntCytvsNuc"]] 
signifi <- mydata %>% filter(adj.P.Val < FDR)
enrch <- mydata %>% filter(abs(logFC) > FC)
finDat <- mydata %>%  group_by(smallRNA) %>% 
  mutate(EnRNA=ifelse(smallRNA %in% enrch$smallRNA &&
                      smallRNA %in% signifi$smallRNA,
                      smallRNA,
            ifelse(smallRNA %in% enrch$smallRNA, "not_significant_match_genome",
                   "not_Enriched_match_genome"))) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  mutate(DB=as.factor(DB)) %>% 
  select(smallRNA,logFC,adj.P.Val,DB)

#multiplot
cyt <- mydata %>% filter(adj.P.Val < FDR,logFC > FC)
nuc <- mydata %>% filter(adj.P.Val < FDR,logFC < -FC)
enrch2 <- enrch %>% filter(adj.P.Val < FDR) %>% 
  mutate(EnRNA=smallRNA) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  mutate(DB=as.factor(DB)) %>% 
  select(smallRNA,logFC,adj.P.Val,DB)

newList <- list(histogram = enrch2,
                'volcano plot'= finDat) %>% 
  rbindlist(use.names = TRUE, idcol = "facet") %>%  as_tibble()

ggplot()+
  facet_grid(rows = vars(facet), scales = "free_y")+
  geom_point(data = filter(newList,facet=="volcano plot"),
             mapping=aes(x = finDat$logFC,
                         y=-log10(adj.P.Val),
                         colour= DB),position = "jitter")+
  geom_histogram(data = filter(newList,facet=="histogram"), 
                 mapping=aes(x= enrch2$logFC, fill= enrch2$DB))+
  theme_bw()+
  scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "chartreuse4", "darkorchid1", #"firebrick1",
                                "darkorange1",  "orange", "black","bisque4"))+
  scale_fill_manual(values = c("gold","cornflowerblue","grey",
                               "chartreuse4", "darkorchid1",#"firebrick1",
                               "darkorange1", "orange"))+
labs(title =str_glue("{nrow(cyt)} smallRNAs identified enriched in non-treated Cytosol,
                      {nrow(nuc)} smallRNAs identified enriched in non-treated Nucleus
                      FDR < {FDR}, |LFC| > 2"))

# piRNAs only
finDatpiR <- finDat %>% 
  filter(str_detect(smallRNA,"piRNA")) 

enrchpiR <- enrch2 %>% 
  filter(str_detect(smallRNA,"piRNA"))

newList1 <- list(histogram = enrchpiR,
                'volcano plot'= finDatpiR) %>% 
  rbindlist(use.names = TRUE, idcol = "facet") %>%  as_tibble()

ggplot()+
  facet_grid(rows = vars(facet), scales = "free_y")+
  geom_point(data = filter(newList1,facet=="volcano plot"),
             mapping=aes(x = finDatpiR$logFC,
                         y=-log10(adj.P.Val),
                         colour= DB),position = "jitter")+
  geom_histogram(data = filter(newList1,facet=="histogram"), 
                 mapping=aes(x= enrchpiR$logFC, fill= enrchpiR$DB))+
  theme_bw()+
  scale_color_manual(values = c("chartreuse4","black","grey"))+
  scale_fill_manual(values = c("chartreuse4"))+
  labs(title =str_glue("{nrow(filter(enrchpiR,logFC>0))} Enriched piRNAs in Cytosol, {nrow(filter(enrchpiR,logFC<0))} Enriched piRNAs in Nucleus before treament\n FDR < {FDR}, |LFC| > {FC}"),
     x="LogFC",
     color="Selected RNAs")


#  4.1   Compare results to find methylated IPP ---------
name_file <- "CYT_vs_NUC"
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")

# load the enriched entries
enrichedCyt <- read_delim(str_glue("complete_TrevsUnt_Cyt_x1_voomQ_{name_file}.txt"),delim = "\t") %>% 
  filter( Cyt_NuclogFC > lfc, Cyt_Nucadj.P.Val < fdr)
enrichedNuc <- read_delim(str_glue("complete_TrevsUnt_Nuc_x1_voomQ_{name_file}.txt"),delim = "\t")  %>% 
  filter( Cyt_NuclogFC > lfc, Cyt_Nucadj.P.Val < fdr)

#LFCstable %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% filter(DB=="noAnnot")
#filter(str_detect(smallRNA,"piRNA_m")) %>% separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% select(sRNA) %>%  write_delim(path="IPP_pirna_targets.txt",delim = "\t")

path <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/LFCs_C_N_2_input_2_results.txt"

IPP <- read_delim(path,delim = "\t")

# IPP and Treated Cyt
MethIPPcyt <- inner_join(IPP,enrichedCyt,by="smallRNA") %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,topIPP_INPUT_vs_noAb,
         Cyt_Treat_vs_NonTr = Cyt_NuclogFC,adj.P.Val=Cyt_Nucadj.P.Val)

# IPP and Treated Nuc
MethIPPnuc <- inner_join(IPP,enrichedNuc,by="smallRNA") %>% 
  select(smallRNA,topIPP_C_vs_INPUT,topIPP_N_vs_INPUT,topIPP_INPUT_vs_noAb,
         Nuc_Treat_vs_NonTr = Cyt_NuclogFC,adj.P.Val=Cyt_Nucadj.P.Val)

# save the tables
write_delim(MethIPPcyt,path=str_glue("Methylated_Cyt_IPP_{name_file}_FDR001.txt"),delim = "\t")
write_delim(MethIPPnuc,path=str_glue("Methylated_Nuc_IPP_{name_file}_FDR001.txt"),delim = "\t")

piMethIPPcyt <- MethIPPcyt %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% 
  filter(str_detect(DB,"piRNA")) %>% 
  select(-DB)

piMethIPPnuc <- MethIPPnuc %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_") %>% 
  filter(str_detect(DB,"piRNA")) %>% 
  select(-DB)

piMethIPPcyt %>% select(sRNA) %>%
  write_delim("piMethIPPcyt_list.txt",delim = "\t")

piMethIPPnuc %>% select(sRNA) %>%
  write_delim("piMethIPPnuc_list.txt",delim = "\t")

```
# Relative ratios  Cytosol treated                                          
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))

library(data.table);library(tidyverse);library(rafalib);library(edgeR);library(wesanderson)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/")
# 1.1 load the data ------
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_CYTOSOL_treatment/Treated_non_Rawcounts.txt"
DT2 <- as_tibble(fread(path ,header = TRUE)) %>%
  as_tibble %>%  column_to_rownames("smallRNA")

mat <- as.matrix(DT2)
group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
# 2.1 filtering ------------------------------------
cutoff <- cpm(10,mean(x$samples$lib.size))

keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)
# 2.2 normalization -----
x1_TMM <- calcNormFactors(x1, method = "TMM")
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE) %>% 
  as_tibble(rownames = "smallRNA")
# 2.3 find the relative ratio -----
cpm_TMM_Ratio <- cpm_TMM %>% 
  group_by(smallRNA) %>% 
  mutate(Cyt_mean = mean(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_median = median(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_sd = sd(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_mad = mad(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         Cyt_NAIO4_mean = mean(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         Cyt_NAIO4_median = median(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         Cyt_NAIO4_sd = sd(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         Cyt_NAIO4_mad = mad(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4))) 
select(smallRNA, Cyt_mean, Cyt_median, Cyt_sd, Cyt_NAIO4_mean, Cyt_NAIO4_median,
       Cyt_NAIO4_sd, Cyt_NAIO4_mad)
cpm_TMM_Ratio <- cpm_TMM_Ratio %>% 
  mutate(Ratio_mean = (Cyt_NAIO4_mean +0.1)/ (Cyt_mean+0.1),
         Ratio_median = (Cyt_NAIO4_median +0.1)/ (Cyt_median+0.1))
# 2.4 visualization of mean, median, etc----
#mean to median
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Ratio_mean ),
                         y = log2(Ratio_median),
                         colour= DB ))
# mean to sd
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Cyt_mean),
                         y = Cyt_sd,
                         colour= DB ))
# 2.5 visualization of methylated, not methylated smallRNAs, etc----
lfc <- 2
my_data_allRNA <- cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  mutate(Status = ifelse(
    Cyt_NAIO4_mean == 0,"not-methylated",
    ifelse(log2(Ratio_mean) > lfc,DB,
    ifelse(between(log2(Ratio_mean),-lfc,lfc),"partially-methylated",
           "not-methylated"))))  
my_data_allRNA %>% 
 # unite("smallRNA",DB,sRNA,sep = "_match_") %>% 
  write_delim(str_glue("Relative_ratio_Cytosol_{date}.txt"),delim = "\t")  

my_data_allRNA %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Cyt_mean + 0.1),
                         y = log2(Cyt_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(filter(my_data_allRNA,Status == DB))} smallRNAs found methylated in Cytosol after treament, Relative ratio > {2^lfc}"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="smallRNAs") +
  theme_bw() +
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" ))
#1 "gold" = ensembl #2 "cornflowerblue" = miRNA 
#3 "grey" = noAnnot 
#4 "black"  = not_Enriched 
#5 "chartreuse4" = piRNA  
#6 "darkorchid1" = rfam 
#7 "firebrick1" = rRNA 
#8 "darkorange1" = tRNA 
#9 "orange" = tRNA_mature

# 2.6 visualization of methylated, not methylated piRNAs, etc----
lfc <- 2
my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"piR|noAn")) %>% 
  mutate(Status = ifelse(Status == "noAnnot"|Status == "piRNA","methylated",Status)) %>% 
  write_tsv(str_glue("Relative_ratio_Cytosol_{date}.txt"))

my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"piR")) %>% 
  mutate(Status = ifelse(Status == "piRNA","methylated",Status)) 

pirnas <- filter(my_data,Status == "methylated")

my_data %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Cyt_mean + 0.1),
                         y = log2(Cyt_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(pirnas)} piRNAs found methylated in Cytosol after treament, Relative ratio > {2^lfc}"),
     x="Non-Treated piRNAs (log2 average cpm)",
     y="NaIO4 oxidized piRNAs (log2 average cpm)",
     color="smallRNAs") +
  theme_bw() +
  scale_y_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  scale_x_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  coord_equal()+
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" )[c(6,4,5)])

# 2.7 visualization of methylated, not methylated notAnnot, etc----
lfc <- 2
my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"noAn")) %>% 
  mutate(Status = ifelse(Status == "noAnnot","methylated",Status))

pirnas <- filter(my_data,Status == "methylated")

my_data %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Cyt_mean + 0.1),
                         y = log2(Cyt_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(pirnas)} not Annotated sequences found methylated in Cytosol after treament, Relative ratio > {2^lfc}"),
     x="Non-Treated not Annotated (log2 average cpm)",
     y="NaIO4 oxidized not Annotated (log2 average cpm)",
     color="smallRNAs") +
  theme_bw() +
  scale_y_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  scale_x_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  coord_equal()+
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" )[c(6,4,5)])

# 3.1 cross with IPP----
# load data
my_data <- read_delim("Relative_ratio_Cytosol_Mar_25_2019.txt", delim = "\t")

pathN <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/N_enriched_RNA.txt"
pathC <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/C_enriched_RNA.txt"

IPPN <- read_delim(pathN,delim = "\t")
IPPC <- read_delim(pathC,delim = "\t")
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% IPPC$smallRNA,str_detect(smallRNA,"piR"))
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% DEs$smallRNA)
DEs %>% filter(smallRNA %in% IPPN$smallRNA,
                  str_detect(smallRNA,"piR"))
```

# Relative ratios  Nucleus treated                                          
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))

library(data.table);library(tidyverse);library(rafalib);library(edgeR);library(wesanderson)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_NUCLEUS_treatment/")
# 1.1 load the data ------
path <- "Treated_non_NUCLEUS_raw_counts.txt"
DT2 <- as_tibble(fread(path ,header = TRUE)) %>%
  as_tibble %>%  column_to_rownames("smallRNA")
mat <- as.matrix(DT2)
group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
# 2.1 filtering ------------------------------------
cutoff <- cpm(10,mean(x$samples$lib.size))

keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)
# 2.2 normalization -----
x1_TMM <- calcNormFactors(x1, method = "TMM")
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE) %>% 
  as_tibble(rownames = "smallRNA")
# 2.3 find the relative ratio -----
cpm_TMM_Ratio <- cpm_TMM %>% 
  group_by(smallRNA) %>% 
  mutate(Nuc_mean = mean(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         Nuc_median = median(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         Nuc_sd = sd(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         Nuc_mad = mad(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         Nuc_NAIO4_mean = mean(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4)),
         Nuc_NAIO4_median = median(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4)),
         Nuc_NAIO4_sd = sd(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4)),
         Nuc_NAIO4_mad = mad(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4))) 
select(smallRNA, Nuc_mean, Nuc_median, Nuc_sd, Nuc_NAIO4_mean, Nuc_NAIO4_median,
       Nuc_NAIO4_sd, Nuc_NAIO4_mad)

cpm_TMM_Ratio <- cpm_TMM_Ratio %>% 
  mutate(Ratio_mean = (Nuc_NAIO4_mean +0.1)/ (Nuc_mean+0.1),
         Ratio_median = (Nuc_NAIO4_median +0.1)/ (Nuc_median+0.1))
# 2.4 visualization of mean, median, etc----
#mean to median
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Ratio_mean ),
                         y = log2(Ratio_median),
                         colour= DB ))
# mean to sd
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Nuc_mean),
                         y = Nuc_sd,
                         colour= DB ))
# 2.5 visualization of methylated, not methylated smallRNAs, etc----
lfc <- 2
compart <- "Nucleus"
my_data_allRNA <- cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  mutate(Status = ifelse(
    Nuc_NAIO4_mean == 0,"not-methylated",
    ifelse(log2(Ratio_mean) > lfc,DB,
    ifelse(between(log2(Ratio_mean),-lfc,lfc),"partially-methylated",
           "not-methylated"))))  
my_data_allRNA %>% 
 # unite("smallRNA",DB,sRNA,sep = "_match_") %>% 
  write_delim(str_glue("Relative_ratio_{compart}_{date}.txt"),delim = "\t")  

my_data_allRNA %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Nuc_mean + 0.1),
                         y = log2(Nuc_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(filter(my_data_allRNA,Status == DB))} smallRNAs found methylated in {compart} after treatment, Relative ratio > {2^lfc}"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="smallRNAs") +
  theme_bw() +
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" ))
#1 "gold" = ensembl #2 "cornflowerblue" = miRNA 
#3 "grey" = noAnnot 
#4 "black"  = not_Enriched 
#5 "chartreuse4" = piRNA  
#6 "darkorchid1" = rfam 
#7 "firebrick1" = rRNA 
#8 "darkorange1" = tRNA 
#9 "orange" = tRNA_mature

# 2.6 visualization of methylated, not methylated noAnnot, etc----
lfc <- 2
compart <- "Nucleus"
my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"piR|noAn")) %>% 
  mutate(Status = ifelse(Status == "noAnnot"|Status == "piRNA","methylated",Status)) %>% 
  write_tsv(str_glue("Relative_ratio_{compart}_{date}.txt"))

my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"noA")) %>% 
  mutate(Status = ifelse(Status == "noAnnot","methylated",Status))
    
pirnas <- filter(my_data,Status == "methylated")

my_data %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Nuc_mean + 0.1),
                         y = log2(Nuc_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(pirnas)} not Annotated sequences found methylated in {compart} after treatment, Relative ratio > {2^lfc}"),
      x="Non-Treated not Annotated sequences (log2 average cpm)",
     y="NaIO4 oxidized not Annotated sequences (log2 average cpm)",
     color="smallRNAs") +
  theme_bw() +
   scale_y_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  scale_x_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  coord_equal()+
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" )[c(6,4,5)])

# 3.1 cross with IPP----
# load data
my_data <- read_delim("Relative_ratio_Nucleus_Mar_26_2019.txt", delim = "\t")

pathN <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/N_enriched_RNA.txt"
pathC <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/C_enriched_RNA.txt"

IPPN <- read_delim(pathN,delim = "\t")
IPPC <- read_delim(pathC,delim = "\t")
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% IPPC$smallRNA,str_detect(smallRNA,"piR"))
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% DEs$smallRNA)
DEs %>% filter(smallRNA %in% IPPN$smallRNA,
                  str_detect(smallRNA,"piR"))
```

# Relative ratios  Total treated                                            
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))

library(data.table);library(tidyverse);library(rafalib);library(edgeR);library(wesanderson)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/")
# 1.1 load the data ------
path <- "TOTAL_raw_counts.txt"
DT2 <- as_tibble(fread(path ,header = TRUE)) %>%
  as_tibble %>%  column_to_rownames("smallRNA")
mat <- as.matrix(DT2)
group <- as.factor(rep(c("treated","non_treated"),3))
batch <- as.factor(rep(c(1:3),each=2))
samples <- data.frame(group,batch,row.names = colnames(mat))

x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
# 2.1 filtering ------------------------------------
cutoff <- cpm(10,mean(x$samples$lib.size))

keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)
# 2.2 normalization -----
x1_TMM <- calcNormFactors(x1, method = "TMM")
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE) %>% 
  as_tibble(rownames = "smallRNA")
# 2.3 find the relative ratio -----
cpm_TMM_Ratio <- cpm_TMM %>% 
  group_by(smallRNA) %>% 
  mutate(Tot_mean = mean(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         Tot_median = median(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         Tot_sd = sd(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         Tot_mad = mad(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         Tot_NAIO4_mean = mean(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         Tot_NAIO4_median = median(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         Tot_NAIO4_sd = sd(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         Tot_NAIO4_mad = mad(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4))) 
select(smallRNA, Tot_mean, Tot_median, Tot_sd, Tot_NAIO4_mean, Tot_NAIO4_median,
       Tot_NAIO4_sd, Tot_NAIO4_mad)
cpm_TMM_Ratio <- cpm_TMM_Ratio %>% 
  mutate(Ratio_mean = (Tot_NAIO4_mean +0.1)/ (Tot_mean+0.1),
         Ratio_median = (Tot_NAIO4_median +0.1)/ (Tot_median+0.1))
# 2.4 visualization of mean, median, etc----
#mean to median
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Ratio_mean ),
                         y = log2(Ratio_median),
                         colour= DB ))
# mean to sd
cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Tot_mean),
                         y = Tot_sd,
                         colour= DB ))
# 2.5 visualization of methylated, not methylated smallRNAs, etc----
lfc <- 2
compart <- "Total"
my_data_allRNA <- cpm_TMM_Ratio %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_" ) %>% 
  mutate(Status = ifelse(
    Tot_NAIO4_mean == 0,"not-methylated",
    ifelse(log2(Ratio_mean) > lfc,DB,
    ifelse(between(log2(Ratio_mean),-lfc,lfc),"partially-methylated",
           "not-methylated"))))  
my_data_allRNA %>% 
  write_delim(str_glue("Relative_ratio_{compart}_{date}.txt"),delim = "\t")  

my_data_allRNA %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Tot_mean + 0.1),
                         y = log2(Tot_NAIO4_mean + 0.1),
                         colour= Status ))+
labs(title =str_glue("{nrow(filter(my_data,Status == DB))} smallRNAs found methylated in {compart} after treatment, Relative ratio > {2^lfc}"),
     x="Non-Treated log2 average counts per million",
     y="Treated log2 average counts per million",
     color="smallRNAs") +
  theme_bw() +
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" ))
#1 "gold" = ensembl #2 "cornflowerblue" = miRNA 
#3 "grey" = noAnnot 
#4 "black"  = not_Enriched 
#5 "chartreuse4" = piRNA  
#6 "darkorchid1" = rfam 
#7 "firebrick1" = rRNA 
#8 "darkorange1" = tRNA 
#9 "orange" = tRNA_mature

# 2.6 visualization of methylated, not methylated piRNAs, etc----
lfc <- 2
compart <- "Total"
my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"piR|noAn")) %>% 
  mutate(Status = ifelse(Status == "noAnnot"|Status == "piRNA","methylated",Status)) %>% 
  write_tsv(str_glue("Relative_ratio_{compart}_{date}.txt"))

my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"piR")) %>% 
  mutate(Status = ifelse(Status == "piRNA","methylated",Status))
   
pirnas <- filter(my_data,Status == "methylated")
my_data %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Tot_mean + 0.1),
                         y = log2(Tot_NAIO4_mean + 0.1),
                         colour= Status ),size = 2)+
labs(title =str_glue("{nrow(pirnas)} piRNAs found methylated in {compart} after treatment, Relative ratio > {2^lfc}"),
     x="Non-Treated piRNAs (log2 average cpm)",
     y="NaIO4 oxidized piRNAs (log2 average cpm)",
     color="smallRNAs") +
  theme_bw() +
  scale_y_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  scale_x_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  coord_equal()+
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" )[c(6,4,5)])
# 2.7 visualization of methylated, not methylated not_annot, etc----
lfc <- 2
compart <- "Total"
my_data <- my_data_allRNA %>% 
  filter(str_detect(DB,"noAn")) %>% 
   mutate(Status = ifelse(Status == "noAnnot","methylated",Status)) 

pirnas <- filter(my_data,Status == "methylated")
my_data %>% 
  ggplot()+
  geom_point(mapping=aes(x = log2(Tot_mean + 0.1),
                         y = log2(Tot_NAIO4_mean + 0.1),
                         colour= Status ),size = 2)+
labs(title =str_glue("{nrow(pirnas)} not Annotated sequences found methylated in {compart} after treatment, Relative ratio > {2^lfc}"),
     x="Non-Treated not Annotated (log2 average cpm)",
     y="NaIO4 oxidized not Annotated (log2 average cpm)",
     color="smallRNAs") +
  theme_bw() +
  scale_y_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  scale_x_continuous(limits = c(-5,10),breaks = seq(-5,10,by=2.5))+
  coord_equal()+
scale_color_manual(values = c("gold", "cornflowerblue", "grey", 
                                "black","bisque4","chartreuse4",
                              "darkorchid1", "firebrick1",
                                "darkorange1",  "orange" )[c(6,4,5)])


# 3.1 cross with IPP----
# load data
my_data <- read_delim("Relative_ratio_Total_Mar_27_2019.txt", delim = "\t")

pathN <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/N_enriched_RNA.txt"
pathC <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/C_enriched_RNA.txt"

IPPN <- read_delim(pathN,delim = "\t")
IPPC <- read_delim(pathC,delim = "\t")
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% IPPC$smallRNA,str_detect(smallRNA,"piR"))
my_data %>% filter(smallRNA %in% IPPN$smallRNA,
                   smallRNA %in% DEs$smallRNA)
DEs %>% filter(smallRNA %in% IPPN$smallRNA,
                  str_detect(smallRNA,"piR"))
```



# Data wrangling of 5,3 UTR and CDS                                         
```{r}
# 1.1 load the  data 5,3UTR and CDS-------------------
name_file <- "CYT_vs_NUC"
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
namesUTRCDS <- list.files(pattern = "UTR_all|CDS_all")
names_UTR3 <- namesUTRCDS[str_detect(namesUTRCDS,"3UTR")]
names_UTR5 <- namesUTRCDS[str_detect(namesUTRCDS,"5UTR")]
names_CDS  <- namesUTRCDS[str_detect(namesUTRCDS,"CDS")]


library(tidyverse);library(data.table); library(biomaRt)
mart <- useDataset("hsapiens_gene_ensembl",useMart("ensembl"))

setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")

# 1.2a UTR3-----
UTR3 <- fread(str_glue("{names_UTR3}"),select = c(1,3)) %>% 
  as_tibble() %>% dplyr::rename("UTR3"= V3,"piRNA" = V1) %>% 
  separate(UTR3,c("UTR3","mRNA")) %>% 
  distinct(piRNA,UTR3)

G_listUTR3 <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = UTR3$UTR3,mart = mart) %>% as_tibble()

UTR3gen <- inner_join(UTR3,G_listUTR3,by= c("UTR3" = "ensembl_gene_id"))  %>% 
  distinct()

names_UTR3 <- str_replace(names_UTR3,".txt","")
UTR3gen %>% 
  write_delim(str_glue("{names_UTR3}_piRNA_genes_{date}.txt"),delim = "\t") %>% 
  dplyr::select(external_gene_name) %>% 
  distinct() %>% write_delim(str_glue("{names_UTR3}genes_{date}.txt"),delim = "\t")

# 1.2b UTR5------
UTR5 <- fread(str_glue("{names_UTR5}"),select = c(1,3)) %>% 
  as_tibble() %>% dplyr::rename("UTR5"=V3,"piRNA"=V1) %>% 
  separate(UTR5,c("UTR5","mRNA")) %>% 
  distinct(piRNA,UTR5)

G_listUTR5 <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = UTR5$UTR5,mart = mart) %>% as_tibble()

UTR5gen <- inner_join(UTR5,G_listUTR5,by= c("UTR5" = "ensembl_gene_id"))  %>% 
  distinct()

names_UTR5 <- str_replace(names_UTR5,".txt","")

UTR5gen %>% 
  write_delim(str_glue("{names_UTR5}_piRNA_genes_{date}.txt"),delim = "\t") %>% 
  dplyr::select(external_gene_name) %>% 
  distinct() %>% 
  write_delim(str_glue("{names_UTR5}genes_{date}.txt"),delim = "\t")

# 1.2c CDS-------
CDS <- fread(str_glue("{names_CDS}"),select = c(1,3)) %>% 
  as_tibble() %>% dplyr::rename("CDS"=V3,"piRNA"=V1) %>% 
  separate(CDS,c("CDS","mRNA")) %>% 
  distinct(piRNA,CDS)

G_listCDS <- getBM(filters = "ensembl_gene_id",
                attributes = c("ensembl_gene_id","external_gene_name"),
                values = CDS$CDS,mart = mart) %>% as_tibble()

CDSgen <- inner_join(CDS,G_listCDS,by= c("CDS" = "ensembl_gene_id"))  %>% 
  distinct()

names_CDS <- str_replace(names_CDS,".txt","")

CDSgen %>% 
  write_delim(str_glue("{names_CDS}_piRNA_genes_{date}.txt"),delim = "\t") %>% 
  dplyr::select(external_gene_name) %>% 
  distinct() %>% write_delim(str_glue("{names_CDS}genes_{date}.txt"),delim = "\t")

# 1.3 collapse genes from 3 tables-------
UTR3gen <- UTR3gen %>% dplyr::rename("EnsID"=UTR3)
UTR5gen <- UTR5gen %>% dplyr::rename("EnsID"=UTR5)
CDSgen  <- CDSgen  %>% dplyr::rename("EnsID"=CDS)

genelist <- list(UTR3gen,UTR5gen,CDSgen)
names(genelist) <- c("UTR3","UTR5","CDS")

fname <- str_replace(names_CDS,"CDS_","")
genReg <- rbindlist(genelist,use.names = TRUE,idcol = "region",fill = TRUE) %>%
  spread(region,external_gene_name,fill = "NA") %>% 
  write_delim(str_glue("{fname}_genes_3C5__{date}.txt"),delim = "\t")

# join(t)s ;) -----
inner_join(UTR3gen,CDSgen, by = "external_gene_name")
inner_join(UTR5gen,CDSgen, by = "external_gene_name")
inner_join(UTR3gen,UTR5gen, by = "external_gene_name")

inner_join(UTR3gen,CDSgen, by = "external_gene_name") %>% 
  inner_join(UTR5gen, by = "external_gene_name") %>% 
  write_delim(str_glue("common_{fname[2]}_genes_{date}.txt"),delim = "\t")
```


# find genomic coordinates                                                  
```{r}
name <- "NoAnnot"
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(GenomicRanges);library(GenomicAlignments);library(data.table)
library(tidyverse);library(rtracklayer);library(ggseqlogo)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
path <- "~/Documents/SPORTS_results/Result/fasta_files/IPP_PIWIL/"
setwd(dir = "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")

files <- list.files(path, pattern = ".bam$")
files <- str_glue("{path}{files}")
bflist <- sapply(files,BamFile,simplify = FALSE)
#sapply(files,indexBam,simplify = FALSE)
slist <- lapply(bflist, seqlengths)
slist1 <- sapply(bflist, seqlengths,simplify = FALSE)
gal <- GAlignmentsList(lapply(files,function(x)readGAlignments(x,use.names = TRUE)))
names(gal) <- str_remove(basename(as.character(files)),"_sorted.bam") %>% 
  str_remove("c_") 
#write_rds(gal,"GAlignmentsList_IPP.rds")
#gal <- read_rds("GAlignmentsList_IPP.rds")
#seqinfo(bflist[[1]])
#readGAlignments(str_glue("{files[1]}"),use.names = TRUE) -> test

# load the IPP enriched
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
enrichedFiles <- list.files(path,pattern = "enriched_RNA.txt")

enrichedC <- read_delim(str_glue("{path}{enrichedFiles[1]}"),delim = "\t") %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_" ) %>% 
  dplyr::select(sRNA,topIPP_C_vs_INPUT)

enrichedN <- read_delim(str_glue("{path}{enrichedFiles[2]}"),delim = "\t") %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_" ) %>% 
  dplyr::select(sRNA,topIPP_N_vs_INPUT)

enriched <- inner_join(enrichedC,enrichedN, by = "sRNA")

# load the treated
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/"

Nuc_Meth_Ipp <- read_delim(str_glue("{path}Methylated_Nuc_IPP_CYT_vs_NUC_FDR001.txt"),delim = "\t") %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_" )

Cyt_Meth_Ipp <- read_delim(str_glue("{path}Methylated_Cyt_IPP_CYT_vs_NUC_FDR001.txt"),delim = "\t") %>% 
  separate(smallRNA,c("DB","sRNA"),sep = "_match_" )

# go to another chunk and upload your dataset
pathRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path = pathRNAs ,pattern = "_smallRNA", recursive = TRUE)

path_genome <- "/home/rstudio/Documents/SPORTS_results/Result/"
list_genomeRNA <- list.files(path = path_genome ,pattern = "_output_match", recursive = TRUE)

# find the file about the experiments of treatement
listRNA <- list_smallRNA[str_detect(list_smallRNA,"COLO205_IPP")]
list_genRNA <- list_genomeRNA [str_detect(list_genomeRNA,"COLO205_IPP")]

date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))

i <- 1
l_NUC_PI <- vector("list",12)
l_NUC_NA <- vector("list",12)
l_CYT_PI <- vector("list",12)
l_CYT_NA <- vector("list",12)

for (x in names(gal)) {
  print(x)
  sample <- fread(str_glue("{pathRNAs}{listRNA[str_detect(listRNA,x)]}")) %>% 
  as_tibble() %>% 
  arrange(V1)  %>% 
  mutate(width = str_length(V6))
  
  sample_out <- fread(str_glue("{path_genome}{list_genRNA[str_detect(list_genRNA,x)]}")) %>% 
  as_tibble() %>% 
  arrange(V1)  %>% 
  dplyr::select(V1:V6) %>% 
  mutate(width = str_length(V6))

t1 <- sample %>% dplyr::select(V1) %>% distinct() %>% nrow
t2 <- sample_out %>% dplyr::select(V1) %>% distinct() %>% nrow
print(identical(t1,t2))

if(identical(distinct(sample,V1,.keep_all= TRUE)$width,sample_out$width)== TRUE){
  
  #subset results files
  sample_NUC <- sample %>% filter(V4 %in% Nuc_Meth_Ipp$sRNA)
  sample_out_NUC <- sample_out %>% filter(V1 %in% sample_NUC$V1)
  txt_NUC <- inner_join(sample_NUC,sample_out_NUC, by = c("V1","width","V2")) %>% 
    dplyr::select(RNA_type,counts = "V2",smallRNA = "V4.x",strand = "V3",
            chr = "V4.y", start = "V5",
            seq1 = "V6.x", reverse_Complement = "V6.y",width)

  sample_CYT <- sample %>% filter(V4 %in% Cyt_Meth_Ipp$sRNA)
  sample_out_CYT <- sample_out %>% filter(V1 %in% sample_CYT$V1)
  txt_CYT <- inner_join(sample_CYT,sample_out_CYT, by = c("V1","width","V2")) %>% 
    dplyr::select(RNA_type,counts = "V2",smallRNA = "V4.x",strand = "V3",
            chr = "V4.y", start = "V5",
            seq1 = "V6.x", reverse_Complement = "V6.y",width)
  
  txt_NUC_PI <- txt_NUC %>% filter(RNA_type == "piRNA_match") %>% distinct(seq1)
  txt_NUC_NA <- txt_NUC %>% filter(RNA_type == "noAnnot_match") %>% distinct(seq1)
  
  txt_CYT_PI <- txt_CYT %>% filter(RNA_type == "piRNA_match") %>% distinct(seq1)
  txt_CYT_NA <- txt_CYT %>% filter(RNA_type == "noAnnot_match") %>% distinct(seq1)
  
  l_NUC_PI[[x]] <- str_sub(txt_NUC_PI$seq1,1,15)
  l_NUC_NA[[x]] <- str_sub(txt_NUC_NA$seq1,1,15)
  
  l_CYT_PI[[x]] <- str_sub(txt_CYT_PI$seq1,1,15)
  l_CYT_NA[[x]] <- str_sub(txt_CYT_NA$seq1,1,15)
  
  #subset GRanges
  GR_sample   <- gal[[x]]
  names(GR_sample) <- str_remove(names(GR_sample),"_.+")

  GR_sample_NUC <- GR_sample[names(GR_sample) %in% sample_out_NUC$V1]
  GR_sample_CYT <- GR_sample[names(GR_sample) %in% sample_out_CYT$V1]

  n1_NUC <- names(GR_sample_NUC)
  n1_CYT <- names(GR_sample_CYT)
  n2_NUC <- sapply(as.list(n1_NUC), function(y)filter(sample_NUC,V1==y)$V4[1])
  n2_CYT <- sapply(as.list(n1_CYT), function(y)filter(sample_CYT,V1==y)$V4[1])
  n2_NUC <- str_remove(n2_NUC,"\\s.+")
  n2_CYT <- str_remove(n2_CYT,"\\s.+")
  
  # change the names in Granges
  names(GR_sample_NUC) <- n2_NUC
  names(GR_sample_CYT) <- n2_CYT
  
  # save bed files and their txts
  last_pat <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/bedfiles/"
  export(GR_sample_NUC,con = str_glue("{last_pat}{x}_NUC_{date}.bed"),
         format = "BED")
  write_delim(txt_NUC,str_glue("{last_pat}{x}_NUC_{date}.txt"),
              delim = "\t")
  export(GR_sample_CYT,con = str_glue("{last_pat}{x}_CYT_{date}.bed"),
         format = "BED")
  write_delim(txt_CYT,str_glue("{last_pat}{x}_CYT_{date}.txt"),
              delim = "\t")
}

else {break()}
  
}

require(ggseqlogo)
data(ggseqlogo_sample)
seqs_dna


ggseqlogo(l_CYT_PI[13:24],method = "prob")
ggseqlogo(l_NUC_PI[13:24],method = "prob")

# write_rds(l_NUC_PI[13:24],str_glue("{last_pat}l_NUC_PI.rds"))
# write_rds(l_NUC_NA[13:24],str_glue("{last_pat}l_NUC_NA.rds")) 
# write_rds(l_CYT_PI[13:24],str_glue("{last_pat}l_CYT_PI.rds")) 
# write_rds(l_CYT_NA[13:24],str_glue("{last_pat}l_CYT_NA.rds")) 

#l_CYT_PI <- read_rds(str_glue("{last_pat}l_CYT_PI.rds"))
ggseqlogo(l_CYT_PI, method = "prob")

```


# Cancer snp piRNAs                                                         
```{r}
library(tidyverse)
piRNACRC <- tribble (~piRNA, ~Sequence,
	"piR_000335", 	"TTTCCATTGGACCGTTCGTGTTCTGTGGT", 	
 	'piR_005132',	"CCCCCCGCAGGTCTACATCAACTATTACGA", 	
 	"piR_015481", "CCGCAGGGTCTACATCAACTTTACGACA", 	
	"piR_021520", 	"CCCCCGCAGGGTCTACATCAACTATTA", 	
 	"piR_015551", 	"GCTCAGGGACGGTGTGTAGCCTTGAAGACA", 
 	"piR_020980", 	"TAAAGACAGACTTGCAGTTCAGCCACAGGA", 
 	"piR_001549", 	"GGTTATGCTGCATTAGCTATTAGGCTGATA", 
 	"piR_005938", 	"GCGGTCTTTGCTCTCGGCTGTGTCACAGA", 	
 	'piR_002587', 	"TTCGTGTTGATGTTGCCGATGATGCTGA") %>% 
  mutate(width = str_length(Sequence))
piRbase <- tribble (~piRNA, ~Sequence,
	"piR_hsa-679", 	"ACCACAGAACACGAACGGTCCCAATGGAAA", 	
 	'piR-hsa-7400',	"TCGTAATAGTTGATGTAGACCCTGCGGGGGG", 	
 	"piR-hsa-21417", "TGTCGTAATAGTTGATGTAGACCCTGCGG", 	
	"piR-hsa-29786", 	"TAATAGTTGATGTAGACCCTGCGGGGG", 	
 	"piR-hsa-21517", 	"TGTCTTCAAGGCTACACACCGTCCCTGAGC", 
 	"piR-hsa-29056", 	"TAAAGACAGACTTGCAGTTCAGCCACAGGA", 
 	"piR-hsa-2363", 	"TATCAGCCTAATAGCTAATGTCAGCATAACC", 
 	"piR-hsa-8401", 	"TCTGTGACACAGCCGAGAGCAAAGACCGC", 	
 	'piR-hsa-3789', 	"TCAGCATCATCGGCAACATCAACACGAA") %>% 
  mutate(width = str_length(Sequence))


seq <- paste(fasta)
smallRNA <- names(fasta)
tb <- tibble(seq,smallRNA)
tb %>% filter(seq %in% piRNACRC$Sequence)# piR-hsa-29056
tb %>% filter(seq %in% piRbase$Sequence)

```


# Create Genomic Ranges of sequences to map the reads                       
```{r}
library(tidyverse);library(data.table);library(rtracklayer)
library(Rsamtools);library(GenomicRanges);library(ggbio)

# import bed file of piRBase
gr_obj <- import("~/Documents/hsa.bed")
genome(gr_obj) <- "hg38"

# gr_list <- split(gr_obj,gr_obj$name)
fasta <- readDNAStringSet("/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa")
mirBAse <- readDNAStringSet("/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/miRBase_22.1/miRBase_22-hsa.fa")

# 1.1 import piRNAs, raw reads of TOTAL etc-----
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment")
path_ <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/Methylated_IPP_tables_FDR001_filt2.txt"

piRNAfil <- read_delim(path_,delim = "\t") %>% 
  filter(str_detect(smallRNA,"piRN")) %>% 
  separate(smallRNA,c("match","piRNA"),sep = "_match_") %>% 
  select(piRNA,topIPP_N_vs_INPUT,topIPP_C_vs_INPUT)

# go to another chunk and upload your dataset
pathRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path = pathRNAs ,pattern = "_smallRNA", recursive = TRUE)

path_genome <- "/home/rstudio/Documents/SPORTS_results/Result/"
list_genomeRNA <- list.files(path = path_genome ,pattern = "_output_match", recursive = TRUE)

# find the file about the experiments of treatement
listRNA <- list_smallRNA[str_detect(list_smallRNA,"TOTAL")]
list_genRNA <- list_genomeRNA [str_detect(list_genomeRNA,"TOTAL")]

# load data to data.table
setwd(pathRNAs)
DTRNA <- rbindlist( sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

setwd(path_genome)
DTGEN <- rbindlist( sapply(list_genRNA,fread,select = c(1,2,3,4,5,6),simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

# filter the datasets
DT <- DTRNA %>% as_tibble() %>% filter(V4 %in% piRNAfil$piRNA) %>% 
  mutate(length=str_count(V6)) %>% 
  select(FileName,reads=V1,score=V2,seqnames=V4,Seq=V6,length)

DTGENfil <-DTGEN %>% as_tibble() %>% filter(V6 %in% pir1_TOTAL_1_NaIO4$Seq,str_detect(FileName,title))

#DTGENfil1 <- DTGENfil %>% add_column(piRNA= filter(DT,Seq==DTGENfil$V6)$seqnames)

# subset the granges object
gr_obj <- gr_obj[gr_obj$name %in% piRNAfil$piRNA]
names(gr_obj) <- gr_obj$name
fasta <- fasta[piRNAfil$piRNA]
gr_obj$seq <- fasta[names(gr_obj)]


# 1.2 Search piRNAs-------
i <- "piR-hsa-2011747"
i <- piRNAfil$piRNA[30]


title1 <- "TOTAL_1_NaIO4"

for (i in seq(along=piRNAfil$piRNA)){
  namT <- piRNAfil$piRNA[i]
piRNA_GR <- DT %>% filter(str_detect(FileName,title1),
                                     seqnames == namT) %>% 
  select(seqnames,score,reads,seqnames,Seq,length) %>% 
add_row(seqnames = namT, 
        score = sum(.$score),
        reads = namT,
        Seq = toString(fasta[namT]),
        length = width(fasta[namT])) %>% 
  add_column(strand = as.factor(gr_obj[namT]@strand)) %>% 
  mutate( 
       start = as_tibble(str_locate(string = toString(fasta[namT]),
                                    pattern =  .$Seq))$start,
       end = as_tibble(str_locate(string = toString(fasta[namT]),
                                                  pattern =  .$Seq))$end) %>% 
  arrange(score)
GR <- makeGRangesFromDataFrame(piRNA_GR,keep.extra.columns = TRUE)
ggplot()+
  ggbio::geom_alignment(GR,aes(color = reads,fill = reads %in% namT ),binwidth=1,
                        main=paste0(title1,", ",
                                    as.factor(runValue(gr_obj[namT]@strand))," strand"))+
  theme_minimal()+
  scale_x_continuous(breaks = seq(0,max(GR$length+1),by=1))+
  labs(fill = str_glue("{namT}"))+
  scale_fill_manual(labels=c("read","complete sequence"),values = c("black","orange"))


ggsave(filename = paste0(sub("-","_",namT),"_",title1,"_reads.pdf"))
}

paste0(sub("-","_",i),"_reads") 


```
# Let's make some Heatmaps                                                  
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))

library(data.table);library(tidyverse)
library(ComplexHeatmap);library(edgeR);library(circlize)
# 1.1 Cyt_Nuc Dataset with IPP subset------
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/")

path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/"
x1_voomQ <- read_rds(str_glue("{path}x1_voomQ_filt2_CYT_vs_NUC.rds"))

path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
x1_voomTMM <- read_rds(str_glue("{path}x1_voomTMM.rds"))
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")

# "create" the matrix
mat <- x1_voomQ$E
names(mat) <- rownames(mat)
sub_C_N_enriched <- C_N_enriched %>% 
  filter(smallRNA %in% names(mat)) %>% 
  filter(str_detect(smallRNA,"piR"))

# subset the matrix
mat <- subset(mat,rownames(mat) %in% sub_C_N_enriched$smallRNA)
mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow

#prepare the matrix
scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat))  
df$cell_line <- df$cell_line %>% 
  str_replace("CITOSOL","CYTOSOL")  %>% 
  str_replace("NUCLEO","NUCLEUS")

ha <- HeatmapAnnotation(df,col = list( cell_line = c("CYTOSOL_4C_NaIO4" = "goldenrod",
                                          "CYTOSOL_5C_NaIO4" = "goldenrod",
                                          "CYTOSOL_6C_NaIO4" = "goldenrod",
                                          "CYTOSOL_4C" = "brown2",
                                          "CYTOSOL_5C" = "brown2",
                                          "CYTOSOL_6C" = "brown2",
                                          "NUCLEUS_4N" = "royalblue1", 
                                          "NUCLEUS_5N" = "royalblue1", 
                                          "NUCLEUS_6N" = "royalblue1",
                                          "NUCLEUS_4N_NaIO4" = "sienna2", 
                                          "NUCLEUS_5N_NaIO4" = "sienna2", 
                                          "NUCLEUS_6N_NaIO4" = "sienna2")))
f <- colorRamp2(c(-1,median(scaled_mat),1),c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(scaled_mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",clustering_distance_columns = "spearman",
        clustering_method_columns ="ward.D2",
        show_row_names = FALSE,name= "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} smallRNAs"),
        column_title_side="top",
        km = 2)

# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
        column_order = c(2,4,6,1,3,5,8,10,12,7,9,11),
        heatmap_legend_param = cm,
        show_row_names = FALSE,
        name = "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} smallRNAs"),
        column_title_side = "top",
        km = 2)

# 1.2 Heatmap of median cpm  between Cyt Nuc IPP, treated and non treated -------
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")

# load cpm tables
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/"
list.files(path,pattern = "lcpm")

CYT_NUC_lcpm <- read_delim("lcpm_TMM_CYT_vs_NUC.txt",delim = "\t") %>% 
  group_by(smallRNA) %>% 
  mutate(Cytosol_Treated = median(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,
                                    CITOSOL_6C_NaIO4)),
         Cytosol = median(c(CITOSOL_4C,CITOSOL_5C,
                                    CITOSOL_6C)),
         Nucleus_Treated = median(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,
                                    NUCLEO_6N_NaIO4)),
         Nucleus = median(c(NUCLEO_4N,NUCLEO_5N,
                                    NUCLEO_6N))) %>% 
  select(smallRNA,Cytosol_Treated,Cytosol,Nucleus_Treated,Nucleus) %>% 
  ungroup()

path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")
list.files(path,"lcpm")

IPP_lcpm <- read_delim(str_glue("{path}IPP_PIWIL_TMM_logcpm_counts.txt"),delim = "\t") %>% 
  group_by(smallRNA) %>% 
  mutate(Input = median(c(COLO205_IPP_INPUT_1,COLO205_IPP_INPUT_2,COLO205_IPP_INPUT_3)),
         N_ab = median(c(COLO205_IPP_N_Ab_1,COLO205_IPP_N_Ab_2,COLO205_IPP_N_Ab_3)),
         C_ab = median(c(COLO205_IPP_C_Ab_1,COLO205_IPP_C_Ab_2,COLO205_IPP_C_Ab_3))) %>%
  select(smallRNA, Input, N_ab, C_ab) %>% ungroup()

# subset the datasets
C_N_enriched <- C_N_enriched %>% 
  filter(str_detect(smallRNA,"piRN"))

CYT_NUC_lcpm <- CYT_NUC_lcpm %>% 
  filter(smallRNA %in% C_N_enriched$smallRNA)
IPP_lcpm <- IPP_lcpm %>% 
    filter(smallRNA %in% CYT_NUC_lcpm$smallRNA)

#  create the matrix
mat1 <- inner_join(CYT_NUC_lcpm,IPP_lcpm) %>% 
  column_to_rownames("smallRNA") %>% 
  as.matrix()
scaled_mat <- t(scale(t(as.matrix(mat1))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat)) 

f <- colorRamp2(c(-2,median(scaled_mat),2),c("blue","white","yellow"))

ha <- HeatmapAnnotation(df = df)
# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
        #column_order = c(1,2,6,5,7,3,4),
        heatmap_legend_param = cm,
        show_row_names = FALSE,
        name = "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(IPP_lcpm)} smallRNAs"),
        column_title_side = "top",
        km = 2)

#  create the matrix treat
mat1 <- inner_join(CYT_NUC_lcpm,IPP_lcpm) %>% 
  select(smallRNA,Cytosol_Treated,Input,Nucleus_Treated) %>% 
  column_to_rownames("smallRNA")
scaled_mat <- t(scale(t(as.matrix(mat1))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat)) 

f <- colorRamp2(c(-0.5,median(scaled_mat),0.5),c("blue","white","yellow"))

ha <- HeatmapAnnotation(df)
# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
        #column_order = c(1,2,5,3,4),
        heatmap_legend_param = cm,
        show_row_names = FALSE,
        name = "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(IPP_lcpm)} smallRNAs"),
        column_title_side = "top",
        km = 4)

#  create the matrix NONtreat
mat1 <- inner_join(CYT_NUC_lcpm,IPP_lcpm) %>% 
  select(smallRNA,Cytosol,Input,Nucleus) %>% 
  column_to_rownames("smallRNA")
scaled_mat <- t(scale(t(as.matrix(mat1))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat)) 

f <- colorRamp2(c(-0.5,median(scaled_mat),0.5),c("blue","white","yellow"))

ha <- HeatmapAnnotation(df)
# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
        #column_order = c(1,2,5,3,4),
        heatmap_legend_param = cm,
        show_row_names = FALSE,
        name = "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(IPP_lcpm)} smallRNAs"),
        column_title_side = "top",
        km = 2)

# put all together
mat1 <- inner_join(CYT_NUC_lcpm,IPP_lcpm) %>% 
  select(smallRNA,Cytosol,N_ab,Input,C_ab,Nucleus) %>% 
  column_to_rownames("smallRNA") %>% 
  as.matrix()
scaled_mat <- t(scale(t(as.matrix(mat1))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat)) 

f <- colorRamp2(c(-1.5,median(scaled_mat),1.5),c("blue","white","yellow"))

ha <- HeatmapAnnotation(df)
# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
       # column_order = c(1,2,6,5,7,3,4),
        heatmap_legend_param = cm,
        show_row_names = FALSE,
        name = "log2 cpm", 
        column_title = str_glue("Heatmap of {nrow(IPP_lcpm)} piRNAs"),
        column_title_side = "top",
        km = 2)



# 1.3 Heatmap of piR in IPP inpu C, N ----
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/")
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
x1_voomTMM <- read_rds(str_glue("{path}x1_voomTMM.rds"))
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")
sub_C_N_enriched <- C_N_enriched %>% 
  filter(str_detect(smallRNA,"piR"))
# "create" the matrix and subset
mat <- subset.matrix(x1_voomTMM$E, 
                     select = -c(COLO205_IPP_Control_noAb_1,
                        COLO205_IPP_Control_noAb_2,
                        COLO205_IPP_Control_noAb_3)) 
mat <- subset(mat,rownames(mat) %in% sub_C_N_enriched$smallRNA)
mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow
# create a scaled mat
scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat))  
df$cell_line <- df$cell_line %>% 
  str_remove("COLO205_IPP_")  

ha <- HeatmapAnnotation(df,
                        col = list( cell_line = c("INPUT_1" = "goldenrod",
                                                  "INPUT_2" = "goldenrod", 
                                                  "INPUT_3" = "goldenrod",
                                                  "C_Ab_1" = "royalblue1", 
                                                  "C_Ab_2" = "royalblue1",
                                                  "C_Ab_3" = "royalblue1",
                                          "N_Ab_1" = "sienna2", 
                                          "N_Ab_2" = "sienna2", 
                                          "N_Ab_3" = "sienna2")))
f <- colorRamp2(c(-1,median(scaled_mat),1),c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(scaled_mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        #clustering_distance_columns = "spearman",
        #clustering_method_columns ="ward.D2",
        column_order = c(4,5,6,1,2,3,7,8,9),
        cluster_columns = FALSE,
        show_row_names = FALSE,name= "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} piRNAs in COLO205"),
        column_title_side="top",
        km = 1)

# find the enriched
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")
sub_C_N_enriched <- C_N_enriched %>% 
  filter(str_detect(smallRNA,"piR"))
#  mat with cpm
mat <- read_delim("IPP_PIWIL_TMM_cpm_counts.txt",delim = "\t") %>%
  select(-contains("_Control_")) %>% 
  filter(smallRNA  %in% sub_C_N_enriched$smallRNA) %>% 
  column_to_rownames("smallRNA") %>%  as.matrix()

mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow
# create a scaled mat
hist(mat)
df <- data.frame(cell_line = colnames(mat))  
df$cell_line <- df$cell_line %>% 
  str_remove("COLO205_IPP_")  

ha <- HeatmapAnnotation(df,
                        col = list( cell_line = c("INPUT_1" = "goldenrod",
                                                  "INPUT_2" = "goldenrod", 
                                                  "INPUT_3" = "goldenrod",
                                                  "C_Ab_1" = "royalblue1", 
                                                  "C_Ab_2" = "royalblue1",
                                                  "C_Ab_3" = "royalblue1",
                                          "N_Ab_1" = "sienna2", 
                                          "N_Ab_2" = "sienna2", 
                                          "N_Ab_3" = "sienna2")))
f <- colorRamp2(c(min(mat),mean(mat),10),c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        #clustering_distance_columns = "spearman",
        #clustering_method_columns ="ward.D2",
        column_order = c(4,5,6,1,2,3,7,8,9),
        cluster_columns = FALSE,
        show_row_names = FALSE,name= "cpm", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} piRNAs in COLO205"),
        column_title_side="top",
        km = 1)

# 1.4 Heatmap of piR in IPP inpu C, N median values----
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/")
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")
sub_C_N_enriched <- C_N_enriched %>% 
  filter(str_detect(smallRNA,"piR"))
# load lcpms and calculate median
list.files(path,".txt")
mat <- read_delim(str_glue("{path}IPP_PIWIL_TMM_logcpm_counts.txt" ),delim = "\t") %>% 
  select(-contains("_Control_")) %>% 
  filter(smallRNA  %in% sub_C_N_enriched$smallRNA) %>% 
  group_by(smallRNA) %>% 
  mutate(Input = median(c(COLO205_IPP_INPUT_1,COLO205_IPP_INPUT_2,COLO205_IPP_INPUT_3)),
         N_ab = median(c(COLO205_IPP_N_Ab_1,COLO205_IPP_N_Ab_2,COLO205_IPP_N_Ab_3)),
         C_ab = median(c(COLO205_IPP_C_Ab_1,COLO205_IPP_C_Ab_2,COLO205_IPP_C_Ab_3))) %>%
  select(smallRNA, Input,C_ab, N_ab ) %>% 
  arrange(desc(Input)) %>% 
  ungroup() %>% column_to_rownames("smallRNA") %>% as.matrix()
mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow
# create a scaled mat
scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat))  

ha <- HeatmapAnnotation(df,
                        col = list( cell_line = c("Input" = "goldenrod",
                                               #   "INPUT_2" = "goldenrod", 
                                                #  "INPUT_3" = "goldenrod",
                                                  "C_ab" = "royalblue1", 
                                                #  "C_Ab_2" = "royalblue1",
                                                #  "C_Ab_3" = "royalblue1",
                                        #  "N_ab" = "sienna2", 
                                        #  "N_Ab_2" = "sienna2", 
                                          "N_ab" = "sienna2")))
f <- colorRamp2(c(min(scaled_mat),mean(scaled_mat),max(scaled_mat)),
                c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(scaled_mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        #clustering_distance_columns = "spearman",
        #clustering_method_columns ="ward.D2",
        #column_order = c(4,5,6,1,2,3,7,8,9),
        cluster_columns = FALSE,
        show_row_names = FALSE,name= "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} piRNAs in COLO205"),
        column_title_side="top",
        row_title = "median values",
        row_title_side = c("left"),
        km = 1)

# load cpms and calculate median

list.files(path,".txt")
mat <- read_delim(str_glue("{path}IPP_PIWIL_TMM_cpm_counts.txt" ),delim = "\t") %>% 
  select(-contains("_Control_")) %>% 
  filter(smallRNA  %in% sub_C_N_enriched$smallRNA) %>% 
  group_by(smallRNA) %>% 
  mutate(Input = median(c(COLO205_IPP_INPUT_1,COLO205_IPP_INPUT_2,COLO205_IPP_INPUT_3)),
         N_ab = median(c(COLO205_IPP_N_Ab_1,COLO205_IPP_N_Ab_2,COLO205_IPP_N_Ab_3)),
         C_ab = median(c(COLO205_IPP_C_Ab_1,COLO205_IPP_C_Ab_2,COLO205_IPP_C_Ab_3))) %>%
  select(smallRNA, Input,C_ab, N_ab ) %>% 
  arrange(desc(Input)) %>% 
  ungroup() %>% column_to_rownames("smallRNA") %>% as.matrix()
mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow
# create a scaled mat

#scaled_mat <- t(scale(t(as.matrix(mat))))
hist(mat)
df <- data.frame(samples = colnames(mat))  

ha <- HeatmapAnnotation(df,
                        col = list( samples = c("Input" = "goldenrod",
                                               #   "INPUT_2" = "goldenrod", 
                                                #  "INPUT_3" = "goldenrod",
                                                  "C_ab" = "royalblue1", 
                                                #  "C_Ab_2" = "royalblue1",
                                                #  "C_Ab_3" = "royalblue1",
                                        #  "N_ab" = "sienna2", 
                                        #  "N_Ab_2" = "sienna2", 
                                          "N_ab" = "sienna2")))
f <- colorRamp2(c(min(mat),median(mat),2),
                c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,
        #clustering_distance_rows = "spearman",
        #clustering_method_rows="ward.D2",
        cluster_rows = FALSE,
        #clustering_distance_columns = "spearman",
        #clustering_method_columns ="ward.D2",
        #column_order = c(4,5,6,1,2,3,7,8,9),
        cluster_columns = FALSE,
        show_row_names = FALSE,name= "counts \nper\nmillion", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} piRNAs
                                found enriched in PIWIL1, COLO205"),
        column_title_side="top",
        row_title = "median values",
        row_title_side = c("left"),
        km = 1)

#with noAnnot 
C_N_enriched <- read_delim(str_glue("{path}LFCs_C_N_2_input_2_results.txt"),delim = "\t")
sub_C_N_enriched <- C_N_enriched %>% 
  filter(str_detect(smallRNA,"noAnnot"))

list.files(path,".txt")
mat <- read_delim(str_glue("{path}IPP_PIWIL_TMM_cpm_counts.txt" ),delim = "\t") %>% 
  select(-contains("_Control_")) %>% 
  filter(smallRNA  %in% sub_C_N_enriched$smallRNA) %>% 
  group_by(smallRNA) %>% 
  mutate(Input = median(c(COLO205_IPP_INPUT_1,COLO205_IPP_INPUT_2,COLO205_IPP_INPUT_3)),
         N_ab = median(c(COLO205_IPP_N_Ab_1,COLO205_IPP_N_Ab_2,COLO205_IPP_N_Ab_3)),
         C_ab = median(c(COLO205_IPP_C_Ab_1,COLO205_IPP_C_Ab_2,COLO205_IPP_C_Ab_3))) %>%
  select(smallRNA, Input,C_ab, N_ab ) %>% 
  arrange(desc(Input)) %>% 
  ungroup() %>% column_to_rownames("smallRNA") %>% as.matrix()
mat %>% as_tibble() %>% nrow
sub_C_N_enriched%>% nrow
# create a scaled mat

#scaled_mat <- t(scale(t(as.matrix(mat))))
hist(mat)
df <- data.frame(samples = colnames(mat))  

ha <- HeatmapAnnotation(df,
                        col = list( samples = c("Input" = "goldenrod",
                                               #   "INPUT_2" = "goldenrod", 
                                                #  "INPUT_3" = "goldenrod",
                                                  "C_ab" = "royalblue1", 
                                                #  "C_Ab_2" = "royalblue1",
                                                #  "C_Ab_3" = "royalblue1",
                                        #  "N_ab" = "sienna2", 
                                        #  "N_Ab_2" = "sienna2", 
                                          "N_ab" = "sienna2")))
f <- colorRamp2(c(min(mat),median(mat),3),
                c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,
        #clustering_distance_rows = "spearman",
        #clustering_method_rows="ward.D2",
        cluster_rows = FALSE,
        #clustering_distance_columns = "spearman",
        #clustering_method_columns ="ward.D2",
        #column_order = c(4,5,6,1,2,3,7,8,9),
        cluster_columns = FALSE,
        show_row_names = FALSE,name= "counts \nper\nmillion", 
        column_title = str_glue("Heatmap of {nrow(sub_C_N_enriched)} not annotated reads
                                found enriched in PIWIL1, COLO205"),
        column_title_side="top",
        row_title = "median values",
        row_title_side = c("left"),
        km = 1)
```

# 1st Dataset, PCA and hypergeometric test Cell lines, one Biol rep         
```{r}
library(data.table);library(tidyverse);library(rafalib)
path <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/PCA_hypergeo/")

data_filt <- "HT115|HT29|RKO|SW1417|CACO2|SW403|HCT116|COLO205"

#  1.1 list files merged and outputs-------
list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

#list_output<- list.files(path,pattern = "_output", recursive = TRUE)
# find the file about the experiments Testis
# list_smallRNA[str_detect(list_smallRNA,"Testis")] -> list_smallRNA_Testis
# list_smallRNA_Testis[str_detect(list_smallRNA_Testis,"Non")]->list_smallRNA_Testis
# list_output[str_detect(list_output,"Testis")] -> list_output_Testis
# list_output_Testis[str_detect(list_output_Testis,"Non")]->list_output_Testis

# find the file about the experiments Cell lines
list_smallRNA[str_detect(list_smallRNA,"1_Cell")] -> list_smallRNA_Cell
#list_output[str_detect(list_output,"1_Cell")] -> list_output_Cell

list_smallRNA_Cell <- subset(list_smallRNA_Cell,str_detect(list_smallRNA_Cell,data_filt))
list_smallRNA_Cell <- str_glue("{path}{list_smallRNA_Cell}")
# combine lists
# listRNA <- c(list_smallRNA_Cell,list_smallRNA_Testis)
# listout <- c(list_output_Testis,list_output_Cell)

#load data to data.table
DTRNA <- rbindlist(sapply(list_smallRNA_Cell,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

DTRNA <- DTRNA %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")
DTRNA <- DTRNA %>% 
  filter(str_detect(smallRNA,"piRNA")) %>% 
  mutate(length = str_length(V6)) %>% 
  write_rds("DTRNA.rds")

#  1.2 load data-------
tbRNA <- read_rds("tbRNA.rds")
#DTRNA <- read_rds("DTRNA.rds")
DT <- read_delim("Cell_lines_norm_TMMwzp_cpm_counts.txt","\t") %>%
  filter(str_detect(smallRNA,"piR")) %>% 
  separate(smallRNA,c("DB","sRNA"),"_match_")

# subset complete seq files with the normlized
tbRNA <- DTRNA %>% filter(smallRNA %in% DT$smallRNA) %>% 
  write_rds("tbRNA.rds")

tbRNA <- tbRNA %>% 
  mutate(FileName = str_remove(basename(FileName),"_smallRNA.txt")) %>% 
  select(FileName,smallRNA, V1, V2, V6) %>% 
  distinct(V6, .keep_all = TRUE)

#  2.1 find mature sequences------
# load mature seq
library(Rsamtools)
fasta <- readDNAStringSet("~/Documents/Databases/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa")
mseq  <- fasta[DT$sRNA]
# make it a tibble
matureTB <- tibble( matureSEQ = paste(mseq),
                    smallRNA = names(mseq),
                    mature_len=width(mseq))

# add a col with the mature seq
tbRNA <- tbRNA %>% 
  separate(smallRNA,c("DB","sRNA"),"_match_") %>% 
  inner_join(matureTB, by = c("sRNA"="smallRNA")) %>% 
  write_rds("tbRNAMature.rds")

# raw reads seq
tbRaw <- tbRNA %>% 
  mutate(FileName = str_remove(basename(FileName),"_smallRNA.txt"),
         length = str_length(V6)) %>% 
  select(FileName, V1, V2, V6, length) %>% 
  distinct(V6, .keep_all = TRUE)
# mature
tbMature <- tbRNA %>% 
  mutate(FileName = str_remove(basename(FileName),"_smallRNA.txt"))

#  2.2 find the logos------
library(ggseqlogo)
distinct(tbRaw,FileName)$FileName
distinct(tbMature,FileName)$FileName
# raw seq
l_cell_lines_PI <- vector("list",8)

ggseqlogo(sapply(distinct(tbRaw,FileName)$FileName, function(x){
  l_cell_lines_PI[[x]] <- str_sub(filter(tbRaw,FileName == x)$V6,1,15)
  
}), method = "prob", ncol = 4)

# mature seq
l_cell_lines_PI <- vector("list",8)

logos_cellines <- ggseqlogo(sapply(distinct(tbMature,FileName)$FileName, function(x){
  l_cell_lines_PI[[x]] <- str_sub(filter(tbMature,FileName == x)$matureSEQ,1,15)
  
}), method = "prob", ncol = 4)

#  2.3 find length distributions-------
tbRaw %>% 
  #filter(!FileName == "CACO2") %>% 
  ggplot( aes(x = length )) +
  geom_histogram(breaks = seq(15,35, by = 1),
                 color = "red")+
    theme_minimal() +
  facet_wrap(~ FileName, nrow = 2)

tbMature %>% 
  #filter(!FileName == "CACO2") %>% 
  ggplot( aes(x = mature_len )) +
  geom_histogram(breaks = seq(22,35, by = 1),
                 color = "red")+
    theme_minimal() +
  facet_wrap(~ FileName, nrow = 2)

## 3.1 using data table------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,data_filt)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as_vector(sumsDT$Reads),as_vector(round(newSUms$sums))))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,NUCLEO_5N])

path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/PCA_hypergeo/Cell_lines_raw_counts.txt" # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  4.1 Identify features of piRNA reads------

# load dataset
DT2 <- fread(path) %>%  as_tibble()
cpm_TMMwzp <- fread("Cell_lines_norm_TMMwzp_cpm_counts.txt") %>% as_tibble()
lcpm_TMMwzp <- fread("Cell_lines_norm_TMMwzp_logcpm_counts.txt") %>% as_tibble()

# check piRNAs found
DT2 %>% 
  filter(str_detect(smallRNA,"piRNA")) %>% 
  filter_all(all_vars(.>0)) %>% 
  arrange(desc(COLO205)) 

cpm_TMMwzp %>% 
  filter(str_detect(smallRNA,"piRNA")) %>% 
  arrange(desc(COLO205)) 

Lmat <-lcpm_TMMwzp %>% 
  filter(str_detect(smallRNA,"piRNA")) %>% 
  gather(key = cell_lines, value = value, 2:ncol(lcpm_TMMwzp)) %>% 
  spread(key = names(lcpm_TMMwzp)[1],value = "value") %>% 
  column_to_rownames("cell_lines")


library(ggbiplot)
pca <- prcomp(Lmat,scale. = TRUE)
ggbiplot(pca,groups = group,var.axes= FALSE
         )+
  theme_minimal()+
  aes(colour = group, shape = status)+
  geom_point(size = 5)

#  5.1 NOISEQ for pairwise  (data packaging)-------
library(NOISeq)
#  load the data
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/PCA_hypergeo/Cell_lines_raw_counts.txt"

DT2 <- read_delim(path,delim = "\t") %>%  
  as_tibble() 

mat <- DT2 %>% 
  column_to_rownames("smallRNA") %>% as.matrix()

group <- as.factor(colnames(mat))
batch <- as.factor(1:8)
status <- as.factor(c("MSS_low","MSS_high",
                      "MSI_no","MSS_low","MSS_low","MSI_no","MSS_high","MSS_low"))
samples <- data.frame(group,batch,status,row.names = colnames(mat))
mybiotypes <-  separate(DT2,smallRNA,c("DB","sRNA"),sep = "_match_") %>% 
  select(DB)

mybiotypes <- mybiotypes$DB
names(mybiotypes) <-  DT2$smallRNA
mydata <- readData(data = mat, factors = samples, biotype = mybiotypes)

#  5.2 NOISEQ explore data with -------
mybiodetection <- dat(mydata, k = 0, type = "biodetection", factor = NULL)
mycountsbio <- dat(mydata, factor = NULL, type = "countsbio")
#
par(mfrow = c(1, 1))
explo.plot(mycountsbio, samples = NULL, toplot = 1, plottype = "barplot")
#  5.3 NOISEQ filter data------
myfilt <- filtered.data(mat, factor = group,
                        norm = FALSE, depth = NULL,
                        method = 1, cv.cutoff = 100, cpm = 0.3)
myfiltdata <- readData(data = myfilt, factors = samples, biotype = mybiotypes)

#  5.4 NOISEQ DE with one sample -----
myresults <- noiseq(myfiltdata, 
                    factor = "group", 
                    conditions = c("COLO205","CACO2"),
                    k = NULL,
                    norm = "tmm", pnr = 0.2,
                    nss = 5, v = 0.02, lc = 1, replicates = "no")

mynoiseDE <- degenes(myresults, q = 0.9, M = NULL)
mynoiseDE %>% 
  as_tibble(rownames = "smallRNA") %>% 
  arrange(desc(M,prob)) %>% 
  filter(str_detect(smallRNA,"piR"))

DE.plot(myresults, q = 0.9 , graphic = "MD")
DE.plot(myresults, q = 0.9 , graphic = "expr", log.scale = TRUE)

list_compar <- list(CACO2 = c("COLO205","CACO2"),
                    HCT116 = c("COLO205","HCT116"),
                    HT115 = c("COLO205","HT115"),
                    HT29 = c("COLO205","HT29"),
                    RKO = c("COLO205","RKO"),
                    SW1417 =c("COLO205","SW1417"),
                    SW403 = c("COLO205","SW403"))

myresults <- lapply(list_compar, function(x){
  noiseq(myfiltdata, 
      factor = "group", 
      conditions = x,
      k = NULL,
      norm = "tmm", pnr = 0.2,
      nss = 5, v = 0.02, lc = 1, replicates = "no")
} )

mynoiseDE <- lapply(myresults,function(x){
  degenes(x, q = 0.9, M = NULL)
})

mynoiseDE2 <- lapply(as.list(names(mynoiseDE)), function(x){
  mynoiseDE[[x]] %>% as_tibble(rownames = "smallRNA")
})
names(mynoiseDE2) <- names(mynoiseDE)

complete_DEs <- rbindlist(mynoiseDE2,use.names = TRUE,idcol = "comparison",fill = TRUE) %>%
  as_tibble() %>% 
  write_rds("NOISEQ_complete_DEs.rds")

#complete_DEs <- read_rds("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/PCA_hypergeo/NOISEQ_complete_DEs.rds")

piRNADE <- complete_DEs %>% 
  filter(Biotype == "piRNA") 

piRname <- piRNADE %>% 
  select(smallRNA,comparison,M) %>% 
  spread(key = comparison, M) %>% drop_na()

piRNADE %>% 
  filter(smallRNA %in% piRname$smallRNA)

#  5.5 NOISEQ  heatmap -----
library(ComplexHeatmap);library(circlize)
mat <- fread("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/only_cell_lines/PCA_hypergeo/Cell_lines_norm_TMMwzp_cpm_0.3filter_counts.txt") %>% 
  as_tibble() %>% 
  filter(smallRNA %in% piRname$smallRNA) %>% 
  separate(smallRNA,c("DB","sRNA"), sep = "_match_") %>% 
  select(-DB) %>% 
  column_to_rownames("sRNA") %>% as.matrix()

scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat)) 

f <- colorRamp2(c(-2,median(scaled_mat),2),c("blue","white","yellow"))


ha <- HeatmapAnnotation(df)
# plot Heatmap
# cluster only rows
Heatmap(scaled_mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = TRUE,
        #column_order = c(1,2,6,5,7,3,4),
        heatmap_legend_param = cm,
        show_row_names = TRUE,
        name = "z-score equivalent", 
        column_title = str_glue("Heatmap of {nrow(mat)} piRNAs"),
        column_title_side = "top",
        km = 1)

f <- colorRamp2(c(min(mat),mean(mat),max(mat)),c("blue","white","yellow"))
Heatmap(mat,
        top_annotation = ha,
        col=f,
        show_row_dend = TRUE,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = TRUE,
        #column_order = c(1,2,6,5,7,3,4),
        heatmap_legend_param = cm,
        show_row_names = TRUE,
        name = "cpm", 
        column_title = str_glue("Heatmap of {nrow(mat)} piRNAs"),
        column_title_side = "top",
        km = 1)

#  4.1.0 data pre-processing with edger----------------------------------------
library(edgeR);library(wesanderson)
# DT2 <- as_tibble(fread("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/Cell_lines_testis_raw_counts.txt" ,header = TRUE))
#DT2 <- as_tibble(DT_last)
DT2 <- DT2 %>% as_tibble %>%  column_to_rownames("smallRNA")

#mat <- as.matrix(DT2)
mat <- as.matrix(DT2)
group <- as.factor(colnames(mat))
batch <- as.factor(1:8)
status <- as.factor(c("MSS_low","MSS_high",
                      "MSI_no","MSS_low","MSS_low","MSI_no","MSS_high","MSS_low"))
samples <- data.frame(group,batch,status,row.names = colnames(mat))
x <- DGEList(counts = mat,samples = samples,lib.size = colSums(mat),norm.factors = rep(1,ncol(mat)))

## 4.1.1 filtering 1 cpm------------------------------------
cpm <- cpm(x)
cpm(10,mean(x$samples$lib.size))
lcpm <- cpm(x,log=TRUE,prior.count=5)
table(rowSums(x$counts==0)==9)
keep.exprs <- rowSums(cpm>1)>=1
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
#x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
#lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

#cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_TMMwzpiRNA

#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "Cell_lines_testis_normalized_TMM_logcpm_noSK_counts.txt",delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
#  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "all_mat_Cell_testis_noSK_norm_TMMwzp_logcpm_counts.txt",delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
#  filter(str_detect(smallRNA,"piR")) %>% 
  write_delim(path = "all_mat_Cell_testis_noSK_norm_TMMwzp_cpm_counts.txt",delim = "\t")
  
  
## 4.1.2 density plot-------------------------  
pal1 <- c("#9986A5", "#79402E", "#CCBA72",
          "#FD6467","#F3DF6C","#CEAB07","#D5D5D3","#24281A","#FF0000")
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=2, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")
lcpm <- cpm(x1, log=TRUE, prior.count=2)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,0.21), las=2, 
     main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

## 4.1.3 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)
hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

## 4.1.4 boxplots---------------------------------------------------
par(mfrow=c(1,3))
par(mar=c(8,3,2,3))
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col=pal1, main="")
title(main="A. Unnormalized data",ylab="Log-cpm")


#-----------------TMM
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMM, las=2, col=pal1, main="")
title(main="B. TMM Normalized data",ylab="Log-cpm")


#-----------------TMMwzp
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
boxplot(lcpm_TMMwzp, las=2, col=pal1, main="")
title(main="C. TMMwzp Normalized data",ylab="Log-cpm")

## 4.1.5 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(x1_TMMwzp$counts))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")
legend("topleft",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )


hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")

## 4.1.6 MDS plot, all RNA----------------------------------------------

par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(1,2))

legend("topleft",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS.DGEList(x1_TMMwzp,labels = x1$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzp,labels = x1_TMMwzp$samples$group,
                col=as.numeric(x1_TMMwzp$samples$status),dim.plot = c(5,6))

legend("topleft",c("4","5","6"),
       fill = as.numeric(x1_TMMwzp$targets$status),cex=1.5,inset = c(.01,.09))

## 4.1.7 mat filter for piRNAs-------------------------------------------
mat_piR <- as.matrix(x1_TMM$counts %>% 
  as_tibble(rownames=NA) %>% 
  rownames_to_column(var="smallRNA") %>%   
  filter(str_detect(smallRNA,"piRNA_m"))%>%  
  column_to_rownames("smallRNA"))
 x1_TMMwzpiRNA<- x1_TMMwzp[rownames(mat_piR),,keep.lib.sizes=TRUE]
 lcpm_TMMwzpiRNA <- lcpm_TMMwzp[rownames(mat_piR),]
## 4.1.8 piRNAs Hierarchical, clustering---------------------------------------
par(mfrow=c(1,3))

d <- dist(t(lcpm_TMMwzpiRNA))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-ward.D2",main="")



hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-complete",
          main="Hierarchical clustering with only piRNAs")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = as.numeric(x1_TMMwzp$samples$status),
          labels = x1_TMMwzp$samples$group,xlab = "euclidean-average",main="")
legend("topright",legend=levels(x1_TMMwzp$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzp$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )
## 4.1.9 piRNAs MDS plot mat1------------------------------------------------
mat_piR


par(mfrow=c(1,3))
par(mar=c(8,5,2,3))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(1,2))

legend("topleft",legend=levels(x1_TMMwzpiRNA$samples$status),
       fill =as.numeric(as.factor( levels(x1_TMMwzpiRNA$samples$status))),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(3,4))

plotMDS.DGEList(x1_TMMwzpiRNA,labels = x1_TMMwzpiRNA$samples$group,
                method = "logFC",
                col=as.numeric(x1_TMMwzpiRNA$samples$status),dim.plot = c(5,6))

  

#
```

# spike-ins                                                                 
```{r}
library(data.table);library(tidyverse);library(edgeR)
#  old data---------
setwd("/home/rstudio/Documents/piRNAs")
list.files(pattern = "spike")
x <- readRDS("completeData_x_spike_ins.rds")
cpm <- cpm(x)
keep.exprs <- rowSums(cpm>1)>=3
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
x1_TMM <- calcNormFactors(x1, method = "TMM")
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE,log = TRUE,prior.count = 5)
x1_TMM$samples$batch <- c(rep(c(4,5,6),each = 2,2),rep(c(1,2,3),each = 2)) 

design <- model.matrix(~0+ x1_TMM$samples$treat + x1_TMM$samples$batch)
colnames(design) <- c(levels(x1_TMM$samples$treat),"batch")
rownames(design) <- rownames(x1_TMM$samples)

voom_TMMwzp <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
vfit <- lmFit(voom_TMMwzp)
cm <- makeContrasts(
  treat_V_non = treated-untreated,
  levels = design)
vfit <- contrasts.fit(vfit,cm)
vfit <- eBayes(vfit,robust = TRUE)

topTreat_V_non <- topTable(vfit,coef = "treat_V_non",
                         number = nrow(vfit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA") %>% 
  arrange(logFC)

topTreat_V_non %>% 
  filter(str_detect(smallRNA,"spike"))
spikesins <- cpm_TMM %>% 
  as_tibble(rownames = "smallRNA") %>% 
  filter(str_detect(smallRNA,"spike|Uni"))
lspikesins <- lcpm_TMM %>% 
  as_tibble(rownames = "smallRNA") %>% 
  filter(str_detect(smallRNA,"spike|Uni"))


#  plot the differences for methylated spike ins--------
spikesins <- spikesins %>% 
  mutate("Spikes ins" = ifelse(str_detect(smallRNA,"spikeB"),"Non Methylated piRNA",
                       ifelse(str_detect(smallRNA,"spikeM"),"Methylated piRNA","miRNA")))

gspikes <- gather(spikesins, "sample", "cpm",matches("^CIT|^NUC|^TOT")) %>% 
  group_by(sample,`Spikes ins`) %>% 
  mutate(Sums= cpm/sum(cpm))

gspikes %>% 
  group_by(sample) %>% 
  filter(!`Spikes ins` == "miRNA",str_detect(sample,"TOT")) %>% 
ggplot( aes(x = sample , y = cpm, fill = `Spikes ins`) ) + 
    geom_bar(stat = "identity", position = "fill")+
    theme_minimal() +
  scale_y_continuous("Percentage of counts per million",
                       labels = scales::percent,breaks = seq(0,1,by=0.25))+
  coord_flip()
    

gspikes %>% ungroup() %>% 
  filter(!`Spikes ins` == "miRNA",str_detect(sample,"TOT")) %>% 
  select(-Meth,-Sums) %>% 
  mutate(Samples = rep(c("treated","non-treated"), each = 4,3)) %>% 
  group_by(sample,`Spikes ins`) %>% 
  mutate(MED = median(cpm)) %>% 
ggplot( aes(x = Samples , y = MED, fill = `Spikes ins`) ) + 
    geom_bar(stat = "identity", position = "fill")+
    theme_minimal() +
  scale_y_continuous("Percentage of Median counts per million",
                       labels = scales::percent,breaks = seq(0,1,by=0.25))+
  coord_flip()


#  spikes-ins in TOTAL samples-------

# find the spikes ins TOTAL samples
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_TOTAL_treatment/")
path  <-  "/home/rstudio/Documents/piRNAs/"
list.files(path , pattern = "spike")
spikes <- read_delim(str_glue("{path}spikes_ins.txt"), delim = "\t") %>% 
  filter(str_detect(spike_in,"spike")) %>% 
  select(smallRNA = spike_in ,starts_with("TOTAL"))

# load the TOTAL samples reads table
list.files(pattern = ".txt")
TOTAL <- read_delim(str_glue("TOTAL_raw_counts.txt"), delim = "\t")

DT <- bind_rows(spikes,TOTAL) %>% 
  write_delim("TOTAL_spike_raw_counts.txt")


#  visualization of results--------

fil <- read_delim(str_glue("Complete_list_Treat_vs_No_TOTAL_SPIKE_INS.txt"),delim = "\t") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))

# set values of filtering
lfc = 1
fdr = 0.01

enriched <- fil %>% 
  filter( adj.P.Val < fdr)

x1_voomTMM <- read_rds(str_glue("x1_voomTMM_SPIKES_INS_TOTAL.rds"))

Exs <- x1_voomTMM$E %>% 
  as_tibble(rownames = "smallRNA") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))

my_data <- inner_join(Exs,fil,by="smallRNA")
my_data <- my_data %>% 
  group_by(smallRNA) %>% 
  mutate(TOTAL_meanExpr = mean(c(TOTAL_1,TOTAL_2,TOTAL_3)),
         TOTAL_NAIO4_meanExpr = mean(c(TOTAL_1_NaIO4,TOTAL_2_NaIO4,TOTAL_3_NaIO4)),
         EnRNA=ifelse(smallRNA %in% enriched$smallRNA,
                      smallRNA,"not_Enriched_match_genome")) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  select(smallRNA,TOTAL_meanExpr,TOTAL_NAIO4_meanExpr,logFC,AveExpr,DB)

# plot the enrichment 
my_data %>% 
ggplot()+
  geom_point(mapping=aes(x=TOTAL_meanExpr,
                         y=TOTAL_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(enriched)} Enriched smallRNAs in TOTAL after treament, FDR < {fdr}"),
     x="Non-Treated log2 average expression",
     y="Treated log2 average expression",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_data$DB)),
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","darkorange1",  "orange"))
#1 ensembl #2 miRNA #3 noAnnot #4 not_Enriched 
#5 piRNA #6 rfam #7 rRNA #8 tRNA #9 tRNA_mature

# only pirnas
filtered <- filter(enriched,str_detect(smallRNA,"piRNA"))
filtered <- filter(enriched,str_detect(smallRNA,"spike|piRNA"))
# plot the enrichment 

my_data %>% 
    filter(str_detect(smallRNA,"spike|piR")) %>%
ggplot()+
  geom_point(mapping=aes(x=TOTAL_meanExpr,
                         y=TOTAL_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(filtered)} Enriched piRNAs in TOTAL after treament, FDR < {fdr}"),
     x="Non-Treated log2 average cpm",
     y="Treated log2 average cpm",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_data$DB))[c(4:5,8:11)],
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","deeppink","darkviolet","navy","red4" ,"darkorange1","orange")[c(4:5,8:11)])



#  spikes-ins in CYT and NUC samples-------
data_filt <- c("CITOSOL","NUCLEO")
# find the spikes ins CYT and NUC samples
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/")
path  <-  "/home/rstudio/Documents/piRNAs/"
list.files(path , pattern = "spike")
spikes <- read_delim(str_glue("{path}spikes_ins.txt"), delim = "\t") %>% 
  filter(str_detect(spike_in,"spike")) %>% 
  select(smallRNA = spike_in ,starts_with(data_filt[1]),starts_with(data_filt[2]))

# load the CYTOSOL & NUCLEUS samples reads table
list.files(pattern = ".txt")
CYT_NUC <- read_delim(str_glue("Treated_non_CYT_vs_NUC_raw_counts.txt"), delim = "\t")
librSums <- CYT_NUC %>% select(-smallRNA) %>% colSums()

# DT <- bind_rows(spikes,CYT_NUC) %>% 
  write_delim("CYT_NUC_spike_raw_counts.txt")

  
cpm_TMM_spikes <- cpm_TMM %>% 
  as_tibble(rownames = "smallRNA") %>% 
  group_by(smallRNA) %>% 
  mutate(medianCYT_NaIO4 = median(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         medianCYT = median(c(CITOSOL_4C, CITOSOL_5C, CITOSOL_6C)),
         medianNUC_NaIO4 = median(c(NUCLEO_4N_NaIO4, NUCLEO_5N_NaIO4, NUCLEO_6N_NaIO4)),
         medianNUC = median(c(NUCLEO_4N, NUCLEO_5N, NUCLEO_6N))) %>% 
  select(smallRNA,medianCYT_NaIO4,medianCYT,medianNUC_NaIO4,medianNUC)

cpm_TMM_spikes <- cpm_TMM_spikes %>% 
  mutate(LFC_CYT = log2(medianCYT_NaIO4/medianCYT),
         LFC_NUC = log2(medianNUC_NaIO4/medianNUC))
  
#  visualization of results--------
#CYT
filCyt <- read_delim(str_glue("complete_TrevsUnt_Cyt_voomQ_CYT_NUC_SPIKE_INS.txt"),delim = "\t") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))

filCytNOSPIKES <- read_delim(str_glue("complete_TrevsUnt_Cyt_x1_voomQ_CYT_vs_NUC.txt"),delim = "\t") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))


#NUC
filNuc <- read_delim(str_glue("complete_TrevsUnt_Nuc_voomQ_CYT_NUC_SPIKE_INS.txt"),delim = "\t") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))


# set values of filtering
lfc = 1
fdr = 0.01

#CYT
enrichedCyt <- filCyt %>% 
  filter( Cyt_Nucadj.P.Val < fdr,Cyt_NuclogFC > lfc)
#NUC
enrichedNuc <- filNuc %>% 
  filter( Cyt_Nucadj.P.Val < fdr, Cyt_NuclogFC > lfc)

x1_voomQ <- read_rds(str_glue("x1_voomQ_filt2_CYT_NUC_SPIKE_INS.rds"))

Exs <- x1_voomQ$E %>% 
  as_tibble(rownames = "smallRNA") %>% 
  mutate(smallRNA = ifelse(str_detect(smallRNA,"spike"),
                           str_glue("{smallRNA}_match_genome"),smallRNA))

#CYT
my_dataCyt <- Exs %>% 
  select(smallRNA,starts_with("CITOSOL")) %>% 
  inner_join(filCyt,by="smallRNA")
#NUC
my_dataNuc <- Exs %>% 
  select(smallRNA,starts_with("NUCLEO")) %>% 
  inner_join(filNuc,by="smallRNA")

#CYT
my_dataCyt <- my_dataCyt %>% 
  group_by(smallRNA) %>% 
  mutate(CITOSOL_meanExpr = median(c(CITOSOL_4C,CITOSOL_5C,CITOSOL_6C)),
         CITOSOL_NAIO4_meanExpr = median(c(CITOSOL_4C_NaIO4,CITOSOL_5C_NaIO4,CITOSOL_6C_NaIO4)),
         EnRNA=ifelse(smallRNA %in% enrichedCyt$smallRNA,
                      smallRNA,"not_Enriched_match_genome")) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  select(smallRNA,CITOSOL_meanExpr,CITOSOL_NAIO4_meanExpr,Cyt_NuclogFC,Cyt_NucAveExpr,DB)

#NUC
my_dataNuc <- my_dataNuc %>% 
  group_by(smallRNA) %>% 
  mutate(NUCLEO_meanExpr = median(c(NUCLEO_4N,NUCLEO_5N,NUCLEO_6N)),
         NUCLEO_NAIO4_meanExpr = median(c(NUCLEO_4N_NaIO4,NUCLEO_5N_NaIO4,NUCLEO_6N_NaIO4)),
         EnRNA=ifelse(smallRNA %in% enrichedNuc$smallRNA,
                      smallRNA,"not_Enriched_match_genome")) %>% 
  separate(EnRNA,c("DB","sRNA"), sep = "_match_") %>% 
  select(smallRNA,NUCLEO_meanExpr,NUCLEO_NAIO4_meanExpr,Cyt_NuclogFC,Cyt_NucAveExpr,DB)

# plot the enrichment 
# CYT
my_dataCyt %>% 
ggplot()+
  geom_point(mapping=aes(x=CITOSOL_meanExpr,
                         y=CITOSOL_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(enrichedCyt)} Statistically significant smallRNAs in CITOSOL after treament, FDR < {fdr}"),
     x="Non-Treated log2 average expression",
     y="Treated log2 average expression",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_dataCyt$DB)),
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","deeppink","darkviolet","navy","red4" ,"darkorange1","orange"))
#1 ensembl #2 miRNA #3 noAnnot #4 not_Enriched 
#5 piRNA #6 rfam #7 rRNA #8 tRNA #9 tRNA_mature

#NUC
my_dataNuc %>% 
ggplot()+
  geom_point(mapping=aes(x=NUCLEO_meanExpr,
                         y=NUCLEO_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(enrichedNuc)} Statistically significant smallRNAs in NUCLEO after treament, FDR < {fdr}"),
     x="Non-Treated log2 average expression",
     y="Treated log2 average expression",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_dataNuc$DB)),
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","deeppink","darkviolet","navy","red4" ,"darkorange1","orange"))

# only piRNAs
filtered <- filter(enriched,str_detect(smallRNA,"piRNA"))
filteredCyt <- filter(enrichedCyt,str_detect(smallRNA,"spike|piRNA"))
filteredNuc <- filter(enrichedNuc,str_detect(smallRNA,"spike|piRNA"))
# plot the enrichment 

#CYT
my_dataCyt %>% 
    filter(str_detect(smallRNA,"spike|piR")) %>%
ggplot()+
  geom_point(mapping=aes(x=CITOSOL_meanExpr,
                         y=CITOSOL_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(filteredCyt)} Statistically significant piRNAs in Cytosol after treament, FDR < {fdr}"),
     x="Non-Treated log2 average cpm",
     y="Treated log2 average cpm",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_dataCyt$DB))[c(4:5,8:11)],
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","deeppink","darkviolet","navy","red4" ,"darkorange1","orange")[c(4:5,8:11)])

#NUC
my_dataNuc %>% 
    filter(str_detect(smallRNA,"spike|piR")) %>%
ggplot()+
  geom_point(mapping=aes(x=NUCLEO_meanExpr,
                         y=NUCLEO_NAIO4_meanExpr,
                         colour= DB))+
labs(title =str_glue("{nrow(filteredNuc)} Statistically significant piRNAs in Nucleus after treament, FDR < {fdr}"),
     x="Non-Treated log2 average cpm",
     y="Treated log2 average cpm",
     color="Selected RNAs") +
  theme_bw()+
scale_color_manual(labels=levels(as.factor(my_dataCyt$DB))[c(4:5,8:11)],
                   values = c("gold","cornflowerblue","grey","black","chartreuse4", "darkorchid1","firebrick1","deeppink","darkviolet","navy","red4" ,"darkorange1","orange")[c(4:5,8:11)])


```

# COLO205 vs HCT116 vs SW1417                                               
```{r}
library(data.table);library(tidyverse);library(rafalib)
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
name_file <- "PIWIL_plus_minus"
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/COLO205_HCT116_SW1417/")
#  1.1 List files merged and outputs of all ----------
# first select the correct tables
data_filt <- "COLO205|HCT116|SW1417|TOTAL"
path <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments COLO205|HCT116|SW1417
listRNA <- list_smallRNA[str_detect(list_smallRNA,data_filt)] 
exclude <- as.list(c("IPP","NaIO4",
                     "7MIR_","INPUT","NUCL","PNF","CIT","Rpph","RppH"))

for(i in exclude){
  listRNA <- listRNA[!str_detect(listRNA,i)]
}

#load data to data.table
listRNA <- sapply(listRNA, function(x)str_glue({path},{x}),simplify=FALSE)
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

#  1.2 using data table ------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>%
  as_tibble %>%  
  unite("smallRNA",RNA_type,V4,sep = "_") 
DT %>% 
write_rds("DT_table.rds")

DT <- read_rds("DT_table.rds")
# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Results_6_Run_Next_Seq_Cell/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,data_filt)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as_vector(sumsDT$Reads),as_vector(round(newSUms$sums))))
sumsDT$Reads
newSUms$sums

# filter low efficient samples : SW1417, SW1417-TOT and COLO205-TOT 
DT_last <- DT_last %>% 
  as_tibble() %>% 
  filter(!FileName == "6_Run_Next_Seq_Cell_all_DBs/Data/SW1417-TOT_smallRNA.txt",
         !FileName == "1_Cell_Lines_all_DBs/Data/SW1417_smallRNA.txt",
         !FileName == "6_Run_Next_Seq_Cell_all_DBs/Data/COLO205-TOT_smallRNA.txt")

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,"HCT116-TOT"])
newSUms

path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/COLO205_HCT116_SW1417/{name_file}_noSW_noBatch_raw_counts_{date}.txt") # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

#  2.1 edgeR and voom Data packaging ------
library(edgeR);library(wesanderson)
path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/COLO205_HCT116_SW1417/{name_file}_noSW_noBatch_raw_counts_Mar_23_2019.txt")
#path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/COLO205_HCT116_SW1417/{name_file}_noSW_raw_countsMar_22_2019.txt")
DT <- read_delim(path,delim = "\t") %>% 
  select(smallRNA,COLO205,
         #COLO205_TOT = `COLO205-TOT`
         TOTAL_1,
         TOTAL_2,TOTAL_3,HCT116, HCT116_TOT = `HCT116-TOT`, HCT116_NonTreat = HCT116_Non_Treated)

DT2 <- as_tibble(DT) %>%
   column_to_rownames("smallRNA")

mat <- as.matrix(DT2)
group <- as.factor(c(rep("COLO205",4),rep("HCT116",3)))# change this
batch <- as.factor(1:7)                                # change this
samples <- data.frame(group,batch,row.names = colnames(mat))
design <- model.matrix(~0+group)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
x <- DGEList(counts = mat,
             samples = samples,
             lib.size = colSums(mat),
             norm.factors = rep(1,ncol(mat)))
#  2.2 edgeR filtering -----
cpm(10,mean(x$samples$lib.size))
cpm  <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
summary(lcpm)

keep.exprs <- filterByExpr.DGEList(x,group = group)
x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]
dim(x);dim(x1)

#  2.3 edgeR density plot -------
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data of {name_file},
                    ~ {round(lcpm.cutoff,3)} log-cpm"), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")
#  2.4 edgeR histogram plot -------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)
hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)
#  2.5 edgeR normalization  --------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM_qual <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp_qual <- voomWithQualityWeights(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ_qual <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)
#new
x1_voomTMM <- voom(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voom(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voom(x1,design = design,normalize.method = "quantile",plot = TRUE)

lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)

cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{name_file}_noBatch_{date}.txt"),sep = "\t",col.names = TRUE)
lcpm_TMM %>% as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("logcpm_TMM_{name_file}_noBatch_{date}.txt"),delim = "\t")

write_rds(x1_voomQ,str_glue("x1_voomQ_{name_file}_noBatch.rds"))
write_rds(x1_voomQ_qual,str_glue("x1_voomQ_qual_{name_file}_noBatch.rds"))
write_rds(x1_voomTMM,str_glue("x1_voomTMM_{name_file}_noBatch.rds"))
write_rds(x1_voomTMM_qual,str_glue("x1_voomTMM_qual_{name_file}_noBatch.rds"))
write_rds(x1_TMM,str_glue("x1_TMM_{name_file}_noBatch.rds"))
#  2.6 edgeR boxplots -------
x1_voomQ <- read_rds(str_glue("x1_voomQ_{name_file}.rds"))
x1_voomTMM <- read_rds(str_glue("x1_voomTMM_{name_file}_noBatch.rds"))

par(mfrow=c(2,3))
par(mar=c(6,5,2,1)+ 0.1)
#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
lcpm_TMMwzp <- cpm(x1_TMMwzp,log=TRUE,prior.count=5)
boxplot(lcpm_TMMwzp, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
lcpm_uQ <- cpm(x1_uQ,log=TRUE,prior.count=5)
boxplot(lcpm_uQ, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
lcpm_RLE <- cpm(x1_RLE,log=TRUE,prior.count=5)
boxplot(lcpm_RLE, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

#raw data
boxplot(lcpm, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")
#voomTMMwzp data
boxplot(x1_voomTMMwzp$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom TMMwzp Normalized filtered data",ylab="Log-cpm")
#voomQ data
boxplot(x1_voomQ$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")
#voomTMM qWe
boxplot(x1_voomQ_qual$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom Quantile quality weights Normalized filtered data",ylab="Log-cpm")
#voomTMM qWe
boxplot(x1_voomTMM_qual$E, las=2, col= c(rep("#899DA4",2),rep("#C93312",3),rep("#FAEFD1",3)), main="",names=group)
title(main="A. Voom TMM quality weights Normalized filtered data",ylab="Log-cpm")


#  2.7 edgeR hierarchical clustering ------
par(mfrow=c(1,3))
d <- dist(t(x1_voomTMM$E))

hc <- hclust(d,method = "ward.D2")
myplclust(hc,lab.col = c(rep("#C93312",4),rep("#899DA4",3)),
          labels = colnames(x1_voomTMM),xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = c(rep("#C93312",4),rep("#899DA4",3)),
          labels = colnames(x1_voomTMM),xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, voomTMM")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = c(rep("#C93312",4),rep("#899DA4",3)),
          labels = colnames(x1_voomTMM),xlab = "euclidean-average",main="")

legend("topright",legend=levels(x1_voomTMM$targets$group),
       fill = c("#C93312","#899DA4"),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

#  2.8 edgeR MDS plots ------
par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)
plotMDS(x1_voomTMM$E,labels = colnames(x1_TMM),
                col=c(rep("#C93312",4),rep("#899DA4",3)),dim.plot = c(1,2))

legend("left",legend=levels(x1_TMM$samples$group),
       fill = c("#C93312","#899DA4"),
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomTMM$E,labels = colnames(x1_TMM),
                col=c(rep("#C93312",5),rep("#899DA4",3)),dim.plot = c(3,4),
        main= str_glue("MDS plot of {name_file} with voom TMM \n normalized counts  not considering batch"))
#  2.9 edgeR DE analysis -------
vfit <- lmFit(x1_voomTMM,design=design)
con_mat <- makeContrasts(
  PIWIL = COLO205-HCT116,
  levels = design)
vfit <- contrasts.fit(vfit,con_mat)
efit <- eBayes(vfit,robust = TRUE)

write_rds(efit,str_glue("efit_x1_voomTMM_noBatch_{name_file}.rds"))
vfit <- read_rds("efit_x1_voomTMM_noBatch_{name_file}.rds")

#PIWIL plus minus
topPIWIL <- topTable(efit,coef = "PIWIL",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA")

topPIWIL %>% 
  filter(adj.P.Val<0.01) %>% 
  filter(str_detect(smallRNA,"piR")) %>%
  arrange(desc(logFC))
  select(smallRNA,logFC) -> LFCstable

write_delim(topPIWIL,path=str_glue("LFCs_voomTMM_noBatch_{name_file}_{date}.txt"),delim = "\t")
#  3.1 Heatmap of cell lines -----
library(ComplexHeatmap);library(circlize)
DEs <- read_delim("LFCs_voomTMM_noBatch_PIWIL_plus_minus_Mar_23_2019.txt", 
                  delim = "\t") %>% 
  filter(adj.P.Val < 0.01, 
         str_detect(smallRNA,"piRN"),abs(logFC) > 2)  

x1_voomTMM <- read_rds(str_glue("x1_voomTMM_PIWIL_plus_minus_noBatch.rds"))
lcpm_TMM <- read_delim("logcpm_TMM_PIWIL_plus_minus_noBatch_Mar_23_2019.txt",
                      delim = "\t")
# "create" the matrix
mat <- x1_voomTMM$E
names(mat) <- rownames(mat)

# subset the matrix
mat <- subset(mat,rownames(mat) %in% DEs$smallRNA)
mat %>% as_tibble() %>% nrow
DEs%>% nrow

#prepare the matrix
scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
df <- data.frame(cell_line = colnames(scaled_mat))  
df$cell_line <- df$cell_line %>% 
  str_replace("TOTAL","COLO")  

ha <- HeatmapAnnotation(df)
f <- colorRamp2(c(-1,median(scaled_mat),1),c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(scaled_mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",clustering_distance_columns = "spearman",
        clustering_method_columns ="ward.D2",
        show_row_names = FALSE,name= "z-score equivalent", 
        column_title = str_glue("Heatmap of DE {nrow(DEs)} piRNAs between COLO205 and HCT116"),
        column_title_side="top")

# lcpm matrix 
# "create" the matrix
mat <- lcpm_TMM %>% 
  column_to_rownames("smallRNA") %>% 
  as.matrix()
names(mat) <- rownames(mat)

# subset the matrix
mat <- subset(mat,rownames(mat) %in% DEs$smallRNA)
mat %>% as_tibble() %>% nrow
DEs%>% nrow

#prepare the matrix
hist(mat)
df <- data.frame(cell_line = colnames(scaled_mat))  
df$cell_line <- df$cell_line %>% 
  str_replace("TOTAL","COLO")  

ha <- HeatmapAnnotation(df)
f <- colorRamp2(c(min(mat),median(mat),max(mat)),c("blue","white","yellow"))

# plot Heatmap
# cluster rows and columns
Heatmap(mat,top_annotation = ha,
        col=f,show_row_dend = TRUE,clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",clustering_distance_columns = "spearman",
        clustering_method_columns ="ward.D2",
        show_row_names = FALSE,name= "cpm", 
        column_title = str_glue("Heatmap of DE {nrow(DEs)} piRNAs between COLO205 and HCT116"),
        column_title_side="top")
#  3.2 cross IPP with identified piRNAs ------
#      PIWIL + - , DE 
DEs <- read_delim("LFCs_voomTMM_noBatch_PIWIL_plus_minus_Mar_23_2019.txt", 
                  delim = "\t") %>% 
  filter(adj.P.Val < 0.01, 
         str_detect(smallRNA,"piRN"),abs(logFC) > 2) 

pathN <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/N_enriched_RNA.txt"
pathC <-  "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/C_enriched_RNA.txt"

IPPN <- read_delim(pathN,delim = "\t")
IPPC <- read_delim(pathC,delim = "\t")
DEs %>% filter(smallRNA %in% IPPN$smallRNA)
DEs %>% filter(smallRNA %in% IPPC$smallRNA)
PIWILn <- inner_join(IPPN, DEs, by = "smallRNA")

PIWILn %>% write_delim("PIWIL_in_IPPN.txt")

```
# run subread for the identification of mRNAs in IPP                        
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(Rsubread);library(tidyverse)
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/")
# 1.1  Build the index                                                      ----
dir.create("Rsubread_index")
path <- "~/Documents/GRCh38.p12.genome.fa"
setwd("Rsubread_index")
#buildindex("GRCh38.primary_assembly",reference = path, gappedIndex=FALSE,
           indexSplit=FALSE, memory=16000,
           TH_subread=100,colorspace=FALSE)
# 1.2  Find long_read files and align them                                  ----
# dir.create("Rsubread_alignments")
path <- "~/Documents/SPORTS_results/Result/"
name_file <- "COLO205_IPP"
list_longReads <- list.files(path,pattern = "_too_long_reads.fa", recursive = TRUE)
list_longIPP <- list_longReads[str_detect(list_longReads,name_file)] 
names_longIPP <- str_remove(basename(list_longIPP),"_too_long_reads.fa")
list_long <- sapply(list_longIPP, function(x)str_glue({path},{x}),simplify=FALSE)
names(names_longIPP) <- names(list_long)
# align them
path_gtf <- "~/Downloads/gencode.v29.chr_patch_hapl_scaff.annotation.gtf.gz"
lapply(names(list_long),function(x)
align(index = "./Rsubread_index/hg38",list_long[x], type = "rna",input_format = "FASTA",
      output_format="BAM",
      output_file = paste("./Rsubread_alignments/",
                          names_longIPP[x],"subread.BAM",sep="_"),
      nBestLocations=10,
      nthreads=7,
      sortReadsByCoordinates = TRUE,
      useAnnotation = TRUE,
      annot.ext = path_gtf,
      isGTF = TRUE))
list.BAM <- list.files(pattern = ".BAM$", recursive = TRUE)
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/"
#list.BAM <- sapply(list.BAM, function(x)str_glue("{path}Rsubread_alignments/{x}"),simplify=FALSE)
fc <- featureCounts(files = unlist(list.BAM),
                    annot.ext =  path_gtf,
                    GTF.attrType.extra = c("gene_type", "gene_name"),
                    isGTFAnnotationFile = TRUE,
                    nthreads = 7,
                    allowMultiOverlap = TRUE,
                    fraction = TRUE,
                    reportReads = "CORE",
                    reportReadsPath = path,
                    strandSpecific = 0)

colnames(fc$counts) <- str_remove(colnames(fc$counts),".+_COLO205_") %>% 
  str_remove("_subread.BAM")
colnames(fc$stat) <- str_remove(colnames(fc$stat),".+_COLO205_") %>% 
  str_remove("_subread.BAM")
fc$targets <- str_remove(fc$targets,".+_COLO205_") %>% 
  str_remove("_subread.BAM")
# fc %>% write_rds(str_glue("feature_counts_{date}.rds"))
# fc <- read_rds(str_glue("feature_counts_Mar_29_2019.rds"))
# 1.3  annotate from ensg to gene symbol                                    ----
library(annotables)
geneID <- fc$annotation$GeneID
genesTib <- grch38 %>% filter(ensgene %in% geneID)

colmdata <-  fc$annotation %>% as_tibble()
Genedata <- inner_join(colmdata, genesTib, by = c("GeneID" = "entrez"))
IPPcounts <- fc$counts %>% as_tibble(rownames = "entrez") %>% 
  mutate(entrez = as.numeric(entrez))
colnames(IPPcounts)[2:length(IPPcounts)] <- str_remove(names_longIPP,"COLO205_")
IPPtargets <- inner_join(IPPcounts,Genedata, by = c("entrez" = "GeneID"))
IPPtargets %>% write_delim(str_glue("IPP_targets_{date}.txt"),delim = "\t")

library("biomaRt")
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
gens <- getBM(attributes = c("hgnc_symbol"), filters = "ensembl_gene_id", 
              values = geneID, mart = ensembl)
# 1.4  limma voom with quantile                                             ----
fc <- read_rds("feature_counts_Mar_29_2019.rds")
library(edgeR)
# 1.5  data packaging                                                       ----
group <- str_replace(colnames(fc$counts),"_[:digit:]$","")  %>%
  str_replace("_Ctrl_","_Control_") %>% 
  str_replace("IPP_","") %>% 
  as_factor()
                    
batch <- str_replace(colnames(fc$counts),"IPP_.+_.+_","") %>% 
  str_replace(".+_","") %>% as_factor()

samples <- data.frame(group,batch,row.names = colnames(fc$counts))

# 1.6  remove duplicates                                                    ----
fc$annotation <- fc$annotation %>% 
  as_tibble() %>% arrange(Chr) %>%
  distinct(gene_name,.keep_all = TRUE)

fc$counts <- fc$counts[fc$annotation$GeneID,]
# 1.7  remove zero entries                                                  ----
keep <- (rowSums(fc$counts) > 0) 
fc$counts <- fc$counts[keep,]
# 1.8  make the DGE list                                                    ----
x <- DGEList(counts = fc$counts,samples = samples,
             lib.size = colSums(fc$counts),
             norm.factors = rep(1,ncol(fc$counts)))
design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
# 1.9  filtering                                                            ----
#keep.exprs <- filterByExpr.DGEList(x,design = design,
                                 #  min.count = 1,
                                  # min.total.count = 3)
keep.exprs <- filterByExpr.DGEList(x,design = design)
                                   #min.count = 1,
                                   #min.total.count = 3)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

counts <- fc$counts[rownames(x1$counts),] %>% 
  as_tibble(rownames = "GeneID") 
  
fc$annotation %>% 
  dplyr::select(GeneID,gene_name) %>% 
  inner_join(counts,by = "GeneID") %>%
  write_delim(str_glue("mRNA_IPP_filtered_data_{date}"), delim = "\t")
# 1.10 limma-voom  with TMM                                                 ----
calcNormFactors
x1_voomQ <- voom(x1,design = design,normalize.method = "quantile",plot = TRUE)
vfit <- lmFit(x1_voomQ,design=design)
con_mat <- makeContrasts(
  C_v_noAb = C_Ab-Control_noAb,
  N_v_noAb = N_Ab-Control_noAb,
  noAb_v_INPUT = Control_noAb-INPUT,
  C_v_INPUT = C_Ab-INPUT,
  N_v_INPUT = N_Ab-INPUT,
  levels = design)

vfit <- contrasts.fit(vfit,con_mat)
efit <- eBayes(vfit,robust = TRUE)

write_rds(efit,str_glue("efit_voomTMM_{date}.rds"))
efit <- read_rds("efit_voomTMM_May_06_2019.rds")
# find the logFCs
#C_v_noAb
C_v_noAb <- topTable(efit,coef = "C_v_noAb",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "GeneID")
#N_v_noAb
N_v_noAb <- topTable(efit,coef = "N_v_noAb",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "GeneID")
#noAb_v_INPUT
noAb_v_INPUT <- topTable(efit,coef = "noAb_v_INPUT",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "GeneID")
#C_v_INPUT
C_v_INPUT <- topTable(efit,coef = "C_v_INPUT",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "GeneID")
#N_v_INPUT
N_v_INPUT <- topTable(efit,coef = "N_v_INPUT",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% as_tibble(rownames = "GeneID")
# 1.11 create a final list with logFCs                                      ----
l <- list(noAb_v_INPUT,C_v_INPUT,N_v_INPUT,C_v_noAb,N_v_noAb) 
names(l) <- c("noAb_v_INPUT","C_v_INPUT",
            "N_v_INPUT", "C_v_noAb","N_v_noAb")
LFCs <- bind_rows(l, .id = "comparison")
# write_rds(LFCs,"comparison_list_Apr_15_2019.rds")
# LFCs <- read_rds("comparison_list.rds")
LFCstable <- LFCs %>% 
  filter(adj.P.Val < 0.05) %>%
  dplyr::select(comparison,GeneID,logFC) %>% 
  spread(comparison,logFC,drop = TRUE ) 

LFCstable <- fc$annotation %>% as_tibble() %>% 
  dplyr::select(GeneID,gene_name) %>% 
  inner_join(LFCstable,by = "GeneID")

LFCstable %>% 
  filter(C_v_INPUT > 2 , N_v_INPUT > 2) 

exclude <- LFCstable %>% 
  filter(noAb_v_INPUT > (1)) 

(RNAs <- LFCstable  %>% 
  filter(C_v_INPUT > 1 , N_v_INPUT > 1, !GeneID %in% exclude$GeneID) )

# x1_voomQ$E %>% 
  as_tibble(rownames = "GeneID") %>% 
  inner_join(RNAs, by = "GeneID") %>% 
  select(gene_name,everything())  %>% 
  write_delim(path = str_glue("mRNA_IPP_005FDR_{date}.txt"),
              delim = "\t")
# 1.12 cross with the predicted targets                                     ----
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/targets/all_IPPgenes_Mar_06_2019.txt"
Targets <- read_delim(path,delim = "\t")
inner_join(RNAs,Targets, by = c("gene_name" = "external_gene_name")) %>% 
  write_delim(str_glue("mRNA_enriched_Frag_{date}.txt"),delim = "\t")

fc$annotation %>%
  as_tibble() %>% 
  dplyr::select(GeneID,gene_name) %>% 
inner_join(N_v_INPUT,by = "GeneID") %>% 
  inner_join(Targets, by = c("gene_name" = "external_gene_name"))

# 2.1  align mRNA COLO205                                                   ----
# dir.create("Rsubread_alignments")
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/mRNA_COLO205/")

# fc %>% write_rds(str_glue("feature_counts_{date}.rds"))
# fc <- read_rds(str_glue("feature_counts_Mar_29_2019.rds"))
path <- "~/Documents/Databases/Rsubread_DB/GRCh38.primary_assembly"
path_gtf <- "~/Downloads/gencode.v30.primary_assembly.annotation.gtf"
"./QC/"
alignmentmRNA <- align(index = path,
      readfile1 = "./QC/QC_fastp/COLO205_CGATGTAT_L006_R1_001.fastq.gz",
      readfile2 = "./QC/QC_fastp/COLO205_CGATGTAT_L006_R2_001.fastq.gz", 
      type = "rna",
      input_format = "FASTQ",
      output_format="BAM",
      output_file = paste("./Rsubread_alignments/COLO205_mRNA_fastp",
                          "subread.BAM",sep="_"),
      unique=TRUE,
      nthreads = parallel::detectCores(),
      sortReadsByCoordinates = TRUE,
      useAnnotation = TRUE,
      annot.ext = path_gtf,
      isGTF = TRUE)

path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/mRNA_COLO205/QC/QC_fastp/"
fc <- featureCounts(
  files = "~/Documents/new_analysis_workflow/results_R/results_with_DBs/mRNA_COLO205/Rsubread_alignments/COLO205_mRNA_fastp_subread.BAM",
                    annot.ext =  path_gtf,
                    GTF.attrType.extra = c("gene_type", "gene_name"),
                    isGTFAnnotationFile = TRUE,
                    nthreads = parallel::detectCores(),
                    #allowMultiOverlap = TRUE,
                    #fraction = TRUE,
                    countMultiMappingReads = F,
                    isPairedEnd = T,
                    reportReads = "CORE",
                    reportReadsPath = path,
                    strandSpecific = 0)

# fc %>% write_rds(str_glue("mRNA_feature_counts_fastp_{date}.rds"))
# fc  <-  read_rds(str_glue("mRNA_feature_counts_Apr_27_2019.rds"))
mRNA <- fc$counts %>% as_tibble(rownames = "GeneID")
names(mRNA) <- c("GeneID","counts")
mRNA <- mRNA %>% 
 inner_join(as_tibble(fc$annotation),by = "GeneID") %>% 
 select(GeneID, gene_name, counts, gene_type, everything())
mRNA %>% write_tsv(str_glue("mRNA_raw_counts_COLO205_{date}.txt"))

# cpm counts
counts <- mRNA %>% select(GeneID,counts) %>% 
  column_to_rownames("GeneID") %>% as.matrix()

x <- DGEList(counts,norm.factors = rep(1,ncol(counts)))

rpkm(x, gene.length = fc$annotation$Length) %>% 
  as_tibble(rownames = "GeneID") %>%
  inner_join(as_tibble(fc$annotation),by = "GeneID")  %>% 
  select(GeneID, gene_name, rpkm = "counts", gene_type, everything()) %>% 
  write_tsv(str_glue("mRNA_rpkm_counts_COLO205_{date}.txt"))
  filter(str_detect(gene_name,"PIWI|AGO")) %>% arrange(gene_name)


cpm(x) %>% as_tibble(rownames = "GeneID") %>%
  inner_join(as_tibble(fc$annotation),by = "GeneID") %>% 
  select(GeneID, gene_name, cpm = "counts", gene_type, everything()) %>% 
  write_tsv(str_glue("mRNA_cpm_counts_COLO205_{date}.txt")) %>% 
  filter(str_detect(gene_name,"PIWI|AGO")) %>% arrange(gene_name)
 

# 2.2  annotate from entrez to gene symbol                                  ----
library(clusterProfiler)
library(org.Hs.eg.db)
eg <-  as.data.frame(bitr(geneID,
                     fromType = "REFSEQ",
                     toType = "SYMBOL",
                     OrgDb = "org.Hs.eg.db"))
eg <- eg[!duplicated(eg$SYMBOL),]

library(biomaRt)
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
gens <- getBM(attributes = c("hgnc_symbol"), filters = "refseq_mrna", values = geneID,
      mart = ensembl)

library(annotables)
geneID <-  fc$annotation$GeneID %>% as.character()
grch38 %>% group_by(entrez) %>% filter(n() > 1) %>% summarise(n = n()) %>% arrange(desc(n))
genesTib <-grch38 %>% filter(entrez %in% geneID)

colmdata <-  fc$annotation %>% as_tibble()
Genedata <- inner_join(colmdata, genesTib, by = c("GeneID" = "entrez"))
counts <- fc$counts %>% as_tibble(rownames = "entrez") %>% 
  mutate(entrez = as.numeric(entrez)) %>% rename(mRNA_counts=2)
mat1 <- counts %>% column_to_rownames("entrez") %>% as.matrix


GENE_counts <- inner_join(counts,Genedata, by = c("entrez" = "GeneID"))
GENE_counts %>% write_delim(str_glue("COLO205_mRNA_{date}.txt"),delim = "\t")
# cross with the IPP targets
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/IPP_targets_Mar_27_2019.txt"
Targets <- read_delim(path,delim = "\t") 
notexp <- Targets %>%  filter(IPP_C_Ab_1 == 0,IPP_C_Ab_2 == 0,IPP_C_Ab_3 == 0,
         IPP_N_Ab_1 == 0,IPP_N_Ab_2 == 0,IPP_N_Ab_3 == 0)
Targets <- Targets %>% filter(!entrez %in% notexp$entrez)
GENE_counts <- GENE_counts %>% filter(!mRNA_counts == 0)
Tar_mrna <- inner_join(GENE_counts, Targets, by = c("ensgene","symbol","biotype","description")) %>% 
  select(mRNA_counts,IPP_C_Ab_1,IPP_C_Ab_2,IPP_C_Ab_3,IPP_N_Ab_1,IPP_N_Ab_2,IPP_N_Ab_3,
         symbol) %>% 
  arrange(desc(mRNA_counts))



# 3.1  align fasta piRNA database                                           ----
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/pirBASE_clean/")
path_gtf <- "~/Downloads/gencode.v29.chr_patch_hapl_scaff.annotation.gtf.gz"
path <- "/home/rstudio/Documents/SPORTS_Db/Homo_sapiens1/piRBase/piR_human.fa"
pathg38 <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_index/hg38"
#align(index = pathg38, 
      path, 
      type = "rna",
      input_format = "FASTA",
      output_format="BAM",
      output_file = "./piRbase_unique_subread.BAM",
      maxMismatches=1,
      unique = TRUE,
      nBestLocations=1,
      nthreads=7,
      sortReadsByCoordinates = TRUE,
      useAnnotation = TRUE,
      annot.ext = path_gtf,
      isGTF = TRUE)
piRbase <- "piRbase_unique_subread.BAM"
piRbaseBam <- GenomicAlignments::readGAlignments(piRbase,
                param = Rsamtools::ScanBamParam(what = c("seq","qname")))
names(piRbaseBam) <- mcols(piRbaseBam)$qname 
GRpiRbase <- granges(piRbaseBam, use.mcols = TRUE)
GRpiRbasef <- mergeByOverlaps(GRpiRbase,GRpiRbase)
GRpiRbasef <- GRpiRbasef %>% as_tibble(rownames = "piRNA")
```

# Find complementarity between smallRNA and targets  enriched, not enriched 
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/")
library(rtracklayer);library(GenomicAlignments);library(tidyverse)
# 1.1 load enriched targets, predicted  and enriched piRNAs                 -------
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/mRNA_enriched_Frag_Apr_01_2019.txt"
# enriched targets, we need the names
mRNA_EN_targets <- read_delim(path,delim = "\t")
# predicted targets, we need the names
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/targets/all_IPP_genes_3C5__Mar_05_2019.txt"

Targets <- read_delim(path,delim = "\t")
inner_join(mRNA_EN_targets, Targets , by = c("gene_name" = "UTR3")) %>% 
  select(-noAb_v_INPUT,-UTR5,-CDS) %>% distinct(piRNA, .keep_all = TRUE)
inner_join(mRNA_EN_targets, Targets , by = c("gene_name" = "CDS")) %>% 
  select(-noAb_v_INPUT,-UTR3,-UTR5) %>% distinct(piRNA, .keep_all = TRUE)
inner_join(mRNA_EN_targets, Targets , by = c("gene_name" = "UTR5")) %>% 
  select(-noAb_v_INPUT,-UTR3,-CDS) %>% distinct(piRNA, .keep_all = TRUE)
# enriched piRNAs, we need the names
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
list.files(path)
C_enriched_RNA <- read_delim(str_glue("{path}C_enriched_RNA.txt"),delim = "\t")
N_enriched_RNA <- read_delim(str_glue("{path}N_enriched_RNA.txt"),delim = "\t")
enriched_piRNAs <- inner_join(C_enriched_RNA,N_enriched_RNA, by = "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  select(smallRNA, topIPP_C_vs_INPUT.x,topIPP_N_vs_INPUT.x,topIPP_INPUT_vs_noAb.x) %>% 
  separate(smallRNA, c("DB","sRNA"),sep = "_match_")
# 1.2 load read names of piRNAs / read names of targets                     ----   
# a)find the smallRNA read files
path_smallRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path = path_smallRNAs ,pattern = "_smallRNA", recursive = TRUE)
list_smallRNA <- list_smallRNA[str_detect(list_smallRNA,"COLO205_IPP")]

fil_name_sRNA <- str_remove(basename(list_smallRNA),"COLO205_IPP_") %>% 
  str_remove("_smallRNA.txt") 
fil_name_sRNA <- fil_name_sRNA[str_detect(fil_name_sRNA,"C_.+|N_.+")]

# b)find the smallRNA BAM files
path_BAMsmallRNAs <- "/home/rstudio/Documents/SPORTS_results/Result/fasta_files/IPP_PIWIL/"
list_BAMsmallRNA <- list.files(path = path_BAMsmallRNAs ,pattern = ".bam$")

# c)find the real target read files
path_real_targ_RNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/"
list_real_targRNA <- list.files(path = path_real_targ_RNAs ,pattern = ".featureCounts$")

# d)find the real target BAM files
path_BAMreal_targRNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_alignments/"
list_BAMreal_targRNA <- list.files(path = path_BAMreal_targRNAs ,pattern = ".BAM$") 
# 1.3 load read files and subset reads by names of enriched piRNA           ----
#annot table for later
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
library(data.table)
for (i in seq(along = fil_name_sRNA)) {
  
x_file <- fil_name_sRNA[i]# file 1

piRNA_enrch_reads <- fread(str_glue("{path_smallRNAs}{list_smallRNA[str_detect(list_smallRNA,x_file)]}")) %>% # change
  as_tibble() %>% 
  filter(V4 %in% enriched_piRNAs$sRNA) %>% 
  distinct(V1, .keep_all = TRUE) %>% 
  select(-RNA_type) %>% 
  mutate(width = str_length(V6)) %>% 
  unite("readNames",V1,V2) %>% 
  arrange(desc(width))
# load the bam smallRNA and subset it by read names
gal_smallRNA <- readGAlignments(str_glue("{path_BAMsmallRNAs}{list_BAMsmallRNA[str_detect(list_BAMsmallRNA,x_file)]}"),
                       param = ScanBamParam(what = c("seq","qname")))
names(gal_smallRNA) <- mcols(gal_smallRNA)$qname

gal_smallRNA <- gal_smallRNA[piRNA_enrch_reads$readNames]# subset
#setwd("enriched_BEDS/")
#export(gal_smallRNA,con = str_glue("{x_file}_enrichedPIRNA_{date}.bed"),format = "BED")

# load the bam of the long reads targets
gal_targetRNA <- readGAlignments(str_glue("{path_BAMreal_targRNAs}{list_BAMreal_targRNA[str_detect(list_BAMreal_targRNA,x_file)]}"),
                       param = ScanBamParam(what = c("seq","qname")))
names(gal_targetRNA) <- mcols(gal_targetRNA)$qname
# load the genes names from feature counts
targetGeneNames <- read_delim(list_real_targRNA[str_detect(list_real_targRNA,x_file)],
                              delim = "\t", col_names = FALSE, col_types = cols(
                                X1 = "c",
                                X2 = "f",
                                X3 = "d",
                                X4 = "c"
                              ) ) %>% 
  filter(X2 %in% c("Assigned","Unassigned_NoFeatures")) %>% 
  mutate(X5 = ifelse(is.na(X4),"no_annot",X4)) %>% 
  select(X1,X5) %>% 
  distinct(X1, .keep_all = TRUE) %>% 
  separate(X5,c("X5","a"), sep = ",") %>% 
  select(X1,X5)
#subset it
#subIPP <- targetGeneNames %>% filter(X4 %in% mRNA_EN_targets$GeneID)
#gal_targetRNA_IPP <- gal_targetRNA[subIPP$X1]


#path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
#fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
Annot_file <- targetGeneNames %>% 
  left_join(as_tibble(fc$annotation), by = c("X5" = "GeneID")) %>% 
  select(X1,gene_name,X5,gene_type,Length)
  

# seqlevels chr21
#gal_targetRNA_chrM <- gal_targetRNA[seqnames(gal_targetRNA) == 'chrM']

# 1.4 mergebyoverlaps 2 BAM files                                           ----
Comple_Targ <- mergeByOverlaps(gal_smallRNA, gal_targetRNA, 
                               minoverlap = 10L,
                               ignore.strand = TRUE) %>% as_tibble()

Comple_Targ <- Comple_Targ %>% select(
  starts_with("gal_"),
  #gal_smallRNA,gal_targetRNA_chr21,
  strandIPP = "strand",strandITarget = "strand.1",
  lengthIPP = "qwidth", lengthTarget = "width.1",
  startIPP = "start",   endIPP = "end",
  startT = "start.1", endT = "end.1",
  seqIPP = "seq", seqT = "seq.1",
  qname, qname.1,
  njunc, njunc.1
)
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/enriched_complment_files/"
piRNA_enrch_reads %>% 
dplyr::select(readNames, V4) %>% 
  inner_join(Comple_Targ, by = c("readNames" = "qname")) %>% 
  inner_join(Annot_file, by = c("qname.1" = "X1")) %>% 
  write_delim(str_glue("{path}{x_file}_complement_{date}.txt"),delim = "\t")

#  filter(strandIPP == "+",strandITarget == "-")
}
# 1.5 mergebyoverlaps 2 BAM files                                           ----

galName <- str_remove(basename(as.character(files[1])),"_subread.BAM") %>% 
  str_remove("^_COLO205_") 





files <- list.files(pattern = "subread.BAM.featureCounts", recursive = TRUE) 
rar1 <- read_delim(files[1],delim = "\t", col_names = FALSE) %>% 
  filter(X2 == "Assigned", X3 == 1) %>% 
  select(X1,X4) %>% 
  distinct(X1, .keep_all = TRUE)
rar1 <- rar1 %>% 
  filter(X4 %in% mRNA_EN_targets$GeneID) 
gal <- gal[rar1$X1]

# random change 1.3 lapply to find seq length                               ----
piRNa <- piRNAT_reads %>% 
  filter(V6 == "AGGGAGATGAAGAGGACAGTGACTGAGAGAC") %>% 
  group_by(V1) %>% mutate( width = str_length(V6)) %>% 
  arrange(desc(width)) %>% ungroup() 
pirnList<- lapply(as.list(piRNa$piRNA),function(x){
  piRNAT_reads %>% 
    filter(piRNA == x) %>% 
    group_by(V1) %>% 
    mutate( width = str_length(V6)) %>% 
    arrange(desc(width))
})

purepiRNA <- piRNAT_reads %>% 
  distinct(V6, .keep_all = TRUE) %>% 
  group_by(V1) %>% 
  mutate( width = str_length(V6)) %>% 
  arrange(desc(width)) %>% ungroup() %>% 
  distinct(piRNA, .keep_all = TRUE)

# random change 1.4 dunno                                                   ----
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_alignments/"

files <- list.files(path, pattern = ".BAM$")
files <- str_glue("{path}{files}")

gal <- readGAlignments(files[1],param = ScanBamParam(what = c("seq","qname")))
names(gal) <- mcols(gal)$qname
pos_gal <- readGAlignments(files[1],param = ScanBamParam(what = c("seq","qname"),
                           flag = scanBamFlag(isMinusStrand = FALSE)))
galName <- str_remove(basename(as.character(files[1])),"_subread.BAM") %>% 
  str_remove("^_COLO205_") 

path <- "~/Documents/SPORTS_results/Result/fasta_files/IPP_PIWIL/"
files_small <- list.files(path, pattern = "_sorted.bam$")
files_small <- str_glue("{path}{files_small}")
pos_small_gal <- readGAlignments(files_small[1],param = ScanBamParam(what = c("seq","qname"),
                           flag = scanBamFlag(isMinusStrand = FALSE)))

countOverlaps(pos_small_gal,pos_gal)
#reads_seq <- mcols(pos_gal)$seq %>% head %>% toString()

match_gen_piRNA <- mcols(piRbaseBam)$qname %>% tibble::enframe()
purepiRNA %>% filter(piRNA %in% match_gen_piRNA$value)
#write_rds(gal,"GAlignmentsList_IPP.rds")
#gal <- read_rds("GAlignmentsList_IPP.rds")
#seqinfo(bflist[[1]])
#readGAlignments(str_glue("{files[1]}"),use.names = TRUE) -> test

# 2.0 Not enriched piRNAs                                                   ----
# 2.1 load not enriched piRNAs                                              ----
# enriched piRNAs, we need the names
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
list.files(path)
not_enriched_piRNA <- read_tsv(str_glue("{path}notEnriched_pirna_001_Apr_23_2019.txt")) %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  separate(smallRNA, c("DB","sRNA"),sep = "_match_") %>% 
  select(sRNA)
# 2.2 load read names of piRNAs / read names of targets                     ----   
# a)find the smallRNA read files
path_smallRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path = path_smallRNAs ,pattern = "_smallRNA", recursive = TRUE)
list_smallRNA <- list_smallRNA[str_detect(list_smallRNA,"COLO205_IPP")]

fil_name_sRNA <- str_remove(basename(list_smallRNA),"COLO205_IPP_") %>% 
  str_remove("_smallRNA.txt") 

# b)find the smallRNA BAM files
path_BAMsmallRNAs <- "/home/rstudio/Documents/SPORTS_results/Result/fasta_files/IPP_PIWIL/"
list_BAMsmallRNA <- list.files(path = path_BAMsmallRNAs ,pattern = ".bam$")

# c)find the real target read files
path_real_targ_RNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/"
list_real_targRNA <- list.files(path = path_real_targ_RNAs ,pattern = ".featureCounts$")

# d)find the real target BAM files
path_BAMreal_targRNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_alignments/"
list_BAMreal_targRNA <- list.files(path = path_BAMreal_targRNAs ,pattern = ".BAM$") 
# 2.3 load read files and subset reads by names of enriched piRNA           ----
#annot table for later
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
library(data.table)
for (i in seq(along = fil_name_sRNA)) {
  
x_file <- fil_name_sRNA[i]# file 1

piRNA_no_enrch_reads <- fread(str_glue("{path_smallRNAs}{list_smallRNA[str_detect(list_smallRNA,x_file)]}")) %>% # change
  as_tibble() %>% 
  filter(V4 %in% not_enriched_piRNA$sRNA) %>% 
  distinct(V1, .keep_all = TRUE) %>% 
  select(-RNA_type) %>% 
  mutate(width = str_length(V6)) %>% 
  unite("readNames",V1,V2) %>% 
  arrange(desc(width))
# load the bam smallRNA and subset it by read names
gal_smallRNA <- readGAlignments(str_glue("{path_BAMsmallRNAs}{list_BAMsmallRNA[str_detect(list_BAMsmallRNA,x_file)]}"),
                       param = ScanBamParam(what = c("seq","qname")))
names(gal_smallRNA) <- mcols(gal_smallRNA)$qname

gal_smallRNA <- gal_smallRNA[piRNA_no_enrch_reads$readNames]# subset

# load the bam of the long reads targets
gal_targetRNA <- readGAlignments(str_glue("{path_BAMreal_targRNAs}{list_BAMreal_targRNA[str_detect(list_BAMreal_targRNA,x_file)]}"),
                       param = ScanBamParam(what = c("seq","qname")))
names(gal_targetRNA) <- mcols(gal_targetRNA)$qname
# load the genes names from feature counts
targetGeneNames <- read_delim(list_real_targRNA[str_detect(list_real_targRNA,x_file)],
                              delim = "\t", col_names = FALSE, col_types = cols(
                                X1 = "c",
                                X2 = "f",
                                X3 = "d",
                                X4 = "c"
                              ) ) %>% 
  filter(X2 %in% c("Assigned","Unassigned_NoFeatures")) %>% 
  mutate(X5 = ifelse(is.na(X4),"no_annot",X4)) %>% 
  select(X1,X5) %>% 
  distinct(X1, .keep_all = TRUE) %>% 
  separate(X5,c("X5","a"), sep = ",") %>% 
  select(X1,X5)

#path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
#fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
Annot_file <- targetGeneNames %>% 
  left_join(as_tibble(fc$annotation), by = c("X5" = "GeneID")) %>% 
  select(X1,gene_name,X5,gene_type,Length)
  
# 1.4 mergebyoverlaps 2 BAM files                                           ----
Comple_Targ <- mergeByOverlaps(gal_smallRNA, gal_targetRNA, 
                               minoverlap = 10L,
                               ignore.strand = TRUE) %>% as_tibble()

Comple_Targ <- Comple_Targ %>% select(
  starts_with("gal_"),
  #gal_smallRNA,gal_targetRNA_chr21,
  strandIPP = "strand",strandITarget = "strand.1",
  lengthIPP = "qwidth", lengthTarget = "width.1",
  startIPP = "start",   endIPP = "end",
  startT = "start.1", endT = "end.1",
  seqIPP = "seq", seqT = "seq.1",
  qname, qname.1,
  njunc, njunc.1
)
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/enriched_complment_files/"
piRNA_no_enrch_reads %>% 
dplyr::select(readNames, V4) %>% 
  inner_join(Comple_Targ, by = c("readNames" = "qname")) %>% 
  inner_join(Annot_file, by = c("qname.1" = "X1")) %>% 
  write_delim(str_glue("{path}{x_file}_no_enrch_complement_{date}.txt"),delim = "\t")

#  filter(strandIPP == "+",strandITarget == "-")
}
# 1.5 mergebyoverlaps 2 BAM files                                           ----

galName <- str_remove(basename(as.character(files[1])),"_subread.BAM") %>% 
  str_remove("^_COLO205_") 





files <- list.files(pattern = "subread.BAM.featureCounts", recursive = TRUE) 
rar1 <- read_delim(files[1],delim = "\t", col_names = FALSE) %>% 
  filter(X2 == "Assigned", X3 == 1) %>% 
  select(X1,X4) %>% 
  distinct(X1, .keep_all = TRUE)
rar1 <- rar1 %>% 
  filter(X4 %in% mRNA_EN_targets$GeneID) 
gal <- gal[rar1$X1]




```

# subset match_genome.fa for pirna reads only                               
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/blast_complement/")
library(rtracklayer);library(GenomicAlignments);library(tidyverse);library(data.table)
# enriched piRNAs, we need the names                                       
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
C_enriched_RNA <- read_delim(str_glue("{path}C_enriched_RNA.txt"),delim = "\t")
N_enriched_RNA <- read_delim(str_glue("{path}N_enriched_RNA.txt"),delim = "\t")
enriched_piRNAs <- inner_join(C_enriched_RNA,N_enriched_RNA, by = "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  select(smallRNA, topIPP_C_vs_INPUT.x,topIPP_N_vs_INPUT.x,topIPP_INPUT_vs_noAb.x) %>% 
  separate(smallRNA, c("DB","sRNA"),sep = "_match_") %>% 
  select(sRNA)
# a)find the smallRNA read files                                            ----
path_smallRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
list_smallRNA <- list.files(path = path_smallRNAs ,pattern = "_smallRNA", recursive = TRUE)
list_smallRNA <- list_smallRNA[str_detect(list_smallRNA,"COLO205_IPP")]

fil_name_sRNA <- str_remove(basename(list_smallRNA),"COLO205_IPP_") %>% 
  str_remove("_smallRNA.txt") 
#fil_name_sRNA <- fil_name_sRNA[str_detect(fil_name_sRNA,"C_.+|N_.+")]
# b)find the smallRNA BAM files                                             ----
path_BAMsmallRNAs <- "/home/rstudio/Documents/SPORTS_results/Result/fasta_files/IPP_PIWIL/"
list_BAMsmallRNA <- list.files(path = path_BAMsmallRNAs ,pattern = ".fasta")
# for loop to make the new .fa files                                        ----
# load files                                                                  
for (i in seq(along = fil_name_sRNA)) {
x_file <- fil_name_sRNA[i]# file 1

piRNA_enrch_reads <- fread(str_glue("{path_smallRNAs}{list_smallRNA[str_detect(list_smallRNA,x_file)]}")) %>% # change
  as_tibble() %>% 
  filter(V4 %in% enriched_piRNAs$sRNA) %>% 
  distinct(V1, .keep_all = TRUE) %>%
  unite("readNames",V1,V2) %>% 
  select(readNames,V4,V6)

fas <- DNAStringSet(piRNA_enrch_reads$V6) 
names(fas) <- piRNA_enrch_reads$readNames
# get the stringset and save to fasta                                       ----
writeXStringSet(fas, str_glue("{x_file}_enriched_subset_piRNA_reads.fasta"))
print(c(x_file,i))
}

# create dirs for dbs blast                                                 ----

filDB <- list.files(pattern = "long")
namedir <- str_replace(filDB,"_too_long_reads.fa", "_blastDB")
sapply(namedir,function(x)dir.create(x))

# make files with pirnas and gene annot from blast results                  ----
cname <- c("query_id", "subject_id", "subject_length", "alignment_length", 
           "q._start", "q._end", "s._start", "s._end", "query_seq",
           "subject_seq", "percent_identity", "identical", "mismatches", 
           "query_sbjct_frames", "subject_strand",
           "percent_query_coverage_per_subject","percent_query_coverage_per_hsp",
           "percent_query_coverage_per_uniq_subject")
files_bl <- list.files(pattern = "blast_.+_enriched") str_detect()
files_bl <- files_bl[!str_detect(files_bl,"not_en")]

path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))


# c)find the real target read files                                         ----
path_real_targ_RNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/"
list_real_targRNA <- list.files(path = path_real_targ_RNAs ,pattern = ".featureCounts$")

# d)find the real target BAM files                                          ----
path_BAMreal_targRNAs <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_alignments/"
list_BAMreal_targRNA <- list.files(path = path_BAMreal_targRNAs ,pattern = ".BAM$") 


# for loop to reannotate blast reads                                        ----
for (i in seq(along = fil_name_sRNA)) {
  
x_file <- fil_name_sRNA[i]
print(x_file)

bl <- read_delim(str_glue("{files_bl[str_detect(files_bl,x_file)]}"),
                 delim = "\t", col_names = cname,comment = "#")

# load the genes names from feature counts
targetGeneNames <- read_delim(str_glue("{path_real_targ_RNAs}{list_real_targRNA[str_detect(list_real_targRNA,x_file)]}"),
                              delim = "\t", col_names = FALSE, col_types = cols(
                                X1 = "c",
                                X2 = "f",
                                X3 = "d",
                                X4 = "c"
                              ) ) %>% 
  filter(X2 %in% c("Assigned","Unassigned_NoFeatures")) %>% 
  mutate(X5 = ifelse(is.na(X4),"no_annot",X4)) %>% 
  select(X1,X5) %>% 
  distinct(X1, .keep_all = TRUE) %>% 
  separate(X5,c("X5","a"), sep = ",") %>% 
  select(X1,X5)

Annot_file <- targetGeneNames %>% 
  left_join(as_tibble(fc$annotation), by = c("X5" = "GeneID")) %>% 
  select(X1,gene_name,X5,gene_type,Length)

piRNA_enrch_reads <- fread(str_glue("{path_smallRNAs}{list_smallRNA[str_detect(list_smallRNA,x_file)]}")) %>% # change
  as_tibble() %>% 
  filter(V4 %in% enriched_piRNAs$sRNA) %>% 
  distinct(V1, .keep_all = TRUE) %>%
  unite("readNames",V1,V2) %>% 
  select(readNames,V4,V6)

path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/enriched_complment_files/"
inner_join(bl,dplyr::select(piRNA_enrch_reads,readNames,V4),by = c("query_id"="readNames")) %>%
  left_join(Annot_file, by = c("subject_id" = "X1")) %>% 
  rename(piRNA= "V4", piRNA_read = "query_id", 
         long_read = "subject_id", gene_name = "gene_name", 
         ENSBL_Gene = "X5", gene_type = "gene_type") %>% 
  write_tsv(str_glue("{path}blast_enriched_{x_file}_{date}.txt"))
}

```


# bash bdtools and blastn                                                   
```{bash}
#bedtools
intersectBed -a '/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/enriched_BEDS/C_Ab_1_enrichedPIRNA_Apr_16_2019.bed'  -b  '/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/Rsubread_alignments/_COLO205_IPP_C_Ab_1_subread.BAM' -S
# makeblastdb
for file in *reads.fa; do  [ -f "$file" ] || continue; regex="${file%%_too_long_reads.fa}";  makeblastdb -in "${file}" -input_type "fasta" -dbtype "nucl" -hash_index -out "./${regex}_blastDB/${regex}_blastn.txt";done

#blastn
export PATH=$PATH:$HOME/Downloads/ncbi-blast-2.9.0+/bin 
cd /home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/blast_complement
for file in *reads.fasta; do  [ -f "$file" ] || continue; regex="${file%%_subset_piRNA_reads.fasta}"; str="${regex%%_enriched}";strh="${str%%_not}"; echo blastn -query  "${file}" -db  "./COLO205_IPP_${strh}_blastDB/COLO205_IPP_${strh}_blastn.txt" -num_threads 8 -out "blast_${regex}_piRNA.txt" -outfmt "7 qseqid sseqid  slen length qstart qend sstart send qseq sseq pident nident mismatch frames  sstrand qcovs qcovhsp qcovus"  -max_target_seqs 10    -subject_besthit -ungapped -word_size 6 -task "blastn-short" -perc_identity 85 -qcov_hsp_perc 60 ;printf "\n ${file} \t ${str} \n" ;done



#makeblastdb
makeblastdb -in input_file -input_type "fasta" -dbtype "nucl" -hash_index -out 


```

# check data from genomic ranges and blast                                  
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/enriched_complment_files/")
library(tidyverse);library(data.table)
list.files()
#dirc <- "my_bl_GR_results"
#dir.create(dirc)
# find files                                                                ----
#blast
blastfiles <- list.files(pattern = "blast")
enriched_bl <- blastfiles[!str_detect(blastfiles,"not_enriched")] 
not_enriched_bl <- blastfiles[str_detect(blastfiles,"not_enriched")]
#granges
GRfiles <- list.files(pattern = "complement")
enriched_GR <- GRfiles[!str_detect(GRfiles,"no_enrch")]
not_enriched_GR <- GRfiles[str_detect(GRfiles,"no_enrch")]
#names for loop
namfil <- enriched_bl %>% str_remove("blast_enriched_") %>% str_remove("_Apr.+")
for (i in seq(along = namfil)) {
  fil <- namfil[i]
# 1st load blast files                                                      ----
enriched_blast <- read_tsv(enriched_bl[str_detect(enriched_bl,fil)],
                           col_types = cols(
  .default = col_double(),
  piRNA_read = col_character(),
  long_read = col_character(),
  query_seq = col_character(),
  subject_seq = col_character(),
  query_sbjct_frames = col_character(),
  subject_strand = col_character(),
  piRNA = col_character(),
  gene_name = col_character(),
  ENSBL_Gene = col_character(),
  gene_type = col_character() )
  )
not_enriched_blast <- read_tsv(not_enriched_bl[str_detect(not_enriched_bl,fil)],
                               col_types = cols(
  .default = col_double(),
  piRNA_read = col_character(),
  long_read = col_character(),
  query_seq = col_character(),
  subject_seq = col_character(),
  query_sbjct_frames = col_character(),
  subject_strand = col_character(),
  piRNA = col_character(),
  gene_name = col_character(),
  ENSBL_Gene = col_character(),
  gene_type = col_character() )
  )

#enriched_blast %>% filter(long_read %in% not_enriched_blast$long_read) %>% 
#  distinct(gene_name)
inner_join(enriched_blast,not_enriched_blast,
           by ="long_read" ,suffix = c("_enrch", "_notenrch") )  %>% 
  write_tsv(str_glue("{dirc}/{fil}_bl_{date}.txt"))

# 2nd load blast  of Granges files                                          ----
enriched_GRang <- read_tsv(enriched_GR[str_detect(enriched_GR,fil)],
                           col_types = cols(
  .default = col_character(),
  lengthIPP = col_double(),
  lengthTarget = col_double(),
  startIPP = col_double(),
  endIPP = col_double(),
  startT = col_double(),
  endT = col_double(),
  njunc = col_double(),
  njunc.1 = col_double(),
  gene_name = col_character(),
  gene_type = col_character(),
  Length = col_character()
))
not_enriched_GRang <- read_tsv(not_enriched_GR[str_detect(not_enriched_GR,fil)],
                               col_types = cols(
  .default = col_character(),
  lengthIPP = col_double(),
  lengthTarget = col_double(),
  startIPP = col_double(),
  endIPP = col_double(),
  startT = col_double(),
  endT = col_double(),
  njunc = col_double(),
  njunc.1 = col_double(),
  gene_name = col_character(),
  gene_type = col_character(),
  Length = col_character()
))
#enriched_GRang %>% filter(qname.1 %in% not_enriched_GRang$qname.1)
inner_join(enriched_GRang,not_enriched_GRang,
           by ="qname.1" ,suffix = c("_enrch", "_notenrch") )  %>% 
  write_tsv(str_glue("{dirc}/{fil}_GR_{date}.txt"))
# 3rd check blast vs granges                                                ----
#enriched_blast %>% filter(piRNA_read %in% enriched_GRang$readNames) %>% distinct(gene_name)
inner_join(enriched_blast,enriched_GRang, by = c("long_read" = "qname.1"),
           suffix = c("_blast", "_GR") ) %>% 
  write_tsv(str_glue("{dirc}/{fil}_blGR_{date}.txt"))
print(fil)
}
# searching for common not enriched gene targets                            ----
bl_en <- list.files(pattern = "blast_enriched")
bl_not_en <- list.files(path = str_glue("{dirc}"), 
pattern = "_bl_.+_26")

namfile <- str_remove(bl_not_en,"_bl.+") 
namfile <- namfile[str_detect(namfile,"C_.+|N_.+")]   
# initialize list
my_list <- list()
for (i in seq(along = namfile)) {
fil <- namfile[i]

rsls_blast_exlu <- read_tsv(str_glue("{dirc}/{bl_not_en[str_detect(bl_not_en,fil)]}"),
                            col_types = cols(
  .default = col_double(),
  piRNA_read_enrch = col_character(),
  long_read = col_character(),
  query_seq_enrch = col_character(),
  subject_seq_enrch = col_character(),
  query_sbjct_frames_enrch = col_character(),
  subject_strand_enrch = col_character(),
  piRNA_enrch = col_character(),
  gene_name_enrch = col_character(),
  ENSBL_Gene_enrch = col_character(),
  gene_type_enrch = col_character(),
  piRNA_read_notenrch = col_character(),
  query_seq_notenrch = col_character(),
  subject_seq_notenrch = col_character(),
  query_sbjct_frames_notenrch = col_character(),
  subject_strand_notenrch = col_character(),
  piRNA_notenrch = col_character(),
  gene_name_notenrch = col_character(),
  ENSBL_Gene_notenrch = col_character(),
  gene_type_notenrch = col_character()
)) %>% #filter(subject_strand_enrch == "minus" | subject_strand_notenrch == "minus") %>%  
  distinct(gene_name_notenrch)

rsls_blast <- read_tsv(bl_en[str_detect(bl_en,fil)],
                       col_types = cols(
  .default = col_double(),
  piRNA_read = col_character(),
  long_read = col_character(),
  query_seq = col_character(),
  subject_seq = col_character(),
  query_sbjct_frames = col_character(),
  subject_strand = col_character(),
  piRNA = col_character(),
  gene_name = col_character(),
  ENSBL_Gene = col_character(),
  gene_type = col_character()
)                       ) %>% 
  filter(subject_strand == "minus", !gene_name %in% rsls_blast_exlu$gene_name_notenrch) %>% 
  distinct(gene_name, .keep_all = TRUE) %>% 
  select(gene_name, gene_type, subject_length, piRNA_read, piRNA)
my_list[[fil]] <- rsls_blast
}

finaltarg <- bind_rows(my_list,.id = "sample") %>% 
  arrange(desc(subject_length)) %>% 
  distinct(gene_name, .keep_all = TRUE) %>% 
  filter(!is.na(gene_name)) %>% 
  write_tsv(str_glue("final_target_Genes_{date}.txt"))

finaltarg %>% 
  inner_join(Targets,by = c("gene_name" = "UTR3")) %>%
  distinct(gene_name, .keep_all = TRUE) %>% 
   write_tsv(str_glue("final_target_plus_pred_UTR3_Genes_{date}.txt"))

finaltarg %>%
  inner_join(Targets,by = c("gene_name" = "CDS")) %>%
  distinct(gene_name, .keep_all = TRUE) %>% 
   write_tsv(str_glue("final_target_plus_pred_CDS_Genes_{date}.txt"))

finaltarg %>%
  inner_join(Targets,by = c("gene_name" = "UTR5")) %>%
  distinct(gene_name, .keep_all = TRUE) %>% 
   write_tsv(str_glue("final_target_plus_pred_UTR5_Genes_{date}.txt"))
  
# search for precursors ---- 

(GR <- list.files(pattern = "_complement_"))
(GR <- GR[str_detect(GR,"C_.+|N_.+")])
(GR_no_en_fls <- GR[str_detect(GR,"no_enrch")])
(GR_en_fls <- GR[!str_detect(GR,"no_enrch")])

path_smallRNAs <- "/home/rstudio/Documents/new_analysis_workflow/1_5_files_withDBs/"
# enriched piRNAs, we need the names
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
list.files(path)
C_enriched_RNA <- read_delim(str_glue("{path}C_enriched_RNA.txt"),delim = "\t")
N_enriched_RNA <- read_delim(str_glue("{path}N_enriched_RNA.txt"),delim = "\t")
enriched_piRNAs <- inner_join(C_enriched_RNA,N_enriched_RNA, by = "smallRNA") %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  select(smallRNA, topIPP_C_vs_INPUT.x,topIPP_N_vs_INPUT.x,topIPP_INPUT_vs_noAb.x) %>% 
  separate(smallRNA, c("DB","sRNA"),sep = "_match_") %>% 
   select(sRNA)
# NOTenriched piRNAs, we need the names
path <- "/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/"
list.files(path)
not_enriched_piRNA <- read_tsv(str_glue("{path}notEnriched_pirna_001_Apr_23_2019.txt")) %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  separate(smallRNA, c("DB","sRNA"),sep = "_match_") %>% 
  select(sRNA)

stat_tib <- tribble(~values,
                    "enriched_piRNA_sample",
                    "enriched_pre_piRNA_sample",
                    "enr_percent",
                    "not_enriched_piRNA_sample",
                    "not_enriched_pre_piRNA_sample",
                    "not_enr_percent"
                    )
for (i in seq(along = namfile)) {
fil <- namfile[i]

en_GR <- read_tsv(GR_en_fls[str_detect(GR_en_fls,fil)], col_types =cols(
  .default = col_character(),
  lengthIPP = col_double(),
  lengthTarget = col_double(),
  startIPP = col_double(),
  endIPP = col_double(),
  startT = col_double(),
  endT = col_double(),
  njunc = col_double(),
  njunc.1 = col_double(),
  Length = col_double()
)  )
reads_en_GR <- en_GR %>% 
  filter((strandIPP == "-" & strandITarget == "-") | (strandIPP == "+" & strandITarget == "+") ) %>% 
   distinct(readNames)  %>% 
  separate(readNames, c("read", "count"), sep = "_", convert = TRUE)
 complete_data <- fread(str_glue("{path_smallRNAs}{list_smallRNA[str_detect(list_smallRNA,fil)]}")) %>% as_tibble() %>% 
filter(str_detect(RNA_type,"piR"))   

en_piRNA_in_dset <- complete_data %>% 
  filter(V4 %in% enriched_piRNAs$sRNA) %>% 
  distinct(V4) %>% nrow()

en_prePIRNA_in_dset <- complete_data %>% 
  filter(V4 %in% enriched_piRNAs$sRNA) %>%
  filter(V1 %in% reads_en_GR$read) %>% 
  distinct(V4) %>% nrow()

n_en_GR <- read_tsv(GR_no_en_fls[str_detect(GR_no_en_fls,fil)], col_types =cols(
  .default = col_character(),
  lengthIPP = col_double(),
  lengthTarget = col_double(),
  startIPP = col_double(),
  endIPP = col_double(),
  startT = col_double(),
  endT = col_double(),
  njunc = col_double(),
  njunc.1 = col_double(),
  Length = col_double()
) )

rs_n_en_GR <- n_en_GR %>% 
  filter((strandIPP == "-" & strandITarget == "-") | (strandIPP == "+" & strandITarget == "+") ) %>% 
  distinct(readNames)  %>% 
  separate(readNames, c("read", "count"), sep = "_", convert = TRUE)

NOTen_piRNA_in_dset <- complete_data %>% 
  filter(V4 %in% not_enriched_piRNA$sRNA) %>% 
  distinct(V4) %>% nrow()

NOTen_prePIRNA_in_dset <- complete_data %>% 
  filter(V4 %in% not_enriched_piRNA$sRNA) %>%
  filter(V1 %in% rs_n_en_GR$read) %>% 
  distinct(V4) %>% nrow()

stat_tib[,fil] <- c(en_piRNA_in_dset, en_prePIRNA_in_dset,
                    round(en_prePIRNA_in_dset/en_piRNA_in_dset,2),
                    NOTen_piRNA_in_dset, NOTen_prePIRNA_in_dset,
                    round(NOTen_prePIRNA_in_dset/NOTen_piRNA_in_dset,2))
}
stat_tib %>% write_tsv("precursors_stricter_results.txt")
#cross tables of predicted with all longreads -----
#load predicted gene targets
path <- "~/Downloads/all_IPPgenes_Mar_06_2019.txt"
Targets <- read_tsv(path)
#load the counts of longreads
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
(long_counts <- fc$counts %>% as_tibble(rownames = "GeneID"))

(long_annot <- fc$annotation %>% 
    as_tibble() %>% 
    arrange(Chr) %>% 
    distinct(gene_name, .keep_all = TRUE) %>% 
    select(GeneID,gene_type,gene_name))

(long_counts <- long_counts %>% 
  inner_join(long_annot,"GeneID") %>% select(gene_name,everything()))

PredTarg_RealTarg <- long_counts %>% 
  filter(gene_name %in% Targets$external_gene_name) %>% 
  write_tsv(str_glue("PredTarg_RealTarg_{date}.txt"))

#load mRNA reads
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/mRNA_COLO205/mRNA_cpm_counts_COLO205_Apr_27_2019.txt"

mRNACOLO <- read_tsv(path)  %>%  
  filter(counts > 0,gene_type == "protein_coding") %>%
  arrange(Chr) %>% 
  distinct(gene_name, .keep_all = TRUE) %>% 
  select(counts,gene_name)

Cab <- PredTarg_RealTarg  %>% 
  filter(IPP_C_Ab_1>0,IPP_C_Ab_2>0,IPP_C_Ab_3>0,
         IPP_N_Ab_1>0,IPP_N_Ab_2>0,IPP_N_Ab_3>0, 
         gene_type == "protein_coding") %>% 
  inner_join(mRNACOLO,"gene_name") %>% 
  select(gene_name,counts,gene_type,GeneID,everything()) %>% 
  write_tsv(str_glue("mRNA_PredTarg_{date}.txt"))
  
# intersect between targets and methylated ------
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/2nd_treatment/targets/all_IPP_genes_3C5__Mar_05_2019.txt"

Targets <- read_delim(path,delim = "\t") %>% 
  select(-EnsID) %>% 
  gather(key = "region", value = "gene_name", CDS,UTR3,UTR5) %>% 
  filter(!is.na(gene_name))
  
meth_ip <- readxl::read_excel("~/Downloads/Comparison methylated and IP piRNAs_30_04_19.xlsx")

Targets <- Targets %>% 
  mutate(methylation = ifelse(piRNA %in% meth_ip$`24 methylated`,"methylated", 
                              ifelse(piRNA %in% meth_ip$`102 not methylated`,"not_methyl","not_identified"))) 

Cab %>% select(gene_name, counts,gene_type, GeneID) %>% 
  inner_join(Targets) %>% 
  write_tsv(str_glue("Targets_Meth_3C5_{date}.txt"))

```

# correlation of IPP                        
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(tidyverse);library(edgeR)
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/correlation_ip/")
# find the  longreads cpm ------
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/"
fc <- read_rds(str_glue("{path}feature_counts_Mar_29_2019.rds"))
fc$annotation <- fc$annotation %>% 
  as_tibble() %>% 
  arrange(Chr) %>% 
  distinct(gene_name,.keep_all = TRUE)

fc$counts <- fc$counts[fc$annotation$GeneID,] 

batch <- str_replace(colnames(fc$counts),"IPP_.+_.+_","") %>% 
  str_replace(".+_","") %>% as_factor()
group <- str_replace(colnames(fc$counts),"_[:digit:]$","")  %>%
  str_replace("_Control_","_Ctrl_") %>% 
  str_replace_all("IPP_|_noAb","") %>% 
  as_factor()
samples <- data.frame(group,batch,row.names = colnames(fc$counts))

x <- DGEList(counts = fc$counts,samples = samples,
             lib.size = colSums(fc$counts),
             norm.factors = rep(1,ncol(fc$counts)))

design <- model.matrix(~0+group+batch)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
#   filtering                                                            ----
keep.exprs <- filterByExpr.DGEList(x,design = design)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)

names(long_reads) <- str_glue("{names(long_reads)}_lR") %>% 
  str_remove_all("IPP_|_noAb") %>% 
  str_replace("Control","Ctrl")
# normalization ------
x1_TMM <- calcNormFactors(x1, method = "TMM")
cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
# make the longreads  dataset -----
long_reads_ds <- cpm_TMM %>% 
  as_tibble(rownames = "GeneID")

names(long_reads_ds)[2:length(long_reads_ds)] <- str_glue("{names(long_reads_ds)[2:length(long_reads_ds)]}_lR") %>% 
  str_remove_all("IPP_|_noAb") %>% 
  str_replace("Control","Ctrl")  

long_reads_ds <- long_reads_ds %>%   
  inner_join(fc$annotation)

# load the targets ----
path <- "~/Downloads/all_IPPgenes_Mar_06_2019.txt"
targets <- read_tsv(path)

# subset longreads with targets -----
long_reads_ds <- long_reads_ds %>% 
  filter(gene_name %in% targets$external_gene_name)

# load piRNA cpm IPP ---
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/3_IPP_PIWIL1_all_DBs/IPP_PIWIL_TMM_logcpm_counts.txt"

piRNA_cpm <- read_tsv(path) %>% 
  filter(str_detect(smallRNA,"piR")) %>% 
  separate(smallRNA, c("small","RNA"), sep = "_match_") %>% 
  select(-small)

names(piRNA_cpm) <- str_glue("{names(piRNA_cpm)}_pi") %>% 
  str_remove_all("COLO205_IPP_|_noAb") %>% 
  str_replace("Control","Ctrl")
#load 279/317 pirna ----
path <- "~/Downloads/all_IPP_genes_3C5__Mar_05_2019.txt"
subpirna <- read_tsv(path) 
subpirna <- subpirna %>% 
  select(-EnsID) %>% 
  gather(key = "region", value = "gene_name", CDS,UTR3,UTR5) %>% filter(!is.na(gene_name))

# subset piRNA -----
piRNA %>% filter(RNA %in% subpirna$piRNA)

# make the dataset   ------
pDT <- subpirna %>%
  inner_join(piRNA_cpm, by = c("piRNA" = "RNA_pi")) %>% 
  inner_join(long_reads_ds, by = c("gene_name" = "gene_name"))

# time to do the plots -----
library(ggpmisc);library(gridExtra)
myformula <- y~x
dat %>% column_to_rownames(var="smallRNA")->dat

cor.test(pDT$Ctrl_3_pi,pDT$Ctrl_3_lR, 
         method = "spearman",conf.level = 0.95)

scatter_plot <- ggplot(pDT, aes(Ctrl_3_pi,Ctrl_3_lR))+ 
  geom_smooth(method = "lm",formula = myformula)+
  stat_poly_eq(formula = myformula,
               aes(label=paste(..eq.label..,..rr.label..,sep="~~~")),
               parse=TRUE) + geom_point()
scatter_plot+labs(title = "mitochondrial tRNAs,rRNAs")
plot_C_N <- scatter_plot+labs(title="mitochondrial tRNAs,rRNAs log2 cpm",x="Cytosol",y="Nucleus")


grid.arrange(plot_C_N,plot_C_N_treat,plot_C_T,plot_N_T,nrow=2,ncol=2)

```


# make tables for pub
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(Rsubread);library(tidyverse);library(data.table)
setwd("~/Documents/new_analysis_workflow/tables_journal/")
#find the names of celllines
path <- "/media/sf_R_virtual_shared_folder/1_Cell_lines_01062016_NextSeq/"
cellines <- list.files(path) %>% str_remove(".fastq")
# find the summaries
path <- "~/Documents/SPORTS_results/"
filesnames <- list.files(path,pattern = "_summary.txt", recursive = TRUE) %>% 
  enframe() %>% 
  select(value) %>%  
 filter(!str_detect(value,"noDBs|furlan|old|Next_Seq|_pcmv6|_Piwi|summary_result"))

#subset for the experiments
Cell_files <- filesnames %>%
  filter(!str_detect(value,"Treated|treated|_IPP|_Piwil1_|CITOSO|NUC|TOT"))

Testis_files <- filesnames %>% 
  filter(str_detect(value,"testis"),
         str_detect(value,"3ug_1|Non_"),
         !str_detect(value,"_tech_2")) 

Treat_files <- filesnames %>% 
  filter(str_detect(value,"CITOSO|NUC|TOT"))

IPP_files <- filesnames %>% 
  filter(str_detect(value,"_IPP"))

#load them and calculate
#IPP_files
DTIPP_files <- rbindlist(sapply(str_glue("{path}{IPP_files$value}"),fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>% 
  as_tibble() %>% mutate(FileName = str_remove(basename(FileName),"_summary.txt")) %>% 
  select(-Sub_Class) %>% filter(!Class =="-",!str_detect(Class,"Unmatch|5_end")) %>% 
  separate(Class, c("Database", "smallRNA"), sep = "-|_M") %>% 
  mutate(smallRNA = ifelse(is.na(smallRNA),"-",smallRNA),
         Database = as_factor(Database))

DTIPP_files %>% 
  group_by(FileName,Database) %>% 
   summarise(read_sum=sum(Reads)) %>%
  ungroup() %>% 
  spread(FileName,read_sum) %>% 
  rename(Annotation = "Database") %>% 
  write_tsv("IPP_files_summary.txt")
#Treat_files
DTTreat_files <- rbindlist(sapply(str_glue("{path}{Treat_files$value}"),fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>% 
  as_tibble() %>% mutate(FileName = str_remove(basename(FileName),"_summary.txt")) %>% 
  select(-Sub_Class) %>% filter(!Class =="-",!str_detect(Class,"Unmatch|5_end")) %>% 
  separate(Class, c("Database", "smallRNA"), sep = "-|_M") %>% 
  mutate(smallRNA = ifelse(is.na(smallRNA),"-",smallRNA),
         Database = as_factor(Database))

DTTreat_files %>% 
  group_by(FileName,Database) %>% 
   summarise(read_sum=sum(Reads)) %>%
  ungroup() %>% 
  spread(FileName,read_sum) %>% 
  rename(Annotation = "Database") %>% 
  write_tsv("Treat_files_summary.txt")
#Testis_Cell_lines
DTTestis_cell <- rbindlist(sapply(str_glue("{path}{Testis_files$value}"),fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>% 
  as_tibble() %>% mutate(FileName = str_remove(basename(FileName),"_summary.txt")) %>% 
  select(-Sub_Class) %>% filter(!Class =="-",!str_detect(Class,"Unmatch|5_end")) %>% 
  separate(Class, c("Database", "smallRNA"), sep = "-|_M") %>% 
  mutate(smallRNA = ifelse(is.na(smallRNA),"-",smallRNA),
         Database = as_factor(Database))

DTTestis_cell %>% 
  group_by(FileName,Database) %>% 
   summarise(read_sum=sum(Reads)) %>%
  ungroup() %>% 
  spread(FileName,read_sum) %>% 
  rename(Annotation = "Database") %>% 
  write_tsv("Testis_Cell_lines_files_summary.txt")

DTT_cell <- rbindlist(sapply(str_glue("{path}{Cell_files$value}"),fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>% 
  as_tibble() %>% mutate(FileName = str_remove(basename(FileName),"_summary.txt")) %>% 
  select(-Sub_Class) %>% filter(!Class =="-",!str_detect(Class,"Unmatch|5_end")) %>% 
  separate(Class, c("Database", "smallRNA"), sep = "-|_M") %>% 
  mutate(smallRNA = ifelse(is.na(smallRNA),"-",smallRNA),
         Database = as_factor(Database))

DTT_cell %>% 
  group_by(FileName,Database) %>% 
   summarise(read_sum=sum(Reads)) %>%
  ungroup() %>% 
  spread(FileName,read_sum) %>% 
  rename(Annotation = "Database") %>% 
  write_tsv("Cell_lines_files_summary.txt")


#find the complete files with all reads-----
path_comple <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/"
files_complete <- list.files(path_comple, pattern = "smallRNA",recursive = TRUE) %>% 
  enframe() %>% 
  select(value) %>%  
 filter(!str_detect(value,"HCT116_Treated|furlan|Next_Seq"))

#subset the files
Cell_dat <- files_complete %>% 
  filter(str_detect(value,"Cell_Lines_"))
Testis_dat <- files_complete %>% 
  filter(str_detect(value,"Testis"))

IPP_dat <- files_complete %>% 
  filter(str_detect(value,"_IPP"))
Treat_dat <- files_complete %>% 
  filter(str_detect(value,"Periodate"))

DT <- fread(str_glue("{path_comple}{IPP_dat$value[1]}"),header = TRUE) 
DT <- DT %>% as_tibble %>%  unite("smallRNA",RNA_type,V4,sep = "_")

DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.( V6)][, .(counts=sum(readcount)), by=.(smallRNA)] %>%
  as_tibble() %>% 
  separate(smallRNA, c("Database", "smallRNA"), sep = "_match_") 

DT_last %>% group_by(Database) %>% 
  summarise(read_sum = sum(counts)) %>% 
  filter(!Database == "tRNA_mature")

#make logos testis colo205 -----
files <- c(Testis_files$value[2],Cell_files$value[2]) %>% as.list()
# heatmap celllines_testis------
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/Cell_lines_Testis_exp/")
 mat <- read_tsv("common_piRNA_Cell_Testis_1cpm_norm.txt") %>% column_to_rownames("smallRNA") %>% as.matrix()
 scaled_mat <- t(scale(t(as.matrix(mat))))
hist(scaled_mat)
f <- colorRamp2(c(-1.5,0,1.5),c("blue","white","yellow"))
df <- data.frame(cell_line = colnames(scaled_mat))
ha <- HeatmapAnnotation(df=df)
Heatmap(scaled_mat,
        
        col = f,
        clustering_distance_rows = "spearman",
        clustering_method_rows="ward.D2",
        cluster_columns = FALSE,
        show_row_names = FALSE,
        name = "z-score equivalent",
        top_annotation = ha,
        column_title = str_glue("{nrow(mat)} piRNAs common"),
        column_title_side = "top")


```





# Align match_genome reads subread                                          
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
library(Rsubread);library(tidyverse)
setwd("~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/")
# 1.2  Find long_read files and align them                                  ----
# dir.create("Rsubread_alignments")
path <- "~/Documents/SPORTS_results/Result/"
name_file <- "COLO205_IPP"
list_match <- list.files(path,pattern = "_match_genome.fa", recursive = TRUE)
list_matchIPP <- list_match[str_detect(list_match,name_file)] 
list_matchIPP <- list_matchIPP[!str_detect(list_matchIPP,"unmatch")]
list_matchIPP <- list_matchIPP[!str_detect(list_matchIPP,"_match_.+_match_")]
list_matchIPP <- list_matchIPP[13:24]
names_matchIPP <- str_remove(basename(list_matchIPP),"_match_genome.fa")
list_matchIPP <- sapply(list_matchIPP, function(x)str_glue({path},{x}),simplify=FALSE)
names(names_matchIPP) <- names(list_matchIPP)
# align them
path_gtf <- "~/Downloads/gencode.v29.chr_patch_hapl_scaff.annotation.gtf.gz"
lapply(names(list_matchIPP), function(x)
align(index = "./Rsubread_index/hg38",
      list_matchIPP[x], 
      type = "rna",
      input_format = "FASTA",
      output_format="BAM",
      output_file = paste("./match_genome_bam_smallReads_alignments/",
                          names_matchIPP[x],"subread.BAM",sep="_"),
      nBestLocations = 5,
      nsubreads = 3,
      nthreads=5,
      sortReadsByCoordinates = TRUE,
      useAnnotation = TRUE,
      annot.ext = path_gtf,
      isGTF = TRUE))

list.BAM <- list.files(pattern = ".BAM$", recursive = TRUE)
path <- "~/Documents/new_analysis_workflow/results_R/results_with_DBs/IPP_Rsubread_mRNA/complementarity/"
#list.BAM <- sapply(list.BAM, function(x)str_glue("{path}Rsubread_alignments/{x}"),simplify=FALSE)
fc <- featureCounts(files = unlist(list.BAM),
                    annot.ext =  path_gtf,
                    GTF.attrType.extra = c("gene_type", "gene_name"),
                    isGTFAnnotationFile = TRUE,
                    nthreads = 7,
                    allowMultiOverlap = TRUE,
                    fraction = TRUE,
                    reportReads = "CORE",
                    reportReadsPath = path,
                    strandSpecific = 0)

colnames(fc$counts) <- str_remove(colnames(fc$counts),".+_COLO205_") %>% 
  str_remove("_subread.BAM")
colnames(fc$stat) <- str_remove(colnames(fc$stat),".+_COLO205_") %>% 
  str_remove("_subread.BAM")
fc$targets <- str_remove(fc$targets,".+_COLO205_") %>% 
  str_remove("_subread.BAM")
# fc %>% write_rds(str_glue("feature_counts_{date}.rds"))
# fc <- read_rds(str_glue("feature_counts_Mar_29_2019.rds"))
cname <- c("query id", "subject id", "subject length", "alignment length", "q. start", "q. end", "s. start", "s. end", "query seq", "subject seq", "% identity", "identical", "mismatches", "query/sbjct frames", "subject strand")
```







Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


# furlan dataset                                                            -----
```{r}
date <- gsub(" ","_",format(Sys.time(), "%b %d %Y"))
name_file <- "furlan"
library(data.table);library(tidyverse);library(rafalib);library(edgeR);library(wesanderson)
setwd("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/furlan/")
# 1.1 List files merged and outputs of all ----------
# first select the correct tables
data_filt <- c("C17|C28|C29|C31|C39|C40|C44|C45|C13_A|C33|C35-|C37|C7|C4-|C25|C30|C3_A|C5|C6|C9")
path <- "~/Documents/new_analysis_workflow/1_5_files_withDBs/furlan/"

list_smallRNA <- list.files(path,pattern = "_smallRNA", recursive = TRUE)

# find the file about the experiments CITOSOL|NUCLEO
listRNA <- list_smallRNA[str_detect(list_smallRNA,data_filt)]

#load data to data.table
listRNA <- sapply(listRNA, function(x)str_glue({path},{x}),simplify=FALSE)
DTRNA <- rbindlist(sapply(listRNA,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName")

# 1.2 using data table------------------
# merge column of smallRNA and RNA_type to have info for files of DBs
DT <- DTRNA %>%
  as_tibble %>%  
  unite("smallRNA",RNA_type,V4,sep = "_")

# count smallRNAs from reads
DT_last <- setDT(DT)[, V2 := as.numeric(V2)][, readcount := V2 / .N, by=.(FileName, V6)][, .(counts=sum(readcount)), by=.(FileName,smallRNA)] %>% as_tibble()

# check sums if the same as output summaries
newSUms <- DT_last %>% 
  group_by(FileName) %>% 
  summarise(sums=sum(counts)) %>% 
  arrange(sums)

path <- "/home/rstudio/Documents/SPORTS_results/Result_furlan_with_DBs/" # CHANGE this
outfiles <- list.files(path,pattern = "_summary.txt",recursive = TRUE ) # CHANGE this
outfiles <- outfiles[str_detect(outfiles,data_filt)]
outfiles <- sapply(outfiles, function(x)str_glue({path},{x}),simplify=FALSE)
sumsDT <- rbindlist(sapply(outfiles,fread,simplify=FALSE), 
                 use.names= TRUE,idcol="FileName") %>%
  as_tibble() %>% 
  filter(Class=="Match_Genome") %>% 
  select(FileName,Reads) %>% arrange(Reads)

table(near(as_vector(sumsDT$Reads),as_vector(round(newSUms$sums))))
sumsDT$Reads
newSUms$sums

# from long to wide table
setDT(DT_last)

DT_last <- dcast.data.table(DT_last,smallRNA ~ FileName,drop = FALSE,fill = 0,value.var ="counts")

# change the colnames
gsub("_smallRNA.txt","",basename(colnames(DT_last))) -> colnames(DT_last)
colnames(DT_last)

# check the sum names
sum(DT_last[,C3_A_ATGTCA_L008_R1_001])

path <- str_glue("/home/rstudio/Documents/new_analysis_workflow/results_R/results_with_DBs/furlan/{name_file}_raw_counts_{date}.txt") # CHANGE THIS^^^^^^^^^^^^^^^^^^

fwrite(DT_last,file = path,sep = "\t",col.names = TRUE)

# 1.3 load  data ------
path <- "furlan_raw_counts_Apr_24_2019.txt"
DT2 <- as_tibble(fread(path ,header = TRUE)) 
fnams <- colnames(DT2)[-1]

colnames(DT2)[2:length(colnames(DT2))] <- str_remove(fnams,"_.+|-.+")

normal <- c("C17","C28","C29","C31")
MSI_L <- c("C39","C40","C44","C45")
MSI_Sp <- c("C13","C33","C35","C37")
MSS <- c("C7","C4","C25","C30","C3","C5","C6","C9")
mat <- DT2 %>% select(smallRNA,normal,MSI_L,MSI_Sp,MSS) %>% 
  column_to_rownames("smallRNA") %>% 
  as.matrix()

group <- as_factor(c(rep("normal",4),rep("cancer",16)))
batch <- as_factor(1:20)
typeMS <- as_factor(c(rep("normal",4), rep("MSI",8),rep("MSS",8)))
type <- as_factor(c(rep("normal",4), rep("lynch",4),
                    rep("sporadic",4),rep("MSS",8)))
colors_group <- wes_palettes$Royal1[1:length(levels(group))]
c_group <- c(rep(colors_group[1],4),rep(colors_group[2],16))
colors_type <- wes_palettes$Royal1[1:length(levels(type))]
c_type <- c(rep(colors_type[1],4),rep(colors_type[2],4),
            rep(colors_type[3],4),rep(colors_type[4],8))
samples <- data.frame(group, type, row.names = colnames(mat))

x <- DGEList(counts = mat, 
             samples = samples, 
             lib.size = colSums(mat),
             norm.factors = rep(1,ncol(mat)))

design <- model.matrix(~group)
colnames(design) <- gsub("group","",colnames(design))
rownames(design) <- rownames(samples)
design_1 <- model.matrix(~0 + type)
colnames(design_1) <- gsub("type","",colnames(design_1))
rownames(design_1) <- rownames(samples)

# 2.1 filtering ------------------------------------
cutoff <- cpm(10,mean(x$samples$lib.size))
cpm  <- cpm(x)
lcpm <- cpm(x,log=TRUE,prior.count=5)
L <- mean(x$samples$lib.size)*1e-6
M <- median(x$samples$lib.size)*1e-6
c(L,M)
log2(5/L)
log2(cutoff)

keep.exprs <- filterByExpr.DGEList(x,type)

x1 <- x[keep.exprs,,keep.lib.sizes=FALSE]

keep.exprs <- rowSums(cpm > as.double(cutoff)) >= 10

x2 <- x[keep.exprs,,keep.lib.sizes=FALSE]

dim(x)
dim(x1)
dim(x2)
# 2.2.1 density plot-------------------------  
lcpm.cutoff <- log2(10/M + 2/L)
pal1 <- c(wes_palettes$GrandBudapest1,
          wes_palettes$GrandBudapest2,
          wes_palettes$Moonrise1)
par(mfrow=c(1,2))
samplenames <- colnames(x$counts)
nsamples <- ncol(x)

plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)

for (i in 2:nsamples){
 den <- density(lcpm[,i])
 lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")

lcpm <- cpm(x1, log=TRUE, prior.count=5)
plot(density(lcpm[,1]), col=pal1[1], lwd=2, ylim=c(0,1.5), las=1, 
     main="", xlab="")
title(main=str_glue("B. Filtered data of {name_file}, ~ {round(log2(cutoff),3)} log-cpm"), xlab="Log-cpm")
abline(v=log2(cutoff), lty=3)
for (i in 2:nsamples){
   den <- density(lcpm[,i])
   lines(den$x, den$y, col=pal1[i], lwd=2)
}
legend("topright", samplenames, text.col=pal1, bty="n")


# 2.2.2 histogram--------------------------------------------------
AveLogCpm_Raw_Data <- aveLogCPM(x)
AveLogCpm_Filtered_Data <-aveLogCPM(x1)

hist(AveLogCpm_Raw_Data)
hist(AveLogCpm_Filtered_Data)

# 2.2.3 normalization---------------------------------------------------------
x1_TMM <- calcNormFactors(x1, method = "TMM")
x1_RLE <- calcNormFactors(x1, method = "RLE")
x1_uQ <- calcNormFactors(x1, method = "upperquartile")
x1_TMMwzp <- calcNormFactors(x1, method = "TMMwzp")
x1_voomTMM <- voomWithQualityWeights(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voomWithQualityWeights(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voomWithQualityWeights(x1,design = design,normalize.method = "quantile",plot = TRUE)

#new
x1_voomTMM <- voom(x1_TMM,design = design,plot = TRUE)
x1_voomTMMwzp <- voom(x1_TMMwzp,design = design,plot = TRUE)
x1_voomQ <- voom(x1,design = design,normalize.method = "quantile",plot = TRUE)


lcpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)
lcpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE, log=TRUE, prior.count=5)

cpm_TMM <- cpm(x1_TMM,normalized.lib.sizes = TRUE)
cpm_TMMwzp <- cpm(x1_TMMwzp,normalized.lib.sizes = TRUE)
cpm_uQ<- cpm(x1_uQ,normalized.lib.sizes = TRUE)
cpm_RLE <- cpm(x1_RLE,normalized.lib.sizes = TRUE)

cpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMM_{name_file}_{date}.txt"),sep = "\t",col.names = TRUE)
lcpm_TMM %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("lcpm_TMM_{name_file}_{date}.txt"),sep = "\t",col.names = TRUE)
cpm_TMMwzp %>% as_tibble(rownames = "smallRNA") %>% 
  fwrite(file = str_glue("cpm_TMMwzp_{name_file}.txt"),sep = "\t",col.names = TRUE)


write_rds(x1_voomQ,str_glue("x1_voomQ_{name_file}_{date}.rds"))
write_rds(x1_voomTMM,str_glue("x1_voomTMM_{name_file}_{date}.rds"))
write_rds(x2_voomTMM,str_glue("x2_voomTMM_{name_file}_{date}.rds"))
x1_voomQ <- read_rds(str_glue("x1_voomQ_filt2_{name_file}.rds"))
x1_TMM %>% write_rds(str_glue("x1_TMM_{name_file}_{date}.rds"))
# 2.2.4 boxplots---------------------------------------------------
par(mfrow=c(1,3))
par(mar=c(6,5,2,1)+ 0.1)

#raw data
lcpm <- cpm(x1,log=TRUE,prior.count=5)
boxplot(lcpm, las=2, col= c_group, main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#TMM data
boxplot(lcpm_TMM, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. TMM Normalized filtered data",ylab="Log-cpm")
#TMMwzp data
boxplot(lcpm_TMMwzp, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. TMMwzp Normalized filtered data",ylab="Log-cpm")
#uQ data
boxplot(lcpm_uQ, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. uQ Normalized filtered data",ylab="Log-cpm")
#RLE data
boxplot(lcpm_RLE, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. RLE Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= rep(wes_palettes$Royal1[1:3],each=2,2), main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")

par(mfrow=c(1,4))
#raw data
boxplot(lcpm, las=2, col= c_group, main="",names=group)
title(main="A. Not Normalized filtered data",ylab="Log-cpm")
#voomTMM data
boxplot(x1_voomTMM$E, las=2, col= c_group, main="",names=group)
title(main="A. Voom TMM Normalized filtered data",ylab="Log-cpm")
#voomTMMwzp data
boxplot(x1_voomTMMwzp$E, las=2, col= c_group, main="",names=group)
title(main="A. Voom TMMwzp Normalized filtered data",ylab="Log-cpm")
#voomQ data
boxplot(x1_voomQ$E, las=2, col= c_group, main="",names=group)
title(main="A. Voom Quantile Normalized filtered data",ylab="Log-cpm")

## save the matrices
#lcpm_TMM %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMM_logcpm_{name_file}.txt"),delim = "\t")
#lcpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_logcpm_{name_file}.txt"),delim = "\t")
#cpm_TMMwzp %>% 
  as_tibble(rownames =  "smallRNA") %>% 
  write_delim(path = str_glue("TMMwzp_cpm_{name_file}.txt"),delim = "\t")
#x1_voomTMMwzp
x1_voomTMMwzp$E %>% as_tibble(rownames="smallRNA") %>% 
  write_delim(path = str_glue("x1_voomTMMwzp_lcpm_{name_file}.txt"),delim = "\t")
  
# 2.2.5 Hierarchical, clustering, allRNA-----------------------------------------
par(mfrow=c(1,3))

d <- dist(t(x1_voomQ$E))

hc <- hclust(d,method = "ward.D2")

myplclust(hc,lab.col = c_group,
          labels = rownames(x1_TMM$samples),xlab = "euclidean-ward.D2",main="")

hc <- hclust(d,method = "complete")
myplclust(hc,lab.col = c_group,
          labels = rownames(x1_TMM$samples),xlab = "euclidean-complete",
          main="Hierarchical clustering with all smallRNAs, voomQ")

hc <- hclust(d,method = "average")
myplclust(hc,lab.col = c_group,
          labels = rownames(x1_TMM$samples),xlab = "euclidean-average",main="")

legend("topright", legend = levels(group),
       fill = c_group[4:5],
       bty="n",
       cex = 1.5, inset = c(.01,.09))
# 2.2.6 MDS plot, all RNA                                                   ----
par(mfrow=c(1,2))
par(mar=c(6,5,2,1)+ 0.1)

plotMDS(x1_voomQ$E,labels = group,
        col = c_group,
        dim.plot = c(1,2))

legend("left",legend = levels(group),
       fill = c_group[4:5],
       bty="n",
       cex = 1.5, inset = c(.01,.09) )

plotMDS(x1_voomQ$E,labels = x1$samples$group,
                col = c_group,dim.plot = c(3,4),
        main=" MDS plot of normal, tumour tissue with voom Quantile \n normalized counts")
# 3.1  DE analysis with limma                                               ----
vfit <- lmFit(x1_voomTMM, design = design)
# check if to use trend = TRUE
max(x1_voomTMM$targets$lib.size)/min(x1_voomTMM$targets$lib.size)>3

# if TRUE (>3) don't use trend = TRUE
# apply eBayes 
efit <- eBayes(vfit,robust = TRUE)
plotSA(efit,main=str_glue("Final model: Mean-variance trend"))
summary(decideTests(efit,p.value = 0.05))
summary(decideTests(efit,p.value = 0.05,lfc=2))
summary(decideTests(efit,p.value = 0.01))
summary(decideTests(efit,p.value = 0.01,lfc=2))

#tfit <- treat(vfit,lfc = 2)
#summary(decideTests(tfit))
  
write_rds(efit,str_glue("efit_x1_voomTMM_{name_file}.rds"))

efit <- read_rds(str_glue("efit_x1_voomTMM_{name_file}.rds"))

#treated-non_treated
top_cancer_vs_control <- topTable(efit,coef = "cancer",
                         number = nrow(efit),
                         adjust.method = "fdr",
                         sort.by = "p") %>% 
  as_tibble(rownames = "smallRNA")

top_cancer_vs_control %>% 
  filter(adj.P.Val<0.05,abs(logFC)>2) %>%
  select(smallRNA,logFC) -> LFCstable

write_delim(top_cancer_vs_control,path=str_glue("LFCs_voomTMM_filt2_treatvsnon_{nafi}.txt"),delim = "\t")

```


